// @wordpress/block-library/build-module/tabs/edit.mjs
// packages/block-library/src/tabs/edit.js
import {
  useBlockProps,
  useInnerBlocksProps,
  BlockContextProvider,
  store as blockEditorStore
} from "@wordpress/block-editor";
import { useSelect, useDispatch } from "@wordpress/data";
import { useMemo, useEffect, useRef } from "@wordpress/element";
import Controls from "./controls.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
var TABS_TEMPLATE = [
  [
    "core/tabs-menu",
    {
      lock: {
        remove: true
      }
    },
    [
      ["core/tabs-menu-item", { anchor: "tab-1-button" }],
      ["core/tabs-menu-item", { anchor: "tab-2-button" }]
    ]
  ],
  [
    "core/tab-panel",
    {
      lock: {
        remove: true
      }
    },
    [
      [
        "core/tab",
        {
          anchor: "tab-1",
          label: "Tab 1"
        },
        [["core/paragraph"]]
      ],
      [
        "core/tab",
        {
          anchor: "tab-2",
          label: "Tab 2"
        },
        [["core/paragraph"]]
      ]
    ]
  ]
];
function Edit({
  clientId,
  attributes,
  setAttributes,
  __unstableLayoutClassNames: layoutClassNames
}) {
  const { anchor, activeTabIndex, editorActiveTabIndex } = attributes;
  useEffect(() => {
    if (editorActiveTabIndex === void 0) {
      setAttributes({ editorActiveTabIndex: activeTabIndex });
    }
  }, []);
  const { removeBlock } = useDispatch(blockEditorStore);
  const { tabs, menuItems } = useSelect(
    (select) => {
      const { getBlocks } = select(blockEditorStore);
      const innerBlocks = getBlocks(clientId);
      const tabPanel = innerBlocks.find(
        (block) => block.name === "core/tab-panel"
      );
      const tabsMenu = innerBlocks.find(
        (block) => block.name === "core/tabs-menu"
      );
      return {
        tabs: tabPanel ? tabPanel.innerBlocks.filter(
          (block) => block.name === "core/tab"
        ) : [],
        menuItems: tabsMenu ? getBlocks(tabsMenu.clientId).filter((b) => b.name === "core/tabs-menu-item").map((b) => ({
          clientId: b.clientId,
          anchor: b.attributes.anchor ?? ""
        })) : []
      };
    },
    [clientId]
  );
  const prevSyncStateRef = useRef(null);
  useEffect(() => {
    const currentTabs = tabs.map((tab) => ({
      clientId: tab.clientId,
      anchor: tab.attributes.anchor ?? ""
    }));
    if (prevSyncStateRef.current === null) {
      prevSyncStateRef.current = {
        tabs: currentTabs,
        menuItems: [...menuItems]
      };
      return;
    }
    const { tabs: prevTabs, menuItems: prevMenuItems } = prevSyncStateRef.current;
    const tabsRemoved = currentTabs.length < prevTabs.length;
    const menuItemsRemoved = menuItems.length < prevMenuItems.length;
    prevSyncStateRef.current = {
      tabs: currentTabs,
      menuItems: [...menuItems]
    };
    if (!tabsRemoved && !menuItemsRemoved || tabsRemoved && menuItemsRemoved) {
      return;
    }
    const currentTabIds = new Set(currentTabs.map((t) => t.clientId));
    const currentMenuItemIds = new Set(
      menuItems.map((m) => m.clientId)
    );
    if (tabsRemoved) {
      prevTabs.forEach((prevTab) => {
        if (currentTabIds.has(prevTab.clientId)) {
          return;
        }
        const expectedMenuAnchor = prevTab.anchor ? `${prevTab.anchor}-button` : null;
        const menuItemToRemove = expectedMenuAnchor ? menuItems.find((m) => m.anchor === expectedMenuAnchor) : null;
        if (menuItemToRemove) {
          removeBlock(menuItemToRemove.clientId, false);
          prevSyncStateRef.current.menuItems = prevSyncStateRef.current.menuItems.filter(
            (m) => m.clientId !== menuItemToRemove.clientId
          );
        }
      });
    } else {
      prevMenuItems.forEach((prevItem) => {
        if (currentMenuItemIds.has(prevItem.clientId)) {
          return;
        }
        const expectedTabAnchor = prevItem.anchor?.replace(/-button$/, "") ?? "";
        const tabToRemove = tabs.find(
          (tab) => (tab.attributes.anchor ?? "") === expectedTabAnchor
        );
        if (tabToRemove) {
          removeBlock(tabToRemove.clientId, false);
          prevSyncStateRef.current.tabs = prevSyncStateRef.current.tabs.filter(
            (t) => t.clientId !== tabToRemove.clientId
          );
        }
      });
    }
  }, [tabs, menuItems, removeBlock]);
  const contextValue = useMemo(() => {
    const tabList = tabs.map((tab, index) => ({
      id: tab.attributes.anchor || `tab-${index}`,
      label: tab.attributes.label || "",
      clientId: tab.clientId,
      index
    }));
    return {
      "core/tabs-list": tabList,
      "core/tabs-id": anchor,
      "core/tabs-activeTabIndex": activeTabIndex,
      "core/tabs-editorActiveTabIndex": editorActiveTabIndex
    };
  }, [tabs, anchor, activeTabIndex, editorActiveTabIndex]);
  const blockProps = useBlockProps({
    className: layoutClassNames
  });
  const innerBlockProps = useInnerBlocksProps(blockProps, {
    __experimentalCaptureToolbars: true,
    template: TABS_TEMPLATE,
    templateLock: false,
    renderAppender: false
  });
  return /* @__PURE__ */ jsx(BlockContextProvider, { value: contextValue, children: /* @__PURE__ */ jsxs("div", { ...innerBlockProps, children: [
    /* @__PURE__ */ jsx(
      Controls,
      {
        clientId,
        attributes,
        setAttributes
      }
    ),
    innerBlockProps.children
  ] }) });
}
var edit_default = Edit;
export {
  edit_default as default
};
