// @wordpress/core-data@7.42.0/build/queried-data/actions.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/queried-data/actions.js
var actions_exports = {};
__export(actions_exports, {
  receiveItems: () => receiveItems,
  receiveQueriedItems: () => receiveQueriedItems,
  removeItems: () => removeItems
});
module.exports = __toCommonJS(actions_exports);
function receiveItems(items, edits, meta) {
  return {
    type: "RECEIVE_ITEMS",
    items,
    persistedEdits: edits,
    meta
  };
}
function removeItems(kind, name, records, invalidateCache = false) {
  return {
    type: "REMOVE_ITEMS",
    itemIds: Array.isArray(records) ? records : [records],
    kind,
    name,
    invalidateCache
  };
}
function receiveQueriedItems(items, query = {}, edits, meta) {
  return {
    ...receiveItems(items, edits, meta),
    query
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  receiveItems,
  receiveQueriedItems,
  removeItems
});
