// @wordpress/core-data@7.42.0/build/queried-data/get-query-parts.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/queried-data/get-query-parts.js
var get_query_parts_exports = {};
__export(get_query_parts_exports, {
  default: () => get_query_parts_default,
  getQueryParts: () => getQueryParts
});
module.exports = __toCommonJS(get_query_parts_exports);
var import_url = require("@wordpress/url");
var import_utils = require("../utils/index.cjs");
function getQueryParts(query) {
  const parts = {
    stableKey: "",
    page: 1,
    perPage: 10,
    offset: void 0,
    fields: null,
    include: null,
    context: "default"
  };
  const keys = Object.keys(query).sort();
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    let value = query[key];
    switch (key) {
      case "page":
        parts[key] = Number(value);
        break;
      case "per_page":
        parts.perPage = Number(value);
        break;
      case "context":
        parts.context = value;
        break;
      default:
        if (key === "offset") {
          const numericOffset = Number(value);
          if (Number.isFinite(numericOffset)) {
            parts.offset = numericOffset;
          }
        }
        if (key === "_fields") {
          parts.fields = (0, import_utils.getNormalizedCommaSeparable)(value) ?? [];
          value = parts.fields.join();
        }
        if (key === "include") {
          if (typeof value === "number") {
            value = value.toString();
          }
          parts.include = ((0, import_utils.getNormalizedCommaSeparable)(value) ?? []).map(Number);
          value = parts.include.join();
        }
        parts.stableKey += (parts.stableKey ? "&" : "") + (0, import_url.addQueryArgs)("", { [key]: value }).slice(1);
    }
  }
  return parts;
}
var get_query_parts_default = (0, import_utils.withWeakMapCache)(getQueryParts);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getQueryParts
});
