// .ignored/@wordpress/core-data/build/hooks/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/hooks/index.ts
var hooks_exports = {};
__export(hooks_exports, {
  __experimentalUseEntityRecord: () => import_use_entity_record.__experimentalUseEntityRecord,
  __experimentalUseEntityRecords: () => import_use_entity_records.__experimentalUseEntityRecords,
  __experimentalUseResourcePermissions: () => import_use_resource_permissions.__experimentalUseResourcePermissions,
  useEntityBlockEditor: () => import_use_entity_block_editor.default,
  useEntityId: () => import_use_entity_id.default,
  useEntityProp: () => import_use_entity_prop.default,
  useEntityRecord: () => import_use_entity_record.default,
  useEntityRecords: () => import_use_entity_records.default,
  useResourcePermissions: () => import_use_resource_permissions.default
});
module.exports = __toCommonJS(hooks_exports);
var import_use_entity_record = __toESM(require("./use-entity-record.cjs"));
var import_use_entity_records = __toESM(require("./use-entity-records.cjs"));
var import_use_resource_permissions = __toESM(require("./use-resource-permissions.cjs"));
var import_use_entity_block_editor = __toESM(require("./use-entity-block-editor.cjs"));
var import_use_entity_id = __toESM(require("./use-entity-id.cjs"));
var import_use_entity_prop = __toESM(require("./use-entity-prop.cjs"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  __experimentalUseEntityRecord,
  __experimentalUseEntityRecords,
  __experimentalUseResourcePermissions,
  useEntityBlockEditor,
  useEntityId,
  useEntityProp,
  useEntityRecord,
  useEntityRecords,
  useResourcePermissions
});
