// .ignored/@wordpress/core-data/build/hooks/use-entity-record.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/hooks/use-entity-record.ts
var use_entity_record_exports = {};
__export(use_entity_record_exports, {
  __experimentalUseEntityRecord: () => __experimentalUseEntityRecord,
  default: () => useEntityRecord
});
module.exports = __toCommonJS(use_entity_record_exports);
var import_data = require("@wordpress/data");
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var import_element = require("@wordpress/element");
var import_use_query_select = __toESM(require("./use-query-select.cjs"));
var import__ = require("../index.cjs");
var EMPTY_OBJECT = {};
function useEntityRecord(kind, name, recordId, options = { enabled: true }) {
  const { editEntityRecord, saveEditedEntityRecord } = (0, import_data.useDispatch)(import__.store);
  const mutations = (0, import_element.useMemo)(
    () => ({
      edit: (record2, editOptions = {}) => editEntityRecord(kind, name, recordId, record2, editOptions),
      save: (saveOptions = {}) => saveEditedEntityRecord(kind, name, recordId, {
        throwOnError: true,
        ...saveOptions
      })
    }),
    [editEntityRecord, kind, name, recordId, saveEditedEntityRecord]
  );
  const { editedRecord, hasEdits, edits } = (0, import_data.useSelect)(
    (select) => {
      if (!options.enabled) {
        return {
          editedRecord: EMPTY_OBJECT,
          hasEdits: false,
          edits: EMPTY_OBJECT
        };
      }
      return {
        editedRecord: select(import__.store).getEditedEntityRecord(
          kind,
          name,
          recordId
        ),
        hasEdits: select(import__.store).hasEditsForEntityRecord(
          kind,
          name,
          recordId
        ),
        edits: select(import__.store).getEntityRecordNonTransientEdits(
          kind,
          name,
          recordId
        )
      };
    },
    [kind, name, recordId, options.enabled]
  );
  const { data: record, ...querySelectRest } = (0, import_use_query_select.default)(
    (query) => {
      if (!options.enabled) {
        return {
          data: null
        };
      }
      return query(import__.store).getEntityRecord(kind, name, recordId);
    },
    [kind, name, recordId, options.enabled]
  );
  return {
    record,
    editedRecord,
    hasEdits,
    edits,
    ...querySelectRest,
    ...mutations
  };
}
function __experimentalUseEntityRecord(kind, name, recordId, options) {
  (0, import_deprecated.default)(`wp.data.__experimentalUseEntityRecord`, {
    alternative: "wp.data.useEntityRecord",
    since: "6.1"
  });
  return useEntityRecord(kind, name, recordId, options);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  __experimentalUseEntityRecord
});
