// @wordpress/viewport/build/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/viewport/src/index.js
var index_exports = {};
__export(index_exports, {
  ifViewportMatches: () => import_if_viewport_matches.default,
  store: () => import_store.store,
  withViewportMatch: () => import_with_viewport_match.default
});
module.exports = __toCommonJS(index_exports);
var import_listener = __toESM(require("./listener.cjs"));
var import_store = require("./store/index.cjs");
var import_if_viewport_matches = __toESM(require("./if-viewport-matches.cjs"));
var import_with_viewport_match = __toESM(require("./with-viewport-match.cjs"));
var BREAKPOINTS = {
  huge: 1440,
  wide: 1280,
  large: 960,
  medium: 782,
  small: 600,
  mobile: 480
};
var OPERATORS = {
  "<": "max-width",
  ">=": "min-width"
};
(0, import_listener.default)(BREAKPOINTS, OPERATORS);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ifViewportMatches,
  store,
  withViewportMatch
});
