// @wordpress/sync@1.42.0/build/providers/http-polling/config.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/sync/src/providers/http-polling/config.ts
var config_exports = {};
__export(config_exports, {
  DEFAULT_CLIENT_LIMIT_PER_ROOM: () => DEFAULT_CLIENT_LIMIT_PER_ROOM,
  MAX_ERROR_BACKOFF_IN_MS: () => MAX_ERROR_BACKOFF_IN_MS,
  MAX_UPDATE_SIZE_IN_BYTES: () => MAX_UPDATE_SIZE_IN_BYTES,
  POLLING_INTERVAL_BACKGROUND_TAB_IN_MS: () => POLLING_INTERVAL_BACKGROUND_TAB_IN_MS,
  POLLING_INTERVAL_IN_MS: () => POLLING_INTERVAL_IN_MS,
  POLLING_INTERVAL_WITH_COLLABORATORS_IN_MS: () => POLLING_INTERVAL_WITH_COLLABORATORS_IN_MS
});
module.exports = __toCommonJS(config_exports);
var DEFAULT_CLIENT_LIMIT_PER_ROOM = 3;
var MAX_ERROR_BACKOFF_IN_MS = 30 * 1e3;
var MAX_UPDATE_SIZE_IN_BYTES = 1 * 1024 * 1024;
var POLLING_INTERVAL_IN_MS = 1e3;
var POLLING_INTERVAL_WITH_COLLABORATORS_IN_MS = 250;
var POLLING_INTERVAL_BACKGROUND_TAB_IN_MS = 25 * 1e3;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_CLIENT_LIMIT_PER_ROOM,
  MAX_ERROR_BACKOFF_IN_MS,
  MAX_UPDATE_SIZE_IN_BYTES,
  POLLING_INTERVAL_BACKGROUND_TAB_IN_MS,
  POLLING_INTERVAL_IN_MS,
  POLLING_INTERVAL_WITH_COLLABORATORS_IN_MS
});
