// @wordpress/core-data@7.42.0/build/private-actions.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/private-actions.js
var private_actions_exports = {};
__export(private_actions_exports, {
  editMediaEntity: () => editMediaEntity,
  receiveEditorAssets: () => receiveEditorAssets,
  receiveEditorSettings: () => receiveEditorSettings,
  receiveRegisteredPostMeta: () => receiveRegisteredPostMeta,
  setCollaborationSupported: () => setCollaborationSupported
});
module.exports = __toCommonJS(private_actions_exports);
var import_api_fetch = __toESM(require("@wordpress/api-fetch"));
var import_name = require("./name.cjs");
function receiveRegisteredPostMeta(postType, registeredPostMeta) {
  return {
    type: "RECEIVE_REGISTERED_POST_META",
    postType,
    registeredPostMeta
  };
}
var editMediaEntity = (recordId, edits = {}, { __unstableFetch = import_api_fetch.default, throwOnError = false } = {}) => async ({ dispatch, resolveSelect }) => {
  if (!recordId) {
    return;
  }
  const kind = "postType";
  const name = "attachment";
  const configs = await resolveSelect.getEntitiesConfig(kind);
  const entityConfig = configs.find(
    (config) => config.kind === kind && config.name === name
  );
  if (!entityConfig) {
    return;
  }
  const lock = await dispatch.__unstableAcquireStoreLock(
    import_name.STORE_NAME,
    ["entities", "records", kind, name, recordId],
    { exclusive: true }
  );
  let updatedRecord;
  let error;
  let hasError = false;
  try {
    dispatch({
      type: "SAVE_ENTITY_RECORD_START",
      kind,
      name,
      recordId
    });
    try {
      const path = `${entityConfig.baseURL}/${recordId}/edit`;
      const newRecord = await __unstableFetch({
        path,
        method: "POST",
        data: {
          ...edits
        }
      });
      if (newRecord) {
        dispatch.receiveEntityRecords(
          kind,
          name,
          newRecord,
          void 0,
          true,
          void 0,
          void 0
        );
        updatedRecord = newRecord;
      }
    } catch (e) {
      error = e;
      hasError = true;
    }
    dispatch({
      type: "SAVE_ENTITY_RECORD_FINISH",
      kind,
      name,
      recordId,
      error
    });
    if (hasError && throwOnError) {
      throw error;
    }
    return updatedRecord;
  } finally {
    dispatch.__unstableReleaseStoreLock(lock);
  }
};
function receiveEditorSettings(settings) {
  return {
    type: "RECEIVE_EDITOR_SETTINGS",
    settings
  };
}
function receiveEditorAssets(assets) {
  return {
    type: "RECEIVE_EDITOR_ASSETS",
    assets
  };
}
var setCollaborationSupported = (supported) => ({ dispatch }) => {
  dispatch({ type: "SET_COLLABORATION_SUPPORTED", supported });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  editMediaEntity,
  receiveEditorAssets,
  receiveEditorSettings,
  receiveRegisteredPostMeta,
  setCollaborationSupported
});
