// @wordpress/block-editor/build-module/store/private-keys.mjs
// packages/block-editor/src/store/private-keys.js
var globalStylesDataKey = /* @__PURE__ */ Symbol("globalStylesDataKey");
var globalStylesLinksDataKey = /* @__PURE__ */ Symbol("globalStylesLinks");
var selectBlockPatternsKey = /* @__PURE__ */ Symbol("selectBlockPatternsKey");
var reusableBlocksSelectKey = /* @__PURE__ */ Symbol("reusableBlocksSelect");
var sectionRootClientIdKey = /* @__PURE__ */ Symbol("sectionRootClientIdKey");
var mediaEditKey = /* @__PURE__ */ Symbol("mediaEditKey");
var getMediaSelectKey = /* @__PURE__ */ Symbol("getMediaSelect");
var isIsolatedEditorKey = /* @__PURE__ */ Symbol("isIsolatedEditor");
var deviceTypeKey = /* @__PURE__ */ Symbol("deviceTypeKey");
var isNavigationOverlayContextKey = /* @__PURE__ */ Symbol(
  "isNavigationOverlayContext"
);
var isNavigationPostEditorKey = /* @__PURE__ */ Symbol("isNavigationPostEditor");
var mediaUploadOnSuccessKey = /* @__PURE__ */ Symbol("mediaUploadOnSuccess");
export {
  deviceTypeKey,
  getMediaSelectKey,
  globalStylesDataKey,
  globalStylesLinksDataKey,
  isIsolatedEditorKey,
  isNavigationOverlayContextKey,
  isNavigationPostEditorKey,
  mediaEditKey,
  mediaUploadOnSuccessKey,
  reusableBlocksSelectKey,
  sectionRootClientIdKey,
  selectBlockPatternsKey
};
