// @wordpress/fields/build/fields/format/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/fields/src/fields/format/index.ts
var format_exports = {};
__export(format_exports, {
  default: () => format_default
});
module.exports = __toCommonJS(format_exports);
var import_i18n = require("@wordpress/i18n");
var import_data = require("@wordpress/data");
var import_core_data = require("@wordpress/core-data");
var POST_FORMATS = [
  { id: "aside", caption: (0, import_i18n.__)("Aside") },
  { id: "audio", caption: (0, import_i18n.__)("Audio") },
  { id: "chat", caption: (0, import_i18n.__)("Chat") },
  { id: "gallery", caption: (0, import_i18n.__)("Gallery") },
  { id: "image", caption: (0, import_i18n.__)("Image") },
  { id: "link", caption: (0, import_i18n.__)("Link") },
  { id: "quote", caption: (0, import_i18n.__)("Quote") },
  { id: "standard", caption: (0, import_i18n.__)("Standard") },
  { id: "status", caption: (0, import_i18n.__)("Status") },
  { id: "video", caption: (0, import_i18n.__)("Video") }
].sort((a, b) => {
  const normalizedA = a.caption.toUpperCase();
  const normalizedB = b.caption.toUpperCase();
  if (normalizedA < normalizedB) {
    return -1;
  }
  if (normalizedA > normalizedB) {
    return 1;
  }
  return 0;
});
var formatField = {
  id: "format",
  label: (0, import_i18n.__)("Format"),
  type: "text",
  Edit: "radio",
  enableSorting: false,
  enableHiding: false,
  filterBy: false,
  getElements: async () => {
    const themeSupports = await (0, import_data.resolveSelect)(import_core_data.store).getThemeSupports();
    return POST_FORMATS.filter(
      (f) => themeSupports?.formats?.includes(
        f.id
      )
    ).map((f) => ({ value: f.id, label: f.caption }));
  }
};
var format_default = formatField;
