// @wordpress/ui@0.9.0/build/form/primitives/field/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/form/primitives/field/index.ts
var field_exports = {};
__export(field_exports, {
  Control: () => import_control.Control,
  Description: () => import_description.Description,
  Details: () => import_details.Details,
  Item: () => import_item.Item,
  Label: () => import_label.Label,
  Root: () => import_root.Root
});
module.exports = __toCommonJS(field_exports);
var import_root = require("./root.cjs");
var import_item = require("./item.cjs");
var import_label = require("./label.cjs");
var import_description = require("./description.cjs");
var import_details = require("./details.cjs");
var import_control = require("./control.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Control,
  Description,
  Details,
  Item,
  Label,
  Root
});
