// @wordpress/block-library@9.42.0/build-module/column/deprecated.mjs
// packages/block-library/src/column/deprecated.js
import clsx from "clsx";
import { InnerBlocks } from "@wordpress/block-editor";
import { jsx } from "react/jsx-runtime";
var deprecated = [
  {
    attributes: {
      verticalAlignment: {
        type: "string"
      },
      width: {
        type: "number",
        min: 0,
        max: 100
      }
    },
    isEligible({ width }) {
      return isFinite(width);
    },
    migrate(attributes) {
      return {
        ...attributes,
        width: `${attributes.width}%`
      };
    },
    save({ attributes }) {
      const { verticalAlignment, width } = attributes;
      const wrapperClasses = clsx({
        [`is-vertically-aligned-${verticalAlignment}`]: verticalAlignment
      });
      const style = { flexBasis: width + "%" };
      return /* @__PURE__ */ jsx("div", { className: wrapperClasses, style, children: /* @__PURE__ */ jsx(InnerBlocks.Content, {}) });
    }
  }
];
var deprecated_default = deprecated;
export {
  deprecated_default as default
};
