// @wordpress/block-editor/build/components/font-sizes/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/font-sizes/index.js
var font_sizes_exports = {};
__export(font_sizes_exports, {
  FontSizePicker: () => import_font_size_picker.default,
  getComputedFluidTypographyValue: () => import_fluid_utils.getComputedFluidTypographyValue,
  getFontSize: () => import_utils.getFontSize,
  getFontSizeClass: () => import_utils.getFontSizeClass,
  getFontSizeObjectByValue: () => import_utils.getFontSizeObjectByValue,
  withFontSizes: () => import_with_font_sizes.default
});
module.exports = __toCommonJS(font_sizes_exports);
var import_utils = require("./utils.cjs");
var import_fluid_utils = require("./fluid-utils.cjs");
var import_font_size_picker = __toESM(require("./font-size-picker.cjs"));
var import_with_font_sizes = __toESM(require("./with-font-sizes.cjs"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FontSizePicker,
  getComputedFluidTypographyValue,
  getFontSize,
  getFontSizeClass,
  getFontSizeObjectByValue,
  withFontSizes
});
