// @wordpress/block-library/build/latest-comments/deprecated.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/latest-comments/deprecated.js
var deprecated_exports = {};
__export(deprecated_exports, {
  default: () => deprecated_default
});
module.exports = __toCommonJS(deprecated_exports);
var v1 = {
  attributes: {
    commentsToShow: {
      type: "number",
      default: 5,
      minimum: 1,
      maximum: 100
    },
    displayAvatar: {
      type: "boolean",
      default: true
    },
    displayDate: {
      type: "boolean",
      default: true
    },
    displayExcerpt: {
      type: "boolean",
      default: true
    }
  },
  isEligible(attributes) {
    return attributes?.displayExcerpt === false;
  },
  migrate(attributes) {
    return {
      ...attributes,
      displayContent: attributes.displayExcerpt ? "excerpt" : "none"
    };
  }
};
var deprecated_default = [v1];
