// @wordpress/block-editor@15.15.0/build/components/rich-text/format-toolbar-container.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/rich-text/format-toolbar-container.js
var format_toolbar_container_exports = {};
__export(format_toolbar_container_exports, {
  default: () => format_toolbar_container_default
});
module.exports = __toCommonJS(format_toolbar_container_exports);
var import_i18n = require("@wordpress/i18n");
var import_components = require("@wordpress/components");
var import_block_controls = __toESM(require("../block-controls/index.cjs"));
var import_format_toolbar = __toESM(require("./format-toolbar/index.cjs"));
var import_navigable_toolbar = __toESM(require("../navigable-toolbar/index.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function InlineToolbar({ popoverAnchor }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Popover,
    {
      placement: "top",
      focusOnMount: false,
      anchor: popoverAnchor,
      className: "block-editor-rich-text__inline-format-toolbar",
      __unstableSlotName: "block-toolbar",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_navigable_toolbar.default,
        {
          className: "block-editor-rich-text__inline-format-toolbar-group",
          "aria-label": (0, import_i18n.__)("Format tools"),
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.ToolbarGroup, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_format_toolbar.default, {}) })
        }
      )
    }
  );
}
var FormatToolbarContainer = ({ inline, editableContentElement }) => {
  if (inline) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(InlineToolbar, { popoverAnchor: editableContentElement });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_block_controls.default, { group: "inline", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_format_toolbar.default, {}) });
};
var format_toolbar_container_default = FormatToolbarContainer;
