// @wordpress/ui/build/dialog/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/dialog/index.ts
var dialog_exports = {};
__export(dialog_exports, {
  Action: () => import_action.Action,
  CloseIcon: () => import_close_icon.CloseIcon,
  Footer: () => import_footer.Footer,
  Header: () => import_header.Header,
  Popup: () => import_popup.Popup,
  Root: () => import_root.Root,
  Title: () => import_title.Title,
  Trigger: () => import_trigger.Trigger
});
module.exports = __toCommonJS(dialog_exports);
var import_action = require("./action.cjs");
var import_close_icon = require("./close-icon.cjs");
var import_footer = require("./footer.cjs");
var import_header = require("./header.cjs");
var import_popup = require("./popup.cjs");
var import_root = require("./root.cjs");
var import_title = require("./title.cjs");
var import_trigger = require("./trigger.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Action,
  CloseIcon,
  Footer,
  Header,
  Popup,
  Root,
  Title,
  Trigger
});
