// @wordpress/block-editor@15.15.0/build/components/global-styles/state-control.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/global-styles/state-control.js
var state_control_exports = {};
__export(state_control_exports, {
  default: () => StateControl
});
module.exports = __toCommonJS(state_control_exports);
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var import_components = require("@wordpress/components");
var import_jsx_runtime = require("react/jsx-runtime");
function StateControl({
  states = [],
  value = "default",
  onChange
}) {
  if (!states || states.length === 0) {
    return null;
  }
  const stateOptions = [
    { label: (0, import_i18n.__)("Default"), value: "default" },
    ...states.map((state) => ({
      label: state.label,
      value: state.value
    }))
  ];
  const getCurrentStateLabel = () => {
    const currentOption = stateOptions.find(
      (option) => option.value === value
    );
    return currentOption?.label || (0, import_i18n.__)("Default");
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.DropdownMenu,
    {
      icon: import_icons.chevronDown,
      label: (0, import_i18n.sprintf)(
        /* translators: %s: Current state (e.g. "Hover", "Focus") */
        (0, import_i18n.__)("State: %s"),
        getCurrentStateLabel()
      ),
      text: getCurrentStateLabel(),
      toggleProps: {
        size: "compact",
        variant: "tertiary",
        iconPosition: "right"
      },
      children: ({ onClose }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.MenuGroup, { label: (0, import_i18n.__)("State"), children: stateOptions.map((option) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.MenuItem,
        {
          onClick: () => {
            onChange(option.value);
            onClose();
          },
          icon: value === option.value ? import_icons.check : null,
          children: option.label
        },
        option.value
      )) })
    }
  );
}
