// @wordpress/ui@0.9.0/build-module/form/primitives/select/trigger.mjs
// packages/ui/src/form/primitives/select/trigger.tsx
import { Select as _Select } from "@base-ui/react/select";
import clsx from "clsx";
import { forwardRef } from "@wordpress/element";
import { chevronDown } from "@wordpress/icons";

// packages/ui/src/utils/css/focus.module.css
if (typeof document !== "undefined" && process.env.NODE_ENV !== "test" && !document.head.querySelector("style[data-wp-hash='6db1282653']")) {
  const style = document.createElement("style");
  style.setAttribute("data-wp-hash", "6db1282653");
  style.appendChild(document.createTextNode("@layer wp-ui-utilities, wp-ui-components, wp-ui-compositions, wp-ui-overrides;@layer wp-ui-utilities{._08e8a2e44959f892__outset-ring--focus,._970d04df7376df67__outset-ring--focus-within-except-active,.c5cb3ee4bddaa8e4__outset-ring--focus-within-visible,.cd83dfc2126a0846__outset-ring--focus-within,.d0541bc9dd9dc7b6__outset-ring--focus-visible,.e25b2bdd7aa21721__outset-ring--focus-except-active,.ecadb9e080e2dfa5__outset-ring--focus-parent-visible{@media not (prefers-reduced-motion){transition:outline .1s ease-out}outline-color:#0000;outline-offset:1px;outline-style:solid;outline-width:0}._08e8a2e44959f892__outset-ring--focus:focus,._970d04df7376df67__outset-ring--focus-within-except-active:focus-within:not(:has(:active)),.c5cb3ee4bddaa8e4__outset-ring--focus-within-visible:focus-within:has(:focus-visible),.cd83dfc2126a0846__outset-ring--focus-within:focus-within,.d0541bc9dd9dc7b6__outset-ring--focus-visible:focus-visible,.e25b2bdd7aa21721__outset-ring--focus-except-active:focus:not(:active),:focus-visible .ecadb9e080e2dfa5__outset-ring--focus-parent-visible{outline-color:var(--wpds-color-stroke-focus-brand,var(--wp-admin-theme-color,#3858e9));outline-width:var(--wpds-border-width-focus,var(--wp-admin-border-width-focus,2px))}}"));
  document.head.appendChild(style);
}
var focus_default = { "outset-ring--focus": "_08e8a2e44959f892__outset-ring--focus", "outset-ring--focus-except-active": "e25b2bdd7aa21721__outset-ring--focus-except-active", "outset-ring--focus-visible": "d0541bc9dd9dc7b6__outset-ring--focus-visible", "outset-ring--focus-within": "cd83dfc2126a0846__outset-ring--focus-within", "outset-ring--focus-within-except-active": "_970d04df7376df67__outset-ring--focus-within-except-active", "outset-ring--focus-within-visible": "c5cb3ee4bddaa8e4__outset-ring--focus-within-visible", "outset-ring--focus-parent-visible": "ecadb9e080e2dfa5__outset-ring--focus-parent-visible" };

// packages/ui/src/utils/css/select-trigger.module.css
if (typeof document !== "undefined" && process.env.NODE_ENV !== "test" && !document.head.querySelector("style[data-wp-hash='fe532ee577']")) {
  const style = document.createElement("style");
  style.setAttribute("data-wp-hash", "fe532ee577");
  style.appendChild(document.createTextNode("@layer wp-ui-utilities, wp-ui-components, wp-ui-compositions, wp-ui-overrides;@layer wp-ui-utilities{._6e4e0445ef20426b__trigger-wrapper{&.fa98c865fb77e675__is-minimal{width:fit-content}}._7893295e3f6d1af7__trigger{align-items:center;background-color:initial;border:none;color:var(--wpds-color-fg-interactive-neutral,#1e1e1e);cursor:var(--wpds-cursor-control,default);display:flex;font-family:inherit;font-size:inherit;gap:var(--wpds-dimension-gap-xs,4px);justify-content:space-between;line-height:1.4;padding-block:4px;padding-inline-end:calc(var(--wp-ui-input-layout-padding-inline) - 4px);padding-inline-start:var(--wp-ui-input-layout-padding-inline);text-align:start;user-select:none;width:100%;&.fa98c865fb77e675__is-minimal{width:auto}&:focus{outline:none}&[data-disabled]{background-color:var(--wpds-color-bg-interactive-neutral-weak-disabled,#0000);color:var(--wpds-color-fg-interactive-neutral-disabled,#8a8a8a)}}.e760c9339965ce84__trigger-value{overflow:hidden;text-overflow:ellipsis;white-space:nowrap;&.c67f9ed393747f3f__is-placeholder{color:var(--wpds-color-fg-interactive-neutral-disabled,#8a8a8a)}}._308bc63e43681f99__trigger-caret{flex:0 0 auto}}"));
  document.head.appendChild(style);
}
var select_trigger_default = { "trigger-wrapper": "_6e4e0445ef20426b__trigger-wrapper", "is-minimal": "fa98c865fb77e675__is-minimal", "trigger": "_7893295e3f6d1af7__trigger", "trigger-value": "e760c9339965ce84__trigger-value", "is-placeholder": "c67f9ed393747f3f__is-placeholder", "trigger-caret": "_308bc63e43681f99__trigger-caret" };

// packages/ui/src/form/primitives/select/trigger.tsx
import { InputLayout } from "../input-layout/index.mjs";
import { Icon } from "../../../icon/index.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
var Trigger = forwardRef(
  function Trigger2({ className, size, variant, children, ...restProps }, ref) {
    return /* @__PURE__ */ jsx(
      InputLayout,
      {
        className: clsx(
          focus_default["outset-ring--focus-within-except-active"],
          select_trigger_default["trigger-wrapper"],
          variant === "minimal" && select_trigger_default["is-minimal"],
          className
        ),
        size,
        isBorderless: variant === "minimal",
        children: /* @__PURE__ */ jsxs(
          _Select.Trigger,
          {
            ...restProps,
            className: clsx(
              select_trigger_default.trigger,
              variant === "minimal" && select_trigger_default["is-minimal"]
            ),
            "data-can-disable-input-layout": true,
            ref,
            children: [
              /* @__PURE__ */ jsx(
                _Select.Value,
                {
                  className: (state) => clsx(
                    select_trigger_default["trigger-value"],
                    state.value === "" && select_trigger_default["is-placeholder"]
                  ),
                  children
                }
              ),
              /* @__PURE__ */ jsx(
                Icon,
                {
                  className: select_trigger_default["trigger-caret"],
                  icon: chevronDown,
                  size: 18
                }
              )
            ]
          }
        )
      }
    );
  }
);
export {
  Trigger
};
