// @wordpress/patterns/build/store/actions.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/patterns/src/store/actions.js
var actions_exports = {};
__export(actions_exports, {
  convertSyncedPatternToStatic: () => convertSyncedPatternToStatic,
  createPattern: () => createPattern,
  createPatternFromFile: () => createPatternFromFile,
  setEditingPattern: () => setEditingPattern
});
module.exports = __toCommonJS(actions_exports);
var import_blocks = require("@wordpress/blocks");
var import_core_data = require("@wordpress/core-data");
var import_block_editor = require("@wordpress/block-editor");
var import_constants = require("../constants.cjs");
var createPattern = (title, syncType, content, categories) => async ({ registry }) => {
  const meta = syncType === import_constants.PATTERN_SYNC_TYPES.unsynced ? {
    wp_pattern_sync_status: syncType
  } : void 0;
  const reusableBlock = {
    title,
    content,
    status: "publish",
    meta,
    wp_pattern_category: categories
  };
  const updatedRecord = await registry.dispatch(import_core_data.store).saveEntityRecord("postType", "wp_block", reusableBlock);
  return updatedRecord;
};
var createPatternFromFile = (file, categories) => async ({ dispatch }) => {
  const fileContent = await file.text();
  let parsedContent;
  try {
    parsedContent = JSON.parse(fileContent);
  } catch (e) {
    throw new Error("Invalid JSON file");
  }
  if (parsedContent.__file !== "wp_block" || !parsedContent.title || !parsedContent.content || typeof parsedContent.title !== "string" || typeof parsedContent.content !== "string" || parsedContent.syncStatus && typeof parsedContent.syncStatus !== "string") {
    throw new Error("Invalid pattern JSON file");
  }
  const pattern = await dispatch.createPattern(
    parsedContent.title,
    parsedContent.syncStatus,
    parsedContent.content,
    categories
  );
  return pattern;
};
var convertSyncedPatternToStatic = (clientId) => ({ registry }) => {
  const patternBlock = registry.select(import_block_editor.store).getBlock(clientId);
  const existingOverrides = patternBlock.attributes?.content;
  function cloneBlocksAndRemoveBindings(blocks) {
    return blocks.map((block) => {
      let metadata = block.attributes.metadata;
      if (metadata) {
        metadata = { ...metadata };
        delete metadata.id;
        delete metadata.bindings;
        if (existingOverrides?.[metadata.name]) {
          for (const [attributeName, value] of Object.entries(
            existingOverrides[metadata.name]
          )) {
            if (!(0, import_blocks.getBlockType)(block.name)?.attributes[attributeName]) {
              continue;
            }
            block.attributes[attributeName] = value;
          }
        }
      }
      return (0, import_blocks.cloneBlock)(
        block,
        {
          metadata: metadata && Object.keys(metadata).length > 0 ? metadata : void 0
        },
        cloneBlocksAndRemoveBindings(block.innerBlocks)
      );
    });
  }
  const patternInnerBlocks = registry.select(import_block_editor.store).getBlocks(patternBlock.clientId);
  registry.dispatch(import_block_editor.store).replaceBlocks(
    patternBlock.clientId,
    cloneBlocksAndRemoveBindings(patternInnerBlocks)
  );
};
function setEditingPattern(clientId, isEditing) {
  return {
    type: "SET_EDITING_PATTERN",
    clientId,
    isEditing
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  convertSyncedPatternToStatic,
  createPattern,
  createPatternFromFile,
  setEditingPattern
});
