// @wordpress/block-editor/build/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/index.js
var index_exports = {};
__export(index_exports, {
  SETTINGS_DEFAULTS: () => import_defaults.SETTINGS_DEFAULTS,
  __experimentalGetBorderClassesAndStyles: () => import_hooks2.getBorderClassesAndStyles,
  __experimentalGetColorClassesAndStyles: () => import_hooks2.getColorClassesAndStyles,
  __experimentalGetDimensionsClassesAndStyles: () => import_hooks2.getDimensionsClassesAndStyles,
  __experimentalGetGapCSSValue: () => import_hooks2.getGapCSSValue,
  __experimentalGetShadowClassesAndStyles: () => import_hooks2.getShadowClassesAndStyles,
  __experimentalGetSpacingClassesAndStyles: () => import_hooks2.getSpacingClassesAndStyles,
  __experimentalUseBorderProps: () => import_hooks2.useBorderProps,
  __experimentalUseColorProps: () => import_hooks2.useColorProps,
  __experimentalUseCustomSides: () => import_hooks2.useCustomSides,
  getDimensionsClassesAndStyles: () => import_hooks2.getDimensionsClassesAndStyles,
  getTypographyClassesAndStyles: () => import_hooks2.getTypographyClassesAndStyles,
  privateApis: () => import_private_apis.privateApis,
  store: () => import_store.store,
  storeConfig: () => import_store.storeConfig,
  useCachedTruthy: () => import_hooks2.useCachedTruthy,
  useStyleOverride: () => import_hooks2.useStyleOverride
});
module.exports = __toCommonJS(index_exports);
var import_hooks = require("./hooks/index.cjs");
var import_hooks2 = require("./hooks/index.cjs");
__reExport(index_exports, require("./components/index.cjs"), module.exports);
__reExport(index_exports, require("./elements/index.cjs"), module.exports);
__reExport(index_exports, require("./utils/index.cjs"), module.exports);
var import_store = require("./store/index.cjs");
var import_defaults = require("./store/defaults.cjs");
var import_private_apis = require("./private-apis.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SETTINGS_DEFAULTS,
  __experimentalGetBorderClassesAndStyles,
  __experimentalGetColorClassesAndStyles,
  __experimentalGetDimensionsClassesAndStyles,
  __experimentalGetGapCSSValue,
  __experimentalGetShadowClassesAndStyles,
  __experimentalGetSpacingClassesAndStyles,
  __experimentalUseBorderProps,
  __experimentalUseColorProps,
  __experimentalUseCustomSides,
  getDimensionsClassesAndStyles,
  getTypographyClassesAndStyles,
  privateApis,
  store,
  storeConfig,
  useCachedTruthy,
  useStyleOverride,
  ...require("./components/index.cjs"),
  ...require("./elements/index.cjs"),
  ...require("./utils/index.cjs")
});
