// @wordpress/block-library/build/template-part/edit/title-modal.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/template-part/edit/title-modal.js
var title_modal_exports = {};
__export(title_modal_exports, {
  default: () => TitleModal
});
module.exports = __toCommonJS(title_modal_exports);
var import_element = require("@wordpress/element");
var import_i18n = require("@wordpress/i18n");
var import_components = require("@wordpress/components");
var import_jsx_runtime = require("react/jsx-runtime");
function TitleModal({ areaLabel, onClose, onSubmit }) {
  const [title, setTitle] = (0, import_element.useState)("");
  const submitForCreation = (event) => {
    event.preventDefault();
    onSubmit(title);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Modal,
    {
      title: (0, import_i18n.sprintf)(
        // Translators: %s as template part area title ("Header", "Footer", etc.).
        (0, import_i18n.__)("Create new %s"),
        areaLabel.toLowerCase()
      ),
      onRequestClose: onClose,
      focusOnMount: "firstContentElement",
      size: "small",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("form", { onSubmit: submitForCreation, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalVStack, { spacing: "5", children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.TextControl,
          {
            label: (0, import_i18n.__)("Name"),
            value: title,
            onChange: setTitle,
            placeholder: (0, import_i18n.__)("Custom Template Part"),
            __next40pxDefaultSize: true
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalHStack, { justify: "right", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              __next40pxDefaultSize: true,
              variant: "tertiary",
              onClick: () => {
                onClose();
                setTitle("");
              },
              children: (0, import_i18n.__)("Cancel")
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              variant: "primary",
              type: "submit",
              accessibleWhenDisabled: true,
              disabled: !title.length,
              __next40pxDefaultSize: true,
              children: (0, import_i18n.__)("Create")
            }
          )
        ] })
      ] }) })
    }
  );
}
