// @wordpress/dataviews/build-module/components/dataviews-search/index.mjs
// packages/dataviews/src/components/dataviews-search/index.tsx
import { __ } from "@wordpress/i18n";
import { useEffect, useRef, memo, useContext } from "@wordpress/element";
import { SearchControl } from "@wordpress/components";
import { useDebouncedInput } from "@wordpress/compose";
import DataViewsContext from "../dataviews-context/index.mjs";
import { jsx } from "react/jsx-runtime";
var DataViewsSearch = memo(function Search({ label }) {
  const { view, onChangeView } = useContext(DataViewsContext);
  const [search, setSearch, debouncedSearch] = useDebouncedInput(
    view.search
  );
  useEffect(() => {
    if (view.search !== debouncedSearch) {
      setSearch(view.search ?? "");
    }
  }, [view.search, setSearch]);
  const onChangeViewRef = useRef(onChangeView);
  const viewRef = useRef(view);
  useEffect(() => {
    onChangeViewRef.current = onChangeView;
    viewRef.current = view;
  }, [onChangeView, view]);
  useEffect(() => {
    if (debouncedSearch !== viewRef.current?.search) {
      onChangeViewRef.current({
        ...viewRef.current,
        page: 1,
        search: debouncedSearch
      });
    }
  }, [debouncedSearch]);
  const searchLabel = label || __("Search");
  return /* @__PURE__ */ jsx(
    SearchControl,
    {
      className: "dataviews-search",
      onChange: setSearch,
      value: search,
      label: searchLabel,
      placeholder: searchLabel,
      size: "compact"
    }
  );
});
var dataviews_search_default = DataViewsSearch;
export {
  dataviews_search_default as default
};
