// @wordpress/blocks@15.15.0/build/api/children.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/blocks/src/api/children.js
var children_exports = {};
__export(children_exports, {
  concat: () => concat,
  default: () => children_default,
  fromDOM: () => fromDOM2,
  getSerializeCapableElement: () => getSerializeCapableElement,
  matcher: () => matcher,
  toHTML: () => toHTML
});
module.exports = __toCommonJS(children_exports);
var import_element = require("@wordpress/element");
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var node = __toESM(require("./node.cjs"));
function getSerializeCapableElement(children) {
  return children;
}
function getChildrenArray(children) {
  (0, import_deprecated.default)("wp.blocks.children.getChildrenArray", {
    since: "6.1",
    version: "6.3",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  return children;
}
function concat(...blockNodes) {
  (0, import_deprecated.default)("wp.blocks.children.concat", {
    since: "6.1",
    version: "6.3",
    alternative: "wp.richText.concat",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  const result = [];
  for (let i = 0; i < blockNodes.length; i++) {
    const blockNode = Array.isArray(blockNodes[i]) ? blockNodes[i] : [blockNodes[i]];
    for (let j = 0; j < blockNode.length; j++) {
      const child = blockNode[j];
      const canConcatToPreviousString = typeof child === "string" && typeof result[result.length - 1] === "string";
      if (canConcatToPreviousString) {
        result[result.length - 1] += child;
      } else {
        result.push(child);
      }
    }
  }
  return result;
}
function fromDOM2(domNodes) {
  (0, import_deprecated.default)("wp.blocks.children.fromDOM", {
    since: "6.1",
    version: "6.3",
    alternative: "wp.richText.create",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  const result = [];
  for (let i = 0; i < domNodes.length; i++) {
    try {
      result.push(node.fromDOM(domNodes[i]));
    } catch (error) {
    }
  }
  return result;
}
function toHTML(children) {
  (0, import_deprecated.default)("wp.blocks.children.toHTML", {
    since: "6.1",
    version: "6.3",
    alternative: "wp.richText.toHTMLString",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  const element = getSerializeCapableElement(children);
  return (0, import_element.renderToString)(element);
}
function matcher(selector) {
  (0, import_deprecated.default)("wp.blocks.children.matcher", {
    since: "6.1",
    version: "6.3",
    alternative: "html source",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  return (domNode) => {
    let match = domNode;
    if (selector) {
      match = domNode.querySelector(selector);
    }
    if (match) {
      return fromDOM2(match.childNodes);
    }
    return [];
  };
}
var children_default = {
  concat,
  getChildrenArray,
  fromDOM: fromDOM2,
  toHTML,
  matcher
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  concat,
  fromDOM,
  getSerializeCapableElement,
  matcher,
  toHTML
});
