// @wordpress/dataviews/build-module/components/dataform-controls/datetime.mjs
// packages/dataviews/src/components/dataform-controls/datetime.tsx
import {
  BaseControl,
  privateApis as componentsPrivateApis
} from "@wordpress/components";
import { useCallback, useEffect, useRef, useState } from "@wordpress/element";
import { __ } from "@wordpress/i18n";
import { dateI18n, getDate, getSettings } from "@wordpress/date";
import { Stack } from "@wordpress/ui";
import { OPERATOR_IN_THE_PAST, OPERATOR_OVER } from "../../constants.mjs";
import RelativeDateControl from "./utils/relative-date-control.mjs";
import getCustomValidity from "./utils/get-custom-validity.mjs";
import parseDateTime from "../../field-types/utils/parse-date-time.mjs";
import { unlock } from "../../lock-unlock.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
var { DateCalendar, ValidatedInputControl } = unlock(componentsPrivateApis);
var formatDateTime = (value) => {
  if (!value) {
    return "";
  }
  return dateI18n("Y-m-d\\TH:i", getDate(value));
};
function CalendarDateTimeControl({
  data,
  field,
  onChange,
  hideLabelFromVision,
  markWhenOptional,
  validity
}) {
  const { id, label, description, setValue, getValue, isValid } = field;
  const fieldValue = getValue({ item: data });
  const value = typeof fieldValue === "string" ? fieldValue : void 0;
  const [calendarMonth, setCalendarMonth] = useState(() => {
    const parsedDate = parseDateTime(value);
    return parsedDate || /* @__PURE__ */ new Date();
  });
  const inputControlRef = useRef(null);
  const validationTimeoutRef = useRef(void 0);
  const previousFocusRef = useRef(null);
  const onChangeCallback = useCallback(
    (newValue) => onChange(setValue({ item: data, value: newValue })),
    [data, onChange, setValue]
  );
  useEffect(() => {
    return () => {
      if (validationTimeoutRef.current) {
        clearTimeout(validationTimeoutRef.current);
      }
    };
  }, []);
  const onSelectDate = useCallback(
    (newDate) => {
      let dateTimeValue;
      if (newDate) {
        const wpDate = dateI18n("Y-m-d", newDate);
        let wpTime;
        if (value) {
          wpTime = dateI18n("H:i", getDate(value));
        } else {
          wpTime = dateI18n("H:i", newDate);
        }
        const finalDateTime = getDate(`${wpDate}T${wpTime}`);
        dateTimeValue = finalDateTime.toISOString();
        onChangeCallback(dateTimeValue);
        if (validationTimeoutRef.current) {
          clearTimeout(validationTimeoutRef.current);
        }
      } else {
        onChangeCallback(void 0);
      }
      previousFocusRef.current = inputControlRef.current && inputControlRef.current.ownerDocument.activeElement;
      validationTimeoutRef.current = setTimeout(() => {
        if (inputControlRef.current) {
          inputControlRef.current.focus();
          inputControlRef.current.blur();
          onChangeCallback(dateTimeValue);
          if (previousFocusRef.current && previousFocusRef.current instanceof HTMLElement) {
            previousFocusRef.current.focus();
          }
        }
      }, 0);
    },
    [onChangeCallback, value]
  );
  const handleManualDateTimeChange = useCallback(
    (newValue) => {
      if (newValue) {
        const dateTime = getDate(newValue);
        onChangeCallback(dateTime.toISOString());
        const parsedDate = parseDateTime(dateTime.toISOString());
        if (parsedDate) {
          setCalendarMonth(parsedDate);
        }
      } else {
        onChangeCallback(void 0);
      }
    },
    [onChangeCallback]
  );
  const { format: fieldFormat } = field;
  const weekStartsOn = fieldFormat.weekStartsOn ?? getSettings().l10n.startOfWeek;
  const {
    timezone: { string: timezoneString }
  } = getSettings();
  let displayLabel = label;
  if (isValid?.required && !markWhenOptional && !hideLabelFromVision) {
    displayLabel = `${label} (${__("Required")})`;
  } else if (!isValid?.required && markWhenOptional && !hideLabelFromVision) {
    displayLabel = `${label} (${__("Optional")})`;
  }
  return /* @__PURE__ */ jsx(
    BaseControl,
    {
      id,
      label: displayLabel,
      help: description,
      hideLabelFromVision,
      children: /* @__PURE__ */ jsxs(Stack, { direction: "column", gap: "lg", children: [
        /* @__PURE__ */ jsx(
          ValidatedInputControl,
          {
            ref: inputControlRef,
            __next40pxDefaultSize: true,
            required: !!isValid?.required,
            customValidity: getCustomValidity(isValid, validity),
            type: "datetime-local",
            label: __("Date time"),
            hideLabelFromVision: true,
            value: formatDateTime(value),
            onChange: handleManualDateTimeChange
          }
        ),
        /* @__PURE__ */ jsx(
          DateCalendar,
          {
            style: { width: "100%" },
            selected: value ? parseDateTime(value) || void 0 : void 0,
            onSelect: onSelectDate,
            month: calendarMonth,
            onMonthChange: setCalendarMonth,
            timeZone: timezoneString || void 0,
            weekStartsOn
          }
        )
      ] })
    }
  );
}
function DateTime({
  data,
  field,
  onChange,
  hideLabelFromVision,
  markWhenOptional,
  operator,
  validity
}) {
  if (operator === OPERATOR_IN_THE_PAST || operator === OPERATOR_OVER) {
    return /* @__PURE__ */ jsx(
      RelativeDateControl,
      {
        className: "dataviews-controls__datetime",
        data,
        field,
        onChange,
        hideLabelFromVision,
        operator
      }
    );
  }
  return /* @__PURE__ */ jsx(
    CalendarDateTimeControl,
    {
      data,
      field,
      onChange,
      hideLabelFromVision,
      markWhenOptional,
      validity
    }
  );
}
export {
  DateTime as default
};
