// @wordpress/block-library/build/comments/edit/placeholder.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/comments/edit/placeholder.js
var placeholder_exports = {};
__export(placeholder_exports, {
  default: () => PostCommentsPlaceholder
});
module.exports = __toCommonJS(placeholder_exports);
var import_block_editor = require("@wordpress/block-editor");
var import_i18n = require("@wordpress/i18n");
var import_data = require("@wordpress/data");
var import_core_data = require("@wordpress/core-data");
var import_element = require("@wordpress/element");
var import_form = __toESM(require("../../post-comments-form/form.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function PostCommentsPlaceholder({ postType, postId }) {
  let [postTitle] = (0, import_core_data.useEntityProp)("postType", postType, "title", postId);
  postTitle = postTitle || (0, import_i18n.__)("Post Title");
  const avatarURL = (0, import_data.useSelect)(
    (select) => select(import_block_editor.store).getSettings().__experimentalDiscussionSettings?.avatarURL,
    []
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "wp-block-comments__legacy-placeholder", inert: "true", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", {
      /* translators: %s: Post title. */
      children: (0, import_i18n.sprintf)((0, import_i18n.__)("One response to %s"), postTitle)
    }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "navigation", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "alignleft", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("a", { href: "#top", children: [
        "\xAB ",
        (0, import_i18n.__)("Older Comments")
      ] }) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "alignright", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("a", { href: "#top", children: [
        (0, import_i18n.__)("Newer Comments"),
        " \xBB"
      ] }) })
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("ol", { className: "commentlist", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("li", { className: "comment even thread-even depth-1", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("article", { className: "comment-body", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("footer", { className: "comment-meta", children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "comment-author vcard", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "img",
            {
              alt: (0, import_i18n.__)("Commenter Avatar"),
              src: avatarURL,
              className: "avatar avatar-32 photo",
              height: "32",
              width: "32",
              loading: "lazy"
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("b", { className: "fn", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("a", { href: "#top", className: "url", children: (0, import_i18n.__)("A WordPress Commenter") }) }),
          " ",
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { className: "says", children: [
            (0, import_i18n.__)("says"),
            ":"
          ] })
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "comment-metadata", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("a", { href: "#top", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("time", { dateTime: "2000-01-01T00:00:00+00:00", children: (0, import_i18n.__)("January 1, 2000 at 00:00 am") }) }),
          " ",
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: "edit-link", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "a",
            {
              className: "comment-edit-link",
              href: "#top",
              children: (0, import_i18n.__)("Edit")
            }
          ) })
        ] })
      ] }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "comment-content", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("p", { children: [
        (0, import_i18n.__)("Hi, this is a comment."),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {}),
        (0, import_i18n.__)(
          "To get started with moderating, editing, and deleting comments, please visit the Comments screen in the dashboard."
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {}),
        (0, import_element.createInterpolateElement)(
          (0, import_i18n.__)(
            "Commenter avatars come from <a>Gravatar</a>."
          ),
          {
            a: (
              // eslint-disable-next-line jsx-a11y/anchor-has-content
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)("a", { href: "https://gravatar.com/" })
            )
          }
        )
      ] }) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "reply", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "a",
        {
          className: "comment-reply-link",
          href: "#top",
          "aria-label": (0, import_i18n.__)(
            "Reply to A WordPress Commenter"
          ),
          children: (0, import_i18n.__)("Reply")
        }
      ) })
    ] }) }) }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "navigation", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "alignleft", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("a", { href: "#top", children: [
        "\xAB ",
        (0, import_i18n.__)("Older Comments")
      ] }) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "alignright", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("a", { href: "#top", children: [
        (0, import_i18n.__)("Newer Comments"),
        " \xBB"
      ] }) })
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_form.default, { postId, postType })
  ] });
}
