// @wordpress/block-editor/build/components/html-element-control/messages.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/html-element-control/messages.js
var messages_exports = {};
__export(messages_exports, {
  htmlElementMessages: () => htmlElementMessages
});
module.exports = __toCommonJS(messages_exports);
var import_i18n = require("@wordpress/i18n");
var htmlElementMessages = {
  a: (0, import_i18n.__)(
    "The <a> element should be used for links that navigate to a different page or to a different section within the same page."
  ),
  article: (0, import_i18n.__)(
    "The <article> element should represent a self-contained, syndicatable portion of the document."
  ),
  aside: (0, import_i18n.__)(
    "The <aside> element should represent a portion of a document whose content is only indirectly related to the document's main content."
  ),
  button: (0, import_i18n.__)(
    "The <button> element should be used for interactive controls that perform an action on the current page, such as opening a modal or toggling content visibility."
  ),
  div: (0, import_i18n.__)(
    "The <div> element should only be used if the block is a design element with no semantic meaning."
  ),
  footer: (0, import_i18n.__)(
    "The <footer> element should represent a footer for its nearest sectioning element (e.g.: <section>, <article>, <main> etc.)."
  ),
  header: (0, import_i18n.__)(
    "The <header> element should represent introductory content, typically a group of introductory or navigational aids."
  ),
  main: (0, import_i18n.__)(
    "The <main> element should be used for the primary content of your document only."
  ),
  nav: (0, import_i18n.__)(
    "The <nav> element should be used to identify groups of links that are intended to be used for website or page content navigation."
  ),
  section: (0, import_i18n.__)(
    "The <section> element should represent a standalone portion of the document that can't be better represented by another element."
  )
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  htmlElementMessages
});
