// @wordpress/dataviews/build-module/components/dataviews-picker-footer/index.mjs
// packages/dataviews/src/components/dataviews-picker-footer/index.tsx
import { Button, CheckboxControl } from "@wordpress/components";
import { useRegistry } from "@wordpress/data";
import { useContext, useMemo, useState } from "@wordpress/element";
import { Stack } from "@wordpress/ui";
import { __ } from "@wordpress/i18n";
import DataViewsPagination from "../dataviews-pagination/index.mjs";
import DataViewsContext from "../dataviews-context/index.mjs";
import getFooterMessage from "../../utils/get-footer-message.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
var EMPTY_ARRAY = [];
function useIsMultiselectPicker(actions) {
  return useMemo(() => {
    return actions?.every((action) => action.supportsBulk);
  }, [actions]);
}
function BulkSelectionCheckbox({
  selection,
  selectedItems,
  onChangeSelection,
  data,
  getItemId
}) {
  const areAllSelected = selectedItems.length === data.length;
  return /* @__PURE__ */ jsx(
    CheckboxControl,
    {
      className: "dataviews-view-table-selection-checkbox",
      checked: areAllSelected,
      indeterminate: !areAllSelected && !!selectedItems.length,
      onChange: () => {
        if (areAllSelected) {
          onChangeSelection(
            selection.filter(
              (id) => !data.some(
                (item) => id === getItemId(item)
              )
            )
          );
        } else {
          const selectionSet = /* @__PURE__ */ new Set([
            ...selection,
            ...data.map((item) => getItemId(item))
          ]);
          onChangeSelection(Array.from(selectionSet));
        }
      },
      "aria-label": areAllSelected ? __("Deselect all") : __("Select all")
    }
  );
}
function ActionButtons({
  actions,
  items,
  selection
}) {
  const registry = useRegistry();
  const [actionInProgress, setActionInProgress] = useState(
    null
  );
  return /* @__PURE__ */ jsx(Stack, { direction: "row", gap: "xs", children: actions.map((action) => {
    if (!("callback" in action)) {
      return null;
    }
    const { id, label, icon, isPrimary, callback } = action;
    const _label = typeof label === "string" ? label : label(items);
    const variant = isPrimary ? "primary" : "tertiary";
    const isInProgress = id === actionInProgress;
    return /* @__PURE__ */ jsx(
      Button,
      {
        accessibleWhenDisabled: true,
        icon,
        disabled: isInProgress || !selection?.length,
        isBusy: isInProgress,
        onClick: async () => {
          setActionInProgress(id);
          await callback(items, {
            registry
          });
          setActionInProgress(null);
        },
        size: "compact",
        variant,
        children: _label
      },
      id
    );
  }) });
}
function DataViewsPickerFooter() {
  const {
    data,
    selection,
    onChangeSelection,
    getItemId,
    actions = EMPTY_ARRAY,
    paginationInfo
  } = useContext(DataViewsContext);
  const isMultiselect = useIsMultiselectPicker(actions);
  const message = getFooterMessage(
    selection.length,
    data.length,
    paginationInfo.totalItems
  );
  const selectedItems = useMemo(
    () => data.filter((item) => selection.includes(getItemId(item))),
    [selection, getItemId, data]
  );
  return /* @__PURE__ */ jsxs(
    Stack,
    {
      direction: "row",
      justify: "space-between",
      align: "center",
      className: "dataviews-footer",
      gap: "sm",
      children: [
        /* @__PURE__ */ jsxs(
          Stack,
          {
            direction: "row",
            className: "dataviews-picker-footer__bulk-selection",
            gap: "md",
            align: "center",
            children: [
              isMultiselect && /* @__PURE__ */ jsx(
                BulkSelectionCheckbox,
                {
                  selection,
                  selectedItems,
                  onChangeSelection,
                  data,
                  getItemId
                }
              ),
              /* @__PURE__ */ jsx("span", { className: "dataviews-bulk-actions-footer__item-count", children: message })
            ]
          }
        ),
        /* @__PURE__ */ jsx(DataViewsPagination, {}),
        Boolean(actions?.length) && /* @__PURE__ */ jsx("div", { className: "dataviews-picker-footer__actions", children: /* @__PURE__ */ jsx(
          ActionButtons,
          {
            actions,
            items: selectedItems,
            selection
          }
        ) })
      ]
    }
  );
}
export {
  DataViewsPickerFooter,
  useIsMultiselectPicker
};
