// @wordpress/reusable-blocks/build/store/actions.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/reusable-blocks/src/store/actions.js
var actions_exports = {};
__export(actions_exports, {
  __experimentalConvertBlockToStatic: () => __experimentalConvertBlockToStatic,
  __experimentalConvertBlocksToReusable: () => __experimentalConvertBlocksToReusable,
  __experimentalDeleteReusableBlock: () => __experimentalDeleteReusableBlock,
  __experimentalSetEditingReusableBlock: () => __experimentalSetEditingReusableBlock
});
module.exports = __toCommonJS(actions_exports);
var import_block_editor = require("@wordpress/block-editor");
var import_blocks = require("@wordpress/blocks");
var import_i18n = require("@wordpress/i18n");
var __experimentalConvertBlockToStatic = (clientId) => ({ registry }) => {
  const oldBlock = registry.select(import_block_editor.store).getBlock(clientId);
  const reusableBlock = registry.select("core").getEditedEntityRecord(
    "postType",
    "wp_block",
    oldBlock.attributes.ref
  );
  const newBlocks = (0, import_blocks.parse)(
    typeof reusableBlock.content === "function" ? reusableBlock.content(reusableBlock) : reusableBlock.content
  );
  registry.dispatch(import_block_editor.store).replaceBlocks(oldBlock.clientId, newBlocks);
};
var __experimentalConvertBlocksToReusable = (clientIds, title, syncType) => async ({ registry, dispatch }) => {
  const meta = syncType === "unsynced" ? {
    wp_pattern_sync_status: syncType
  } : void 0;
  const reusableBlock = {
    title: title || (0, import_i18n.__)("Untitled pattern block"),
    content: (0, import_blocks.serialize)(
      registry.select(import_block_editor.store).getBlocksByClientId(clientIds)
    ),
    status: "publish",
    meta
  };
  const updatedRecord = await registry.dispatch("core").saveEntityRecord("postType", "wp_block", reusableBlock);
  if (syncType === "unsynced") {
    return;
  }
  const newBlock = (0, import_blocks.createBlock)("core/block", {
    ref: updatedRecord.id
  });
  registry.dispatch(import_block_editor.store).replaceBlocks(clientIds, newBlock);
  dispatch.__experimentalSetEditingReusableBlock(
    newBlock.clientId,
    true
  );
};
var __experimentalDeleteReusableBlock = (id) => async ({ registry }) => {
  const reusableBlock = registry.select("core").getEditedEntityRecord("postType", "wp_block", id);
  if (!reusableBlock) {
    return;
  }
  const allBlocks = registry.select(import_block_editor.store).getBlocks();
  const associatedBlocks = allBlocks.filter(
    (block) => (0, import_blocks.isReusableBlock)(block) && block.attributes.ref === id
  );
  const associatedBlockClientIds = associatedBlocks.map(
    (block) => block.clientId
  );
  if (associatedBlockClientIds.length) {
    registry.dispatch(import_block_editor.store).removeBlocks(associatedBlockClientIds);
  }
  await registry.dispatch("core").deleteEntityRecord("postType", "wp_block", id);
};
function __experimentalSetEditingReusableBlock(clientId, isEditing) {
  return {
    type: "SET_EDITING_REUSABLE_BLOCK",
    clientId,
    isEditing
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  __experimentalConvertBlockToStatic,
  __experimentalConvertBlocksToReusable,
  __experimentalDeleteReusableBlock,
  __experimentalSetEditingReusableBlock
});
