// @wordpress/blocks/build/api/parser/fix-custom-classname.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/blocks/src/api/parser/fix-custom-classname.js
var fix_custom_classname_exports = {};
__export(fix_custom_classname_exports, {
  fixCustomClassname: () => fixCustomClassname,
  getHTMLRootElementClasses: () => getHTMLRootElementClasses
});
module.exports = __toCommonJS(fix_custom_classname_exports);
var import_registration = require("../registration.cjs");
var import_serializer = require("../serializer.cjs");
var import_get_block_attributes = require("./get-block-attributes.cjs");
var CLASS_ATTR_SCHEMA = {
  type: "string",
  source: "attribute",
  selector: "[data-custom-class-name] > *",
  attribute: "class"
};
function getHTMLRootElementClasses(innerHTML) {
  const parsed = (0, import_get_block_attributes.parseWithAttributeSchema)(
    `<div data-custom-class-name>${innerHTML}</div>`,
    CLASS_ATTR_SCHEMA
  );
  return parsed ? parsed.trim().split(/\s+/) : [];
}
function fixCustomClassname(blockAttributes, blockType, innerHTML) {
  if (!(0, import_registration.hasBlockSupport)(blockType, "customClassName", true)) {
    return blockAttributes;
  }
  const modifiedBlockAttributes = { ...blockAttributes };
  const { className: omittedClassName, ...attributesSansClassName } = modifiedBlockAttributes;
  const serialized = (0, import_serializer.getSaveContent)(blockType, attributesSansClassName);
  const defaultClasses = getHTMLRootElementClasses(serialized);
  const actualClasses = getHTMLRootElementClasses(innerHTML);
  const customClasses = actualClasses.filter(
    (className) => !defaultClasses.includes(className)
  );
  if (customClasses.length) {
    modifiedBlockAttributes.className = customClasses.join(" ");
  } else if (serialized) {
    delete modifiedBlockAttributes.className;
  }
  return modifiedBlockAttributes;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  fixCustomClassname,
  getHTMLRootElementClasses
});
