// @wordpress/editor/build/components/collaborators-overlay/collaborator-styles.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/editor/src/components/collaborators-overlay/collaborator-styles.ts
var collaborator_styles_exports = {};
__export(collaborator_styles_exports, {
  BORDER_WIDTH: () => BORDER_WIDTH,
  BORDER_WIDTH_FOCUS_FALLBACK: () => BORDER_WIDTH_FOCUS_FALLBACK,
  BUTTON_SIZE_COMPACT: () => BUTTON_SIZE_COMPACT,
  BUTTON_SIZE_SMALL: () => BUTTON_SIZE_SMALL,
  ELEVATION_X_SMALL: () => ELEVATION_X_SMALL,
  FONT_LINE_HEIGHT_SMALL: () => FONT_LINE_HEIGHT_SMALL,
  FONT_SIZE_MEDIUM: () => FONT_SIZE_MEDIUM,
  FONT_SIZE_X_SMALL: () => FONT_SIZE_X_SMALL,
  FONT_WEIGHT_MEDIUM: () => FONT_WEIGHT_MEDIUM,
  GRID_UNIT_05: () => GRID_UNIT_05,
  GRID_UNIT_10: () => GRID_UNIT_10,
  RADIUS_FULL: () => RADIUS_FULL,
  WHITE: () => WHITE
});
module.exports = __toCommonJS(collaborator_styles_exports);
var ELEVATION_X_SMALL = "0 1px 1px rgba(0, 0, 0, 0.03), 0 1px 2px rgba(0, 0, 0, 0.02), 0 3px 3px rgba(0, 0, 0, 0.02), 0 4px 4px rgba(0, 0, 0, 0.01)";
var RADIUS_FULL = "9999px";
var BUTTON_SIZE_COMPACT = "32px";
var BUTTON_SIZE_SMALL = "24px";
var GRID_UNIT_05 = "4px";
var GRID_UNIT_10 = "8px";
var BORDER_WIDTH = "1px";
var BORDER_WIDTH_FOCUS_FALLBACK = "2px";
var WHITE = "#fff";
var FONT_SIZE_X_SMALL = "11px";
var FONT_SIZE_MEDIUM = "13px";
var FONT_WEIGHT_MEDIUM = "499";
var FONT_LINE_HEIGHT_SMALL = "20px";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BORDER_WIDTH,
  BORDER_WIDTH_FOCUS_FALLBACK,
  BUTTON_SIZE_COMPACT,
  BUTTON_SIZE_SMALL,
  ELEVATION_X_SMALL,
  FONT_LINE_HEIGHT_SMALL,
  FONT_SIZE_MEDIUM,
  FONT_SIZE_X_SMALL,
  FONT_WEIGHT_MEDIUM,
  GRID_UNIT_05,
  GRID_UNIT_10,
  RADIUS_FULL,
  WHITE
});
