// @wordpress/block-editor@15.15.0/build/hooks/aria-label.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/hooks/aria-label.js
var aria_label_exports = {};
__export(aria_label_exports, {
  addAttribute: () => addAttribute,
  addSaveProps: () => addSaveProps,
  default: () => aria_label_default
});
module.exports = __toCommonJS(aria_label_exports);
var import_hooks = require("@wordpress/hooks");
var import_blocks = require("@wordpress/blocks");
var import_utils = require("./utils.cjs");
function addAttribute(settings) {
  if (settings?.attributes?.ariaLabel?.type) {
    return settings;
  }
  if ((0, import_blocks.hasBlockSupport)(settings, "ariaLabel")) {
    settings.attributes = {
      ...settings.attributes,
      ariaLabel: {
        type: "string"
      }
    };
  }
  return settings;
}
function addSaveProps(extraProps, blockType, attributes) {
  if ((0, import_blocks.hasBlockSupport)(blockType, "ariaLabel") && !(0, import_utils.shouldSkipSerialization)(blockType, "ariaLabel", "ariaLabel")) {
    extraProps["aria-label"] = attributes.ariaLabel === "" ? null : attributes.ariaLabel;
  }
  return extraProps;
}
var aria_label_default = {
  addSaveProps,
  attributeKeys: ["ariaLabel"],
  hasSupport(name) {
    return (0, import_blocks.hasBlockSupport)(name, "ariaLabel");
  }
};
(0, import_hooks.addFilter)(
  "blocks.registerBlockType",
  "core/ariaLabel/attribute",
  addAttribute
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  addAttribute,
  addSaveProps
});
