// @wordpress/block-editor/build/components/rich-text/format-edit.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/rich-text/format-edit.js
var format_edit_exports = {};
__export(format_edit_exports, {
  default: () => FormatEdit,
  usesContextKey: () => usesContextKey
});
module.exports = __toCommonJS(format_edit_exports);
var import_rich_text = require("@wordpress/rich-text");
var import_element = require("@wordpress/element");
var import_block_context = __toESM(require("../block-context/index.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var DEFAULT_BLOCK_CONTEXT = {};
var usesContextKey = /* @__PURE__ */ Symbol("usesContext");
function Edit({
  onChange,
  onFocus,
  value,
  forwardedRef,
  settings,
  isVisible
}) {
  const {
    name,
    edit: EditFunction,
    [usesContextKey]: usesContext
  } = settings;
  const blockContext = (0, import_element.useContext)(import_block_context.default);
  const context = (0, import_element.useMemo)(() => {
    return usesContext ? Object.fromEntries(
      Object.entries(blockContext).filter(
        ([key]) => usesContext.includes(key)
      )
    ) : DEFAULT_BLOCK_CONTEXT;
  }, [usesContext, blockContext]);
  if (!EditFunction) {
    return null;
  }
  const activeFormat = (0, import_rich_text.getActiveFormat)(value, name);
  const isActive = activeFormat !== void 0;
  const activeObject = (0, import_rich_text.getActiveObject)(value);
  const isObjectActive = activeObject !== void 0 && activeObject.type === name;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    EditFunction,
    {
      isActive,
      isVisible,
      activeAttributes: isActive ? activeFormat.attributes || {} : {},
      isObjectActive,
      activeObjectAttributes: isObjectActive ? activeObject.attributes || {} : {},
      value,
      onChange,
      onFocus,
      contentRef: forwardedRef,
      context
    },
    name
  );
}
function FormatEdit({ formatTypes, ...props }) {
  return formatTypes.map((settings) => /* @__PURE__ */ (0, import_react.createElement)(Edit, { settings, ...props, key: settings.name }));
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  usesContextKey
});
