// @wordpress/dataviews@13.1.0/build/components/dataviews-layouts/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/dataviews/src/components/dataviews-layouts/index.ts
var dataviews_layouts_exports = {};
__export(dataviews_layouts_exports, {
  VIEW_LAYOUTS: () => VIEW_LAYOUTS
});
module.exports = __toCommonJS(dataviews_layouts_exports);
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var import_table = __toESM(require("./table/index.cjs"));
var import_grid = __toESM(require("./grid/index.cjs"));
var import_list = __toESM(require("./list/index.cjs"));
var import_activity = __toESM(require("./activity/index.cjs"));
var import_picker_grid = __toESM(require("./picker-grid/index.cjs"));
var import_picker_table = __toESM(require("./picker-table/index.cjs"));
var import_constants = require("../../constants.cjs");
var import_density_picker = __toESM(require("./utils/density-picker.cjs"));
var import_grid_config_options = __toESM(require("./utils/grid-config-options.cjs"));
var VIEW_LAYOUTS = [
  {
    type: import_constants.LAYOUT_TABLE,
    label: (0, import_i18n.__)("Table"),
    component: import_table.default,
    icon: import_icons.blockTable,
    viewConfigOptions: import_density_picker.default
  },
  {
    type: import_constants.LAYOUT_GRID,
    label: (0, import_i18n.__)("Grid"),
    component: import_grid.default,
    icon: import_icons.category,
    viewConfigOptions: import_grid_config_options.default
  },
  {
    type: import_constants.LAYOUT_LIST,
    label: (0, import_i18n.__)("List"),
    component: import_list.default,
    icon: (0, import_i18n.isRTL)() ? import_icons.formatListBulletsRTL : import_icons.formatListBullets,
    viewConfigOptions: import_density_picker.default
  },
  {
    type: import_constants.LAYOUT_ACTIVITY,
    label: (0, import_i18n.__)("Activity"),
    component: import_activity.default,
    icon: import_icons.scheduled,
    viewConfigOptions: import_density_picker.default
  },
  {
    type: import_constants.LAYOUT_PICKER_GRID,
    label: (0, import_i18n.__)("Grid"),
    component: import_picker_grid.default,
    icon: import_icons.category,
    viewConfigOptions: import_grid_config_options.default,
    isPicker: true
  },
  {
    type: import_constants.LAYOUT_PICKER_TABLE,
    label: (0, import_i18n.__)("Table"),
    component: import_picker_table.default,
    icon: import_icons.blockTable,
    viewConfigOptions: import_density_picker.default,
    isPicker: true
  }
];
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VIEW_LAYOUTS
});
