// @wordpress/fields/build/fields/post-content-info/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/fields/src/fields/post-content-info/index.ts
var post_content_info_exports = {};
__export(post_content_info_exports, {
  default: () => post_content_info_default
});
module.exports = __toCommonJS(post_content_info_exports);
var import_i18n = require("@wordpress/i18n");
var import_post_content_info_view = __toESM(require("./post-content-info-view.cjs"));
var postContentInfoField = {
  label: (0, import_i18n.__)("Post content information"),
  id: "post-content-info",
  type: "text",
  readOnly: true,
  render: import_post_content_info_view.default,
  enableSorting: false,
  enableHiding: false,
  filterBy: false
};
var post_content_info_default = postContentInfoField;
