// @wordpress/block-library/build-module/query-title/variations.mjs
// packages/block-library/src/query-title/variations.js
import { __ } from "@wordpress/i18n";
import { title } from "@wordpress/icons";
var variations = [
  {
    isDefault: true,
    name: "archive-title",
    title: __("Archive Title"),
    description: __(
      "Display the archive title based on the queried object."
    ),
    icon: title,
    attributes: {
      type: "archive"
    },
    scope: ["inserter"]
  },
  {
    isDefault: false,
    name: "search-title",
    title: __("Search Results Title"),
    description: __(
      "Display the search results title based on the queried object."
    ),
    icon: title,
    attributes: {
      type: "search"
    },
    scope: ["inserter"]
  },
  {
    isDefault: false,
    name: "post-type-label",
    title: __("Post Type Label"),
    description: __(
      "Display the post type label based on the queried object."
    ),
    icon: title,
    attributes: {
      type: "post-type"
    },
    scope: ["inserter"]
  }
];
variations.forEach((variation) => {
  if (variation.isActive) {
    return;
  }
  variation.isActive = (blockAttributes, variationAttributes) => blockAttributes.type === variationAttributes.type;
});
var variations_default = variations;
export {
  variations_default as default
};
