// @wordpress/reusable-blocks@5.42.0/build/components/reusable-blocks-menu-items/reusable-blocks-manage-button.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/reusable-blocks/src/components/reusable-blocks-menu-items/reusable-blocks-manage-button.js
var reusable_blocks_manage_button_exports = {};
__export(reusable_blocks_manage_button_exports, {
  default: () => reusable_blocks_manage_button_default
});
module.exports = __toCommonJS(reusable_blocks_manage_button_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_blocks = require("@wordpress/blocks");
var import_data = require("@wordpress/data");
var import_element = require("@wordpress/element");
var import_block_editor = require("@wordpress/block-editor");
var import_url = require("@wordpress/url");
var import_core_data = require("@wordpress/core-data");
var import_store = require("../../store/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function ReusableBlocksManageButton({ clientId }) {
  const [showConfirmDialog, setShowConfirmDialog] = (0, import_element.useState)(false);
  const { canRemove, isVisible, managePatternsUrl } = (0, import_data.useSelect)(
    (select) => {
      const { getBlock, canRemoveBlock } = select(import_block_editor.store);
      const { canUser } = select(import_core_data.store);
      const reusableBlock = getBlock(clientId);
      return {
        canRemove: canRemoveBlock(clientId),
        isVisible: !!reusableBlock && (0, import_blocks.isReusableBlock)(reusableBlock) && !!canUser("update", {
          kind: "postType",
          name: "wp_block",
          id: reusableBlock.attributes.ref
        }),
        // The site editor and templates both check whether the user
        // has edit_theme_options capabilities. We can leverage that here
        // and omit the manage patterns link if the user can't access it.
        managePatternsUrl: canUser("create", {
          kind: "postType",
          name: "wp_template"
        }) ? (0, import_url.addQueryArgs)("site-editor.php", {
          p: "/pattern"
        }) : (0, import_url.addQueryArgs)("edit.php", {
          post_type: "wp_block"
        })
      };
    },
    [clientId]
  );
  const { __experimentalConvertBlockToStatic: convertBlockToStatic } = (0, import_data.useDispatch)(import_store.store);
  if (!isVisible) {
    return null;
  }
  const handleDetach = () => {
    convertBlockToStatic(clientId);
    setShowConfirmDialog(false);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.MenuItem, { href: managePatternsUrl, children: (0, import_i18n.__)("Manage patterns") }),
    canRemove && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.MenuItem, { onClick: () => setShowConfirmDialog(true), children: (0, import_i18n.__)("Disconnect pattern") }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.__experimentalConfirmDialog,
        {
          isOpen: showConfirmDialog,
          onConfirm: handleDetach,
          onCancel: () => setShowConfirmDialog(false),
          confirmButtonText: (0, import_i18n.__)("Disconnect"),
          size: "medium",
          title: (0, import_i18n.__)("Disconnect pattern?"),
          __experimentalHideHeader: false,
          children: (0, import_i18n.__)(
            "Blocks will be separated from the original pattern and will be fully editable. Future changes to the pattern will not apply here."
          )
        }
      )
    ] })
  ] });
}
var reusable_blocks_manage_button_default = ReusableBlocksManageButton;
