// @wordpress/block-editor/build-module/components/block-allowed-blocks/allowed-blocks-control.mjs
// packages/block-editor/src/components/block-allowed-blocks/allowed-blocks-control.js
import { BaseControl, Button } from "@wordpress/components";
import { __ } from "@wordpress/i18n";
import { useState } from "@wordpress/element";
import { useSelect } from "@wordpress/data";
import { hasBlockSupport, store as blocksStore } from "@wordpress/blocks";
import BlockAllowedBlocksModal from "./modal.mjs";
import { store as blockEditorStore } from "../../store/index.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
function BlockAllowedBlocksControl({ clientId }) {
  const [isBlockControlOpened, setIsBlockControlOpened] = useState(false);
  const { blockTypes, selectedBlockNames } = useSelect(
    (select) => {
      const { getBlockAttributes } = select(blockEditorStore);
      return {
        blockTypes: select(blocksStore).getBlockTypes(),
        selectedBlockNames: getBlockAttributes(clientId)?.allowedBlocks
      };
    },
    [clientId]
  );
  const filteredBlockTypes = blockTypes.filter(
    (blockType) => hasBlockSupport(blockType, "inserter", true) && (!blockType.parent || blockType.parent.includes("core/post-content"))
  );
  if (!filteredBlockTypes) {
    return null;
  }
  const selectedBlockTypes = selectedBlockNames === void 0 ? filteredBlockTypes : filteredBlockTypes.filter(
    (blockType) => selectedBlockNames.includes(blockType.name)
  );
  return /* @__PURE__ */ jsxs("div", { className: "block-editor-block-allowed-blocks-control", children: [
    /* @__PURE__ */ jsxs(
      BaseControl,
      {
        help: __(
          "Specify which blocks are allowed inside this container."
        ),
        children: [
          /* @__PURE__ */ jsx(BaseControl.VisualLabel, { children: __("Allowed Blocks") }),
          /* @__PURE__ */ jsx(
            Button,
            {
              __next40pxDefaultSize: true,
              variant: "secondary",
              onClick: () => {
                setIsBlockControlOpened(true);
              },
              className: "block-editor-block-allowed-blocks-control__button",
              children: __("Manage allowed blocks")
            }
          )
        ]
      }
    ),
    isBlockControlOpened && /* @__PURE__ */ jsx(
      BlockAllowedBlocksModal,
      {
        clientId,
        blockTypes: filteredBlockTypes,
        selectedBlockTypes,
        onClose: () => setIsBlockControlOpened(false)
      }
    )
  ] });
}
export {
  BlockAllowedBlocksControl as default
};
