// @wordpress/block-editor@15.15.0/build/components/inserter-listbox/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/inserter-listbox/index.js
var inserter_listbox_exports = {};
__export(inserter_listbox_exports, {
  InserterListboxGroup: () => import_group.default,
  InserterListboxItem: () => import_item.default,
  InserterListboxRow: () => import_row.default,
  default: () => inserter_listbox_default
});
module.exports = __toCommonJS(inserter_listbox_exports);
var import_components = require("@wordpress/components");
var import_element = require("@wordpress/element");
var import_group = __toESM(require("./group.cjs"));
var import_row = __toESM(require("./row.cjs"));
var import_item = __toESM(require("./item.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function InserterListBoxWrapper({ key, children }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_element.Fragment, { children }, key);
}
function InserterListbox({ children }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Composite,
    {
      focusShift: true,
      focusWrap: "horizontal",
      render: InserterListBoxWrapper,
      children
    }
  );
}
var inserter_listbox_default = InserterListbox;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  InserterListboxGroup,
  InserterListboxItem,
  InserterListboxRow
});
