// @wordpress/block-editor/build/components/block-tools/empty-block-inserter.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-tools/empty-block-inserter.js
var empty_block_inserter_exports = {};
__export(empty_block_inserter_exports, {
  default: () => EmptyBlockInserter
});
module.exports = __toCommonJS(empty_block_inserter_exports);
var import_clsx = __toESM(require("clsx"));
var import_cover = __toESM(require("../block-popover/cover.cjs"));
var import_use_block_toolbar_popover_props = __toESM(require("./use-block-toolbar-popover-props.cjs"));
var import_inserter = __toESM(require("../inserter/index.cjs"));
var import_use_selected_block_tool_props = __toESM(require("./use-selected-block-tool-props.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function EmptyBlockInserter({
  clientId,
  __unstableContentRef
}) {
  const {
    capturingClientId,
    isInsertionPointVisible,
    lastClientId,
    rootClientId
  } = (0, import_use_selected_block_tool_props.default)(clientId);
  const popoverProps = (0, import_use_block_toolbar_popover_props.default)({
    contentElement: __unstableContentRef?.current,
    clientId
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_cover.default,
    {
      clientId: capturingClientId || clientId,
      bottomClientId: lastClientId,
      className: (0, import_clsx.default)(
        "block-editor-block-list__block-side-inserter-popover",
        {
          "is-insertion-point-visible": isInsertionPointVisible
        }
      ),
      __unstableContentRef,
      ...popoverProps,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "block-editor-block-list__empty-block-inserter", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_inserter.default,
        {
          position: "bottom right",
          rootClientId,
          clientId,
          __experimentalIsQuick: true
        }
      ) })
    }
  );
}
