// @wordpress/block-library@9.42.0/build/cover/edit/cover-placeholder.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/cover/edit/cover-placeholder.js
var cover_placeholder_exports = {};
__export(cover_placeholder_exports, {
  default: () => CoverPlaceholder
});
module.exports = __toCommonJS(cover_placeholder_exports);
var import_block_editor = require("@wordpress/block-editor");
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var import_blob = require("@wordpress/blob");
var import_shared = require("../shared.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function CoverPlaceholder({
  disableMediaButtons = false,
  children,
  onSelectMedia,
  onError,
  style,
  toggleUseFeaturedImage
}) {
  const onFilesPreUpload = (files) => {
    if (files.length === 1) {
      onSelectMedia({ url: (0, import_blob.createBlobURL)(files[0]) });
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_block_editor.MediaPlaceholder,
    {
      icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_block_editor.BlockIcon, { icon: import_icons.cover }),
      labels: {
        title: (0, import_i18n.__)("Cover")
      },
      onSelect: onSelectMedia,
      allowedTypes: import_shared.ALLOWED_MEDIA_TYPES,
      disableMediaButtons,
      onToggleFeaturedImage: toggleUseFeaturedImage,
      onFilesPreUpload,
      onError,
      style,
      children
    }
  );
}
