// @wordpress/block-editor/build/components/border-radius-control/constants.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/border-radius-control/constants.js
var constants_exports = {};
__export(constants_exports, {
  CORNERS: () => CORNERS,
  DEFAULT_VALUES: () => DEFAULT_VALUES,
  EMPTY_ARRAY: () => EMPTY_ARRAY,
  ICONS: () => ICONS,
  MAX_BORDER_RADIUS_VALUES: () => MAX_BORDER_RADIUS_VALUES,
  MIN_BORDER_RADIUS_VALUE: () => MIN_BORDER_RADIUS_VALUE,
  RANGE_CONTROL_MAX_SIZE: () => RANGE_CONTROL_MAX_SIZE
});
module.exports = __toCommonJS(constants_exports);
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var DEFAULT_VALUES = {
  topLeft: void 0,
  topRight: void 0,
  bottomLeft: void 0,
  bottomRight: void 0
};
var RANGE_CONTROL_MAX_SIZE = 8;
var EMPTY_ARRAY = [];
var CORNERS = {
  all: (0, import_i18n.__)("Border radius"),
  topLeft: (0, import_i18n.__)("Top left"),
  topRight: (0, import_i18n.__)("Top right"),
  bottomLeft: (0, import_i18n.__)("Bottom left"),
  bottomRight: (0, import_i18n.__)("Bottom right")
};
var ICONS = {
  all: import_icons.cornerAll,
  topLeft: import_icons.cornerTopLeft,
  topRight: import_icons.cornerTopRight,
  bottomLeft: import_icons.cornerBottomLeft,
  bottomRight: import_icons.cornerBottomRight
};
var MIN_BORDER_RADIUS_VALUE = 0;
var MAX_BORDER_RADIUS_VALUES = {
  px: 100,
  em: 20,
  rem: 20
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CORNERS,
  DEFAULT_VALUES,
  EMPTY_ARRAY,
  ICONS,
  MAX_BORDER_RADIUS_VALUES,
  MIN_BORDER_RADIUS_VALUE,
  RANGE_CONTROL_MAX_SIZE
});
