// @wordpress/sync@1.42.0/build/errors.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/sync/src/errors.ts
var errors_exports = {};
__export(errors_exports, {
  ConnectionError: () => ConnectionError,
  ConnectionErrorCode: () => ConnectionErrorCode
});
module.exports = __toCommonJS(errors_exports);
var ConnectionErrorCode = /* @__PURE__ */ ((ConnectionErrorCode2) => {
  ConnectionErrorCode2["AUTHENTICATION_FAILED"] = "authentication-failed";
  ConnectionErrorCode2["CONNECTION_EXPIRED"] = "connection-expired";
  ConnectionErrorCode2["CONNECTION_LIMIT_EXCEEDED"] = "connection-limit-exceeded";
  ConnectionErrorCode2["DOCUMENT_SIZE_LIMIT_EXCEEDED"] = "document-size-limit-exceeded";
  ConnectionErrorCode2["UNKNOWN_ERROR"] = "unknown-error";
  return ConnectionErrorCode2;
})(ConnectionErrorCode || {});
var ConnectionError = class extends Error {
  constructor(code = "unknown-error" /* UNKNOWN_ERROR */, message) {
    super(message);
    this.code = code;
    this.name = "ConnectionError";
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ConnectionError,
  ConnectionErrorCode
});
