// @wordpress/dataviews@13.1.0/build/components/dataviews-layouts/grid/composite-grid.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/dataviews/src/components/dataviews-layouts/grid/composite-grid.tsx
var composite_grid_exports = {};
__export(composite_grid_exports, {
  default: () => CompositeGrid
});
module.exports = __toCommonJS(composite_grid_exports);
var import_clsx = __toESM(require("clsx"));
var import_components = require("@wordpress/components");
var import_ui = require("@wordpress/ui");
var import_i18n = require("@wordpress/i18n");
var import_compose = require("@wordpress/compose");
var import_keycodes = require("@wordpress/keycodes");
var import_element = require("@wordpress/element");
var import_lock_unlock = require("../../../lock-unlock.cjs");
var import_dataviews_item_actions = __toESM(require("../../dataviews-item-actions/index.cjs"));
var import_dataviews_selection_checkbox = __toESM(require("../../dataviews-selection-checkbox/index.cjs"));
var import_dataviews_context = __toESM(require("../../dataviews-context/index.cjs"));
var import_dataviews_bulk_actions = require("../../dataviews-bulk-actions/index.cjs");
var import_item_click_wrapper = require("../utils/item-click-wrapper.cjs");
var import_preview_size_picker = require("./preview-size-picker.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var { Badge } = (0, import_lock_unlock.unlock)(import_components.privateApis);
function chunk(array, size) {
  const chunks = [];
  for (let i = 0, j = array.length; i < j; i += size) {
    chunks.push(array.slice(i, i + size));
  }
  return chunks;
}
var GridItem = (0, import_element.forwardRef)(function GridItem2({
  view,
  selection,
  onChangeSelection,
  onClickItem,
  isItemClickable,
  renderItemLink,
  getItemId,
  item,
  actions,
  mediaField,
  titleField,
  descriptionField,
  regularFields,
  badgeFields,
  hasBulkActions,
  config,
  ...props
}, ref) {
  const { showTitle = true, showMedia = true, showDescription = true } = view;
  const hasBulkAction = (0, import_dataviews_bulk_actions.useHasAPossibleBulkAction)(actions, item);
  const id = getItemId(item);
  const instanceId = (0, import_compose.useInstanceId)(GridItem2);
  const isSelected = selection.includes(id);
  const mediaPlaceholder = /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: "dataviews-view-grid__media-placeholder" });
  const rendersMediaField = showMedia && mediaField?.render;
  const renderedMediaField = rendersMediaField ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    mediaField.render,
    {
      item,
      field: mediaField,
      config
    }
  ) : mediaPlaceholder;
  const renderedTitleField = showTitle && titleField?.render ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(titleField.render, { item, field: titleField }) : null;
  let mediaA11yProps;
  let titleA11yProps;
  if (isItemClickable(item) && onClickItem) {
    if (renderedTitleField) {
      mediaA11yProps = {
        "aria-labelledby": `dataviews-view-grid__title-field-${instanceId}`
      };
      titleA11yProps = {
        id: `dataviews-view-grid__title-field-${instanceId}`
      };
    } else {
      mediaA11yProps = {
        "aria-label": (0, import_i18n.__)("Navigate to item")
      };
    }
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ui.Stack,
    {
      direction: "column",
      ...props,
      ref,
      className: (0, import_clsx.default)(
        props.className,
        "dataviews-view-grid__row__gridcell",
        "dataviews-view-grid__card",
        {
          "is-selected": hasBulkAction && isSelected
        }
      ),
      onClickCapture: (event) => {
        props.onClickCapture?.(event);
        if ((0, import_keycodes.isAppleOS)() ? event.metaKey : event.ctrlKey) {
          event.stopPropagation();
          event.preventDefault();
          if (!hasBulkAction) {
            return;
          }
          onChangeSelection(
            selection.includes(id) ? selection.filter((itemId) => id !== itemId) : [...selection, id]
          );
        }
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_item_click_wrapper.ItemClickWrapper,
          {
            item,
            isItemClickable,
            onClickItem,
            renderItemLink,
            className: (0, import_clsx.default)("dataviews-view-grid__media", {
              "dataviews-view-grid__media--placeholder": !rendersMediaField
            }),
            ...mediaA11yProps,
            children: renderedMediaField
          }
        ),
        hasBulkActions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_dataviews_selection_checkbox.default,
          {
            item,
            selection,
            onChangeSelection,
            getItemId,
            titleField,
            disabled: !hasBulkAction
          }
        ),
        !!actions?.length && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "dataviews-view-grid__media-actions", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_item_actions.default, { item, actions, isCompact: true }) }),
        showTitle && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "dataviews-view-grid__title", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_item_click_wrapper.ItemClickWrapper,
          {
            item,
            isItemClickable,
            onClickItem,
            renderItemLink,
            className: "dataviews-view-grid__title-field dataviews-title-field",
            ...titleA11yProps,
            title: titleField?.getValueFormatted({
              item,
              field: titleField
            }) || void 0,
            children: renderedTitleField
          }
        ) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ui.Stack, { direction: "column", gap: "xs", children: [
          showDescription && descriptionField?.render && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            descriptionField.render,
            {
              item,
              field: descriptionField
            }
          ),
          !!badgeFields?.length && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ui.Stack,
            {
              direction: "row",
              className: "dataviews-view-grid__badge-fields",
              gap: "sm",
              wrap: "wrap",
              align: "top",
              justify: "flex-start",
              children: badgeFields.map((field) => {
                return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  Badge,
                  {
                    className: "dataviews-view-grid__field-value",
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      field.render,
                      {
                        item,
                        field
                      }
                    )
                  },
                  field.id
                );
              })
            }
          ),
          !!regularFields?.length && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ui.Stack,
            {
              direction: "column",
              className: "dataviews-view-grid__fields",
              gap: "xs",
              children: regularFields.map((field) => {
                return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_components.Flex,
                  {
                    className: "dataviews-view-grid__field",
                    gap: 1,
                    justify: "flex-start",
                    expanded: true,
                    style: { height: "auto" },
                    direction: "row",
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.Tooltip, { text: field.label, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.FlexItem, { className: "dataviews-view-grid__field-name", children: field.header }) }),
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_components.FlexItem,
                        {
                          className: "dataviews-view-grid__field-value",
                          style: { maxHeight: "none" },
                          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                            field.render,
                            {
                              item,
                              field
                            }
                          )
                        }
                      )
                    ] })
                  },
                  field.id
                );
              })
            }
          )
        ] })
      ]
    }
  );
});
function CompositeGrid({
  data,
  isInfiniteScroll,
  className,
  inert,
  isLoading,
  view,
  fields,
  selection,
  onChangeSelection,
  onClickItem,
  isItemClickable,
  renderItemLink,
  getItemId,
  actions
}) {
  const { paginationInfo, resizeObserverRef } = (0, import_element.useContext)(import_dataviews_context.default);
  const gridColumns = (0, import_preview_size_picker.useGridColumns)();
  const hasBulkActions = (0, import_dataviews_bulk_actions.useSomeItemHasAPossibleBulkAction)(actions, data);
  const titleField = fields.find(
    (field) => field.id === view?.titleField
  );
  const mediaField = fields.find(
    (field) => field.id === view?.mediaField
  );
  const descriptionField = fields.find(
    (field) => field.id === view?.descriptionField
  );
  const otherFields = view.fields ?? [];
  const { regularFields, badgeFields } = otherFields.reduce(
    (accumulator, fieldId) => {
      const field = fields.find((f) => f.id === fieldId);
      if (!field) {
        return accumulator;
      }
      const key = view.layout?.badgeFields?.includes(fieldId) ? "badgeFields" : "regularFields";
      accumulator[key].push(field);
      return accumulator;
    },
    { regularFields: [], badgeFields: [] }
  );
  const size = "900px";
  const totalRows = Math.ceil(data.length / gridColumns);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Composite,
    {
      role: isInfiniteScroll ? "feed" : "grid",
      className: (0, import_clsx.default)("dataviews-view-grid", className, {
        [`has-${view.layout?.density}-density`]: view.layout?.density && ["compact", "comfortable"].includes(
          view.layout.density
        )
      }),
      focusWrap: true,
      "aria-busy": isLoading,
      "aria-rowcount": isInfiniteScroll ? void 0 : totalRows,
      ref: resizeObserverRef,
      inert,
      children: chunk(data, gridColumns).map((row, i) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.Composite.Row,
        {
          render: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              role: "row",
              "aria-rowindex": i + 1,
              "aria-label": (0, import_i18n.sprintf)(
                /* translators: %d: The row number in the grid */
                (0, import_i18n.__)("Row %d"),
                i + 1
              ),
              className: "dataviews-view-grid__row",
              style: {
                gridTemplateColumns: `repeat( ${gridColumns}, minmax(0, 1fr) )`
              }
            }
          ),
          children: row.map((item, indexInRow) => {
            const index = i * gridColumns + indexInRow;
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_components.Composite.Item,
              {
                render: (props) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  GridItem,
                  {
                    ...props,
                    role: isInfiniteScroll ? "article" : "gridcell",
                    "aria-setsize": isInfiniteScroll ? paginationInfo.totalItems : void 0,
                    "aria-posinset": isInfiniteScroll ? index + 1 : void 0,
                    view,
                    selection,
                    onChangeSelection,
                    onClickItem,
                    isItemClickable,
                    renderItemLink,
                    getItemId,
                    item,
                    actions,
                    mediaField,
                    titleField,
                    descriptionField,
                    regularFields,
                    badgeFields,
                    hasBulkActions,
                    config: {
                      sizes: size
                    }
                  }
                )
              },
              getItemId(item)
            );
          })
        },
        i
      ))
    }
  );
}
