// @wordpress/patterns@2.42.0/build/private-hooks.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/patterns/src/private-hooks.js
var private_hooks_exports = {};
__export(private_hooks_exports, {
  useAddPatternCategory: () => useAddPatternCategory
});
module.exports = __toCommonJS(private_hooks_exports);
var import_data = require("@wordpress/data");
var import_core_data = require("@wordpress/core-data");
var import_element = require("@wordpress/element");
var import_category_selector = require("./components/category-selector.cjs");
function useAddPatternCategory() {
  const { saveEntityRecord, invalidateResolution } = (0, import_data.useDispatch)(import_core_data.store);
  const { corePatternCategories, userPatternCategories } = (0, import_data.useSelect)(
    (select) => {
      const { getUserPatternCategories, getBlockPatternCategories } = select(import_core_data.store);
      return {
        corePatternCategories: getBlockPatternCategories(),
        userPatternCategories: getUserPatternCategories()
      };
    },
    []
  );
  const categoryMap = (0, import_element.useMemo)(() => {
    const uniqueCategories = /* @__PURE__ */ new Map();
    userPatternCategories.forEach((category) => {
      uniqueCategories.set(category.label.toLowerCase(), {
        label: category.label,
        name: category.name,
        id: category.id
      });
    });
    corePatternCategories.forEach((category) => {
      if (!uniqueCategories.has(category.label.toLowerCase()) && // There are two core categories with `Post` label so explicitly remove the one with
      // the `query` slug to avoid any confusion.
      category.name !== "query") {
        uniqueCategories.set(category.label.toLowerCase(), {
          label: category.label,
          name: category.name
        });
      }
    });
    return uniqueCategories;
  }, [userPatternCategories, corePatternCategories]);
  async function findOrCreateTerm(term) {
    try {
      const existingTerm = categoryMap.get(term.toLowerCase());
      if (existingTerm?.id) {
        return existingTerm.id;
      }
      const termData = existingTerm ? { name: existingTerm.label, slug: existingTerm.name } : { name: term };
      const newTerm = await saveEntityRecord(
        "taxonomy",
        import_category_selector.CATEGORY_SLUG,
        termData,
        { throwOnError: true }
      );
      invalidateResolution("getUserPatternCategories");
      return newTerm.id;
    } catch (error) {
      if (error.code !== "term_exists") {
        throw error;
      }
      return error.data.term_id;
    }
  }
  return { categoryMap, findOrCreateTerm };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useAddPatternCategory
});
