// @wordpress/block-editor/build/components/responsive-block-control/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/responsive-block-control/index.js
var responsive_block_control_exports = {};
__export(responsive_block_control_exports, {
  default: () => responsive_block_control_default
});
module.exports = __toCommonJS(responsive_block_control_exports);
var import_clsx = __toESM(require("clsx"));
var import_i18n = require("@wordpress/i18n");
var import_element = require("@wordpress/element");
var import_components = require("@wordpress/components");
var import_label = __toESM(require("./label.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function ResponsiveBlockControl(props) {
  const {
    title,
    property,
    toggleLabel,
    onIsResponsiveChange,
    renderDefaultControl,
    renderResponsiveControls,
    isResponsive = false,
    defaultLabel = {
      id: "all",
      label: (0, import_i18n._x)("All", "screen sizes")
    },
    viewports = [
      {
        id: "small",
        label: (0, import_i18n.__)("Small screens")
      },
      {
        id: "medium",
        label: (0, import_i18n.__)("Medium screens")
      },
      {
        id: "large",
        label: (0, import_i18n.__)("Large screens")
      }
    ]
  } = props;
  if (!title || !property || !renderDefaultControl) {
    return null;
  }
  const toggleControlLabel = toggleLabel || (0, import_i18n.sprintf)(
    /* translators: %s: Property value for the control (eg: margin, padding, etc.). */
    (0, import_i18n.__)("Use the same %s on all screen sizes."),
    property
  );
  const toggleHelpText = (0, import_i18n.__)(
    "Choose whether to use the same value for all screen sizes or a unique value for each screen size."
  );
  const defaultControl = renderDefaultControl(
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_label.default,
      {
        property,
        viewport: defaultLabel
      }
    ),
    defaultLabel
  );
  const defaultResponsiveControls = () => {
    return viewports.map((viewport) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_element.Fragment, { children: renderDefaultControl(
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_label.default,
        {
          property,
          viewport
        }
      ),
      viewport
    ) }, viewport.id));
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("fieldset", { className: "block-editor-responsive-block-control", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("legend", { className: "block-editor-responsive-block-control__title", children: title }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "block-editor-responsive-block-control__inner", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.ToggleControl,
        {
          className: "block-editor-responsive-block-control__toggle",
          label: toggleControlLabel,
          checked: !isResponsive,
          onChange: onIsResponsiveChange,
          help: toggleHelpText
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        "div",
        {
          className: (0, import_clsx.default)(
            "block-editor-responsive-block-control__group",
            {
              "is-responsive": isResponsive
            }
          ),
          children: [
            !isResponsive && defaultControl,
            isResponsive && (renderResponsiveControls ? renderResponsiveControls(viewports) : defaultResponsiveControls())
          ]
        }
      )
    ] })
  ] });
}
var responsive_block_control_default = ResponsiveBlockControl;
