// @wordpress/block-editor/build/components/inspector-controls-tabs/utils.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/inspector-controls-tabs/utils.js
var utils_exports = {};
__export(utils_exports, {
  TAB_CONTENT: () => TAB_CONTENT,
  TAB_LIST_VIEW: () => TAB_LIST_VIEW,
  TAB_SETTINGS: () => TAB_SETTINGS,
  TAB_STYLES: () => TAB_STYLES
});
module.exports = __toCommonJS(utils_exports);
var import_icons = require("@wordpress/icons");
var import_i18n = require("@wordpress/i18n");
var TAB_SETTINGS = {
  name: "settings",
  title: (0, import_i18n.__)("Settings"),
  value: "settings",
  icon: import_icons.cog
};
var TAB_STYLES = {
  name: "styles",
  title: (0, import_i18n.__)("Styles"),
  value: "styles",
  icon: import_icons.styles
};
var TAB_CONTENT = {
  name: "content",
  title: (0, import_i18n.__)("Content"),
  value: "content",
  icon: import_icons.page
};
var TAB_LIST_VIEW = {
  name: "list",
  title: (0, import_i18n.__)("List View"),
  value: "list-view",
  icon: import_icons.listView
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TAB_CONTENT,
  TAB_LIST_VIEW,
  TAB_SETTINGS,
  TAB_STYLES
});
