// @wordpress/block-editor/build/components/block-settings-menu/block-mode-toggle.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-settings-menu/block-mode-toggle.js
var block_mode_toggle_exports = {};
__export(block_mode_toggle_exports, {
  default: () => BlockModeToggle
});
module.exports = __toCommonJS(block_mode_toggle_exports);
var import_i18n = require("@wordpress/i18n");
var import_components = require("@wordpress/components");
var import_blocks = require("@wordpress/blocks");
var import_data = require("@wordpress/data");
var import_store = require("../../store/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var noop = () => {
};
function BlockModeToggle({ clientId, onToggle = noop }) {
  const { blockType, mode, enabled } = (0, import_data.useSelect)(
    (select) => {
      const { getBlock, getBlockMode, getSettings } = select(import_store.store);
      const block = getBlock(clientId);
      return {
        mode: getBlockMode(clientId),
        blockType: block ? (0, import_blocks.getBlockType)(block.name) : null,
        enabled: getSettings().codeEditingEnabled && !!block?.isValid
      };
    },
    [clientId]
  );
  const { toggleBlockMode } = (0, import_data.useDispatch)(import_store.store);
  if (!blockType || !(0, import_blocks.hasBlockSupport)(blockType, "html", true) || !enabled) {
    return null;
  }
  const label = mode === "visual" ? (0, import_i18n.__)("Edit as HTML") : (0, import_i18n.__)("Edit visually");
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.MenuItem,
    {
      onClick: () => {
        toggleBlockMode(clientId);
        onToggle();
      },
      children: label
    }
  );
}
