// @wordpress/block-editor/build/components/block-types-list/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-types-list/index.js
var block_types_list_exports = {};
__export(block_types_list_exports, {
  default: () => block_types_list_default
});
module.exports = __toCommonJS(block_types_list_exports);
var import_blocks = require("@wordpress/blocks");
var import_compose = require("@wordpress/compose");
var import_inserter_list_item = __toESM(require("../inserter-list-item/index.cjs"));
var import_inserter_listbox = require("../inserter-listbox/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function chunk(array, size) {
  const chunks = [];
  for (let i = 0, j = array.length; i < j; i += size) {
    chunks.push(array.slice(i, i + size));
  }
  return chunks;
}
function BlockTypesList({
  items = [],
  onSelect,
  onHover = () => {
  },
  children,
  label,
  isDraggable = true
}) {
  const className = "block-editor-block-types-list";
  const listId = (0, import_compose.useInstanceId)(BlockTypesList, className);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_inserter_listbox.InserterListboxGroup, { className, "aria-label": label, children: [
    chunk(items, 3).map((row, i) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_inserter_listbox.InserterListboxRow, { children: row.map((item, j) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_inserter_list_item.default,
      {
        item,
        className: (0, import_blocks.getBlockMenuDefaultClassName)(
          item.id
        ),
        onSelect,
        onHover,
        isDraggable: isDraggable && !item.isDisabled,
        isFirst: i === 0 && j === 0,
        rowId: `${listId}-${i}`
      },
      item.id
    )) }, i)),
    children
  ] });
}
var block_types_list_default = BlockTypesList;
