// .ignored/@wordpress/core-data/build-module/hooks/use-entity-records.mjs
// packages/core-data/src/hooks/use-entity-records.ts
import { addQueryArgs } from "@wordpress/url";
import deprecated from "@wordpress/deprecated";
import { useSelect } from "@wordpress/data";
import { useMemo } from "@wordpress/element";
import useQuerySelect from "./use-query-select.mjs";
import { store as coreStore } from "../index.mjs";
import { unlock } from "../lock-unlock.mjs";
import { getNormalizedCommaSeparable } from "../utils/index.mjs";
var EMPTY_ARRAY = [];
function useEntityRecords(kind, name, queryArgs = {}, options = { enabled: true }) {
  const queryAsString = addQueryArgs("", queryArgs);
  const { data: records, ...rest } = useQuerySelect(
    (query) => {
      if (!options.enabled) {
        return {
          // Avoiding returning a new reference on every execution.
          data: EMPTY_ARRAY
        };
      }
      return query(coreStore).getEntityRecords(kind, name, queryArgs);
    },
    [kind, name, queryAsString, options.enabled]
  );
  const { totalItems, totalPages } = useSelect(
    (select) => {
      if (!options.enabled) {
        return {
          totalItems: null,
          totalPages: null
        };
      }
      return {
        totalItems: select(coreStore).getEntityRecordsTotalItems(
          kind,
          name,
          queryArgs
        ),
        totalPages: select(coreStore).getEntityRecordsTotalPages(
          kind,
          name,
          queryArgs
        )
      };
    },
    [kind, name, queryAsString, options.enabled]
  );
  return {
    records,
    totalItems,
    totalPages,
    ...rest
  };
}
function __experimentalUseEntityRecords(kind, name, queryArgs, options) {
  deprecated(`wp.data.__experimentalUseEntityRecords`, {
    alternative: "wp.data.useEntityRecords",
    since: "6.1"
  });
  return useEntityRecords(kind, name, queryArgs, options);
}
function useEntityRecordsWithPermissions(kind, name, queryArgs = {}, options = { enabled: true }) {
  const entityConfig = useSelect(
    (select) => select(coreStore).getEntityConfig(kind, name),
    [kind, name]
  );
  const { records: data, ...ret } = useEntityRecords(
    kind,
    name,
    {
      ...queryArgs,
      // If _fields is provided, we need to include _links in the request for permission caching to work.
      ...queryArgs._fields ? {
        _fields: [
          .../* @__PURE__ */ new Set([
            ...getNormalizedCommaSeparable(
              queryArgs._fields
            ) || [],
            "_links"
          ])
        ].join()
      } : {}
    },
    options
  );
  const ids = useMemo(
    () => data?.map(
      // @ts-ignore
      (record) => record[entityConfig?.key ?? "id"]
    ) ?? [],
    [data, entityConfig?.key]
  );
  const permissions = useSelect(
    (select) => {
      const { getEntityRecordsPermissions } = unlock(
        select(coreStore)
      );
      return getEntityRecordsPermissions(kind, name, ids);
    },
    [ids, kind, name]
  );
  const dataWithPermissions = useMemo(
    () => data?.map((record, index) => ({
      // @ts-ignore
      ...record,
      permissions: permissions[index]
    })) ?? [],
    [data, permissions]
  );
  return { records: dataWithPermissions, ...ret };
}
export {
  __experimentalUseEntityRecords,
  useEntityRecords as default,
  useEntityRecordsWithPermissions
};
