// @wordpress/block-editor/build/components/rich-text/content.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/rich-text/content.js
var content_exports = {};
__export(content_exports, {
  Content: () => Content,
  valueToHTMLString: () => valueToHTMLString
});
module.exports = __toCommonJS(content_exports);
var import_element = require("@wordpress/element");
var import_blocks = require("@wordpress/blocks");
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var import__ = __toESM(require("./index.cjs"));
var import_utils = require("./utils.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function valueToHTMLString(value, multiline) {
  if (import__.default.isEmpty(value)) {
    const multilineTag = (0, import_utils.getMultilineTag)(multiline);
    return multilineTag ? `<${multilineTag}></${multilineTag}>` : "";
  }
  if (Array.isArray(value)) {
    (0, import_deprecated.default)("wp.blockEditor.RichText value prop as children type", {
      since: "6.1",
      version: "6.3",
      alternative: "value prop as string",
      link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
    });
    return import_blocks.children.toHTML(value);
  }
  if (typeof value === "string") {
    return value;
  }
  return value.toHTMLString();
}
function Content({
  value,
  tagName: Tag,
  multiline,
  format,
  ...props
}) {
  value = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_element.RawHTML, { children: valueToHTMLString(value, multiline) });
  return Tag ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Tag, { ...props, children: value }) : value;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Content,
  valueToHTMLString
});
