// @wordpress/sync/build/utils.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/sync/src/utils.ts
var utils_exports = {};
__export(utils_exports, {
  createYjsDoc: () => createYjsDoc,
  deserializeCrdtDoc: () => deserializeCrdtDoc,
  initializeYjsDoc: () => initializeYjsDoc,
  markEntityAsSaved: () => markEntityAsSaved,
  serializeCrdtDoc: () => serializeCrdtDoc
});
module.exports = __toCommonJS(utils_exports);
var Y = __toESM(require("yjs"));
var buffer = __toESM(require("lib0/buffer"));
var import_config = require("./config.cjs");
function createYjsDoc(documentMeta = {}) {
  const metaMap = new Map(
    Object.entries(documentMeta)
  );
  return new Y.Doc({ meta: metaMap });
}
function initializeYjsDoc(ydoc) {
  const stateMap = ydoc.getMap(import_config.CRDT_STATE_MAP_KEY);
  stateMap.set(import_config.CRDT_STATE_MAP_VERSION_KEY, import_config.CRDT_DOC_VERSION);
}
function markEntityAsSaved(ydoc) {
  const recordMeta = ydoc.getMap(import_config.CRDT_STATE_MAP_KEY);
  recordMeta.set(import_config.CRDT_STATE_MAP_SAVED_AT_KEY, Date.now());
  recordMeta.set(import_config.CRDT_STATE_MAP_SAVED_BY_KEY, ydoc.clientID);
}
function pseudoRandomID() {
  return Math.floor(Math.random() * 1e9);
}
function serializeCrdtDoc(crdtDoc) {
  return JSON.stringify({
    document: buffer.toBase64(Y.encodeStateAsUpdateV2(crdtDoc)),
    updateId: pseudoRandomID()
    // helps with debugging
  });
}
function deserializeCrdtDoc(serializedCrdtDoc) {
  try {
    const { document } = JSON.parse(serializedCrdtDoc);
    const docMeta = {
      [import_config.CRDT_DOC_META_PERSISTENCE_KEY]: true
    };
    const ydoc = createYjsDoc(docMeta);
    const yupdate = buffer.fromBase64(document);
    Y.applyUpdateV2(ydoc, yupdate);
    ydoc.clientID = pseudoRandomID();
    return ydoc;
  } catch (e) {
    return null;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createYjsDoc,
  deserializeCrdtDoc,
  initializeYjsDoc,
  markEntityAsSaved,
  serializeCrdtDoc
});
