// @wordpress/global-styles-engine@1.9.0/build-module/settings/get-palette.mjs
// packages/global-styles-engine/src/settings/get-palette.ts
import { _x } from "@wordpress/i18n";
import { getSetting } from "./get-setting.mjs";
function getPalettes(settings) {
  if (!settings) {
    return {
      disableCustomColors: true,
      disableCustomGradients: true,
      colors: [],
      gradients: [],
      duotones: [],
      hasColorsOrGradients: false
    };
  }
  const [
    enableCustomColors,
    customColors,
    themeColors,
    defaultColors,
    shouldDisplayDefaultColors,
    enableCustomGradients,
    customGradients,
    themeGradients,
    defaultGradients,
    shouldDisplayDefaultGradients,
    shouldDisplayDefaultDuotones,
    customDuotones,
    themeDuotones,
    defaultDuotones
  ] = [
    "color.custom",
    "color.palette.custom",
    "color.palette.theme",
    "color.palette.default",
    "color.defaultPalette",
    "color.customGradient",
    "color.gradients.custom",
    "color.gradients.theme",
    "color.gradients.default",
    "color.defaultGradients",
    "color.defaultDuotone",
    "color.duotone.custom",
    "color.duotone.theme",
    "color.duotone.default"
  ].map((path) => getSetting(settings, path));
  const palettes = {
    disableCustomColors: !enableCustomColors,
    disableCustomGradients: !enableCustomGradients,
    colors: [],
    gradients: [],
    duotones: [],
    hasColorsOrGradients: false
  };
  if (themeColors && themeColors.length) {
    palettes.colors?.push({
      name: _x("Theme", "Indicates this palette comes from the theme."),
      slug: "theme",
      colors: themeColors
    });
  }
  if (shouldDisplayDefaultColors && defaultColors && defaultColors.length) {
    palettes.colors?.push({
      name: _x(
        "Default",
        "Indicates this palette comes from WordPress."
      ),
      slug: "default",
      colors: defaultColors
    });
  }
  if (customColors && customColors.length) {
    palettes.colors?.push({
      name: _x(
        "Custom",
        "Indicates this palette is created by the user."
      ),
      slug: "custom",
      colors: customColors
    });
  }
  if (themeGradients && themeGradients.length) {
    palettes.gradients?.push({
      name: _x("Theme", "Indicates this palette comes from the theme."),
      slug: "theme",
      gradients: themeGradients
    });
  }
  if (shouldDisplayDefaultGradients && defaultGradients && defaultGradients.length) {
    palettes.gradients?.push({
      name: _x(
        "Default",
        "Indicates this palette comes from WordPress."
      ),
      slug: "default",
      gradients: defaultGradients
    });
  }
  if (customGradients && customGradients.length) {
    palettes.gradients?.push({
      name: _x(
        "Custom",
        "Indicates this palette is created by the user."
      ),
      slug: "custom",
      gradients: customGradients
    });
  }
  if (themeDuotones && themeDuotones.length) {
    palettes.duotones?.push({
      name: _x(
        "Theme",
        "Indicates these duotone filters come from the theme."
      ),
      slug: "theme",
      duotones: themeDuotones
    });
  }
  if (shouldDisplayDefaultDuotones && defaultDuotones && defaultDuotones.length) {
    palettes.duotones?.push({
      name: _x(
        "Default",
        "Indicates these duotone filters come from WordPress."
      ),
      slug: "default",
      duotones: defaultDuotones
    });
  }
  if (customDuotones && customDuotones.length) {
    palettes.duotones?.push({
      name: _x(
        "Custom",
        "Indicates these doutone filters are created by the user."
      ),
      slug: "custom",
      duotones: customDuotones
    });
  }
  palettes.hasColorsOrGradients = !!palettes.colors?.length || !!palettes.gradients?.length;
  return palettes;
}
export {
  getPalettes as default
};
