// @wordpress/block-editor/build/components/block-alignment-control/constants.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-alignment-control/constants.js
var constants_exports = {};
__export(constants_exports, {
  BLOCK_ALIGNMENTS_CONTROLS: () => BLOCK_ALIGNMENTS_CONTROLS,
  DEFAULT_CONTROL: () => DEFAULT_CONTROL
});
module.exports = __toCommonJS(constants_exports);
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var BLOCK_ALIGNMENTS_CONTROLS = {
  none: {
    icon: import_icons.alignNone,
    title: (0, import_i18n._x)("None", "Alignment option")
  },
  left: {
    icon: import_icons.positionLeft,
    title: (0, import_i18n.__)("Align left")
  },
  center: {
    icon: import_icons.positionCenter,
    title: (0, import_i18n.__)("Align center")
  },
  right: {
    icon: import_icons.positionRight,
    title: (0, import_i18n.__)("Align right")
  },
  wide: {
    icon: import_icons.stretchWide,
    title: (0, import_i18n.__)("Wide width")
  },
  full: {
    icon: import_icons.stretchFullWidth,
    title: (0, import_i18n.__)("Full width")
  }
};
var DEFAULT_CONTROL = "none";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BLOCK_ALIGNMENTS_CONTROLS,
  DEFAULT_CONTROL
});
