// @wordpress/block-library/build/navigation/edit/navigation-menu-delete-control.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/navigation/edit/navigation-menu-delete-control.js
var navigation_menu_delete_control_exports = {};
__export(navigation_menu_delete_control_exports, {
  default: () => NavigationMenuDeleteControl
});
module.exports = __toCommonJS(navigation_menu_delete_control_exports);
var import_components = require("@wordpress/components");
var import_core_data = require("@wordpress/core-data");
var import_data = require("@wordpress/data");
var import_element = require("@wordpress/element");
var import_i18n = require("@wordpress/i18n");
var import_jsx_runtime = require("react/jsx-runtime");
function NavigationMenuDeleteControl({ onDelete }) {
  const [isConfirmDialogVisible, setIsConfirmDialogVisible] = (0, import_element.useState)(false);
  const id = (0, import_core_data.useEntityId)("postType", "wp_navigation");
  const { deleteEntityRecord } = (0, import_data.useDispatch)(import_core_data.store);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.Button,
      {
        __next40pxDefaultSize: true,
        className: "wp-block-navigation-delete-menu-button",
        variant: "secondary",
        isDestructive: true,
        onClick: () => {
          setIsConfirmDialogVisible(true);
        },
        children: (0, import_i18n.__)("Delete menu")
      }
    ),
    isConfirmDialogVisible && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.__experimentalConfirmDialog,
      {
        isOpen: true,
        onConfirm: () => {
          deleteEntityRecord("postType", "wp_navigation", id, {
            force: true
          });
          onDelete();
        },
        onCancel: () => {
          setIsConfirmDialogVisible(false);
        },
        confirmButtonText: (0, import_i18n.__)("Delete"),
        size: "medium",
        children: (0, import_i18n.__)(
          "Are you sure you want to delete this Navigation Menu?"
        )
      }
    )
  ] });
}
