// @wordpress/block-library/build/navigation/edit/overlay-menu-preview.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/navigation/edit/overlay-menu-preview.js
var overlay_menu_preview_exports = {};
__export(overlay_menu_preview_exports, {
  default: () => OverlayMenuPreview
});
module.exports = __toCommonJS(overlay_menu_preview_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_overlay_menu_icon = __toESM(require("./overlay-menu-icon.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function OverlayMenuPreview({ setAttributes, hasIcon, icon }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.__experimentalToolsPanelItem,
      {
        label: (0, import_i18n.__)("Show icon button"),
        isShownByDefault: true,
        hasValue: () => !hasIcon,
        onDeselect: () => setAttributes({ hasIcon: true }),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.ToggleControl,
          {
            label: (0, import_i18n.__)("Show icon button"),
            help: (0, import_i18n.__)(
              "Configure the visual appearance of the button that toggles the overlay menu."
            ),
            onChange: (value) => setAttributes({ hasIcon: value }),
            checked: hasIcon
          }
        )
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.__experimentalToolsPanelItem,
      {
        label: (0, import_i18n.__)("Icon"),
        isShownByDefault: true,
        hasValue: () => icon !== "handle",
        onDeselect: () => setAttributes({ icon: "handle" }),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_components.__experimentalToggleGroupControl,
          {
            __next40pxDefaultSize: true,
            className: "wp-block-navigation__overlay-menu-icon-toggle-group",
            label: (0, import_i18n.__)("Icon"),
            value: icon,
            onChange: (value) => setAttributes({ icon: value }),
            isBlock: true,
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_components.__experimentalToggleGroupControlOption,
                {
                  value: "handle",
                  "aria-label": (0, import_i18n.__)("handle"),
                  label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_overlay_menu_icon.default, { icon: "handle" })
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_components.__experimentalToggleGroupControlOption,
                {
                  value: "menu",
                  "aria-label": (0, import_i18n.__)("menu"),
                  label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_overlay_menu_icon.default, { icon: "menu" })
                }
              )
            ]
          }
        )
      }
    )
  ] });
}
