// @wordpress/dataviews/build/dataviews-picker/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/dataviews/src/dataviews-picker/index.tsx
var dataviews_picker_exports = {};
__export(dataviews_picker_exports, {
  default: () => dataviews_picker_default
});
module.exports = __toCommonJS(dataviews_picker_exports);
var import_element = require("@wordpress/element");
var import_compose = require("@wordpress/compose");
var import_ui = require("@wordpress/ui");
var import_dataviews_context = __toESM(require("../components/dataviews-context/index.cjs"));
var import_dataviews_layouts = require("../components/dataviews-layouts/index.cjs");
var import_dataviews_filters = require("../components/dataviews-filters/index.cjs");
var import_dataviews_layout = __toESM(require("../components/dataviews-layout/index.cjs"));
var import_dataviews_picker_footer = require("../components/dataviews-picker-footer/index.cjs");
var import_dataviews_search = __toESM(require("../components/dataviews-search/index.cjs"));
var import_dataviews_pagination = require("../components/dataviews-pagination/index.cjs");
var import_dataviews_view_config = __toESM(require("../components/dataviews-view-config/index.cjs"));
var import_field_types = __toESM(require("../field-types/index.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
var isItemClickable = () => false;
var dataViewsPickerLayouts = import_dataviews_layouts.VIEW_LAYOUTS.filter(
  (viewLayout) => viewLayout.isPicker
);
var defaultGetItemId = (item) => item.id;
var EMPTY_ARRAY = [];
function DefaultUI({
  search = true,
  searchLabel = void 0
}) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_ui.Stack,
      {
        direction: "row",
        align: "top",
        justify: "space-between",
        className: "dataviews__view-actions",
        gap: "xs",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_ui.Stack,
            {
              direction: "row",
              gap: "sm",
              justify: "start",
              className: "dataviews__search",
              children: [
                search && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_search.default, { label: searchLabel }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_filters.FiltersToggle, {})
              ]
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ui.Stack, { direction: "row", gap: "xs", style: { flexShrink: 0 }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_view_config.default, {}) })
        ]
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_filters.FiltersToggled, { className: "dataviews-filters__container" }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_layout.default, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_picker_footer.DataViewsPickerFooter, {})
  ] });
}
function DataViewsPicker({
  view,
  onChangeView,
  fields,
  search = true,
  searchLabel = void 0,
  actions = EMPTY_ARRAY,
  data,
  getItemId = defaultGetItemId,
  isLoading = false,
  paginationInfo,
  defaultLayouts: defaultLayoutsProperty,
  selection,
  onChangeSelection,
  children,
  config = { perPageSizes: [10, 20, 50, 100] },
  itemListLabel,
  empty
}) {
  const { infiniteScrollHandler } = paginationInfo;
  const containerRef = (0, import_element.useRef)(null);
  const [containerWidth, setContainerWidth] = (0, import_element.useState)(0);
  const resizeObserverRef = (0, import_compose.useResizeObserver)(
    (resizeObserverEntries) => {
      setContainerWidth(
        resizeObserverEntries[0].borderBoxSize[0].inlineSize
      );
    },
    { box: "border-box" }
  );
  const [openedFilter, setOpenedFilter] = (0, import_element.useState)(null);
  function setSelectionWithChange(value) {
    const newValue = typeof value === "function" ? value(selection) : value;
    if (onChangeSelection) {
      onChangeSelection(newValue);
    }
  }
  const _fields = (0, import_element.useMemo)(() => (0, import_field_types.default)(fields), [fields]);
  const filters = (0, import_dataviews_filters.useFilters)(_fields, view);
  const hasPrimaryOrLockedFilters = (0, import_element.useMemo)(
    () => (filters || []).some(
      (filter) => filter.isPrimary || filter.isLocked
    ),
    [filters]
  );
  const [isShowingFilter, setIsShowingFilter] = (0, import_element.useState)(
    hasPrimaryOrLockedFilters
  );
  (0, import_element.useEffect)(() => {
    if (hasPrimaryOrLockedFilters && !isShowingFilter) {
      setIsShowingFilter(true);
    }
  }, [hasPrimaryOrLockedFilters, isShowingFilter]);
  (0, import_element.useEffect)(() => {
    if (!view.infiniteScrollEnabled || !containerRef.current) {
      return;
    }
    const handleScroll = (0, import_compose.throttle)((event) => {
      const target = event.target;
      const scrollTop = target.scrollTop;
      const scrollHeight = target.scrollHeight;
      const clientHeight = target.clientHeight;
      if (scrollTop + clientHeight >= scrollHeight - 100) {
        infiniteScrollHandler?.();
      }
    }, 100);
    const container = containerRef.current;
    container.addEventListener("scroll", handleScroll);
    return () => {
      container.removeEventListener("scroll", handleScroll);
      handleScroll.cancel();
    };
  }, [infiniteScrollHandler, view.infiniteScrollEnabled]);
  const defaultLayouts = (0, import_element.useMemo)(
    () => Object.fromEntries(
      Object.entries(defaultLayoutsProperty).filter(
        ([layoutType]) => {
          return dataViewsPickerLayouts.some(
            (viewLayout) => viewLayout.type === layoutType
          );
        }
      )
    ),
    [defaultLayoutsProperty]
  );
  if (!defaultLayouts[view.type]) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_dataviews_context.default.Provider,
    {
      value: {
        view,
        onChangeView,
        fields: _fields,
        actions,
        data,
        isLoading,
        paginationInfo,
        isItemClickable,
        selection,
        onChangeSelection: setSelectionWithChange,
        openedFilter,
        setOpenedFilter,
        getItemId,
        containerWidth,
        containerRef,
        resizeObserverRef,
        defaultLayouts,
        filters,
        isShowingFilter,
        setIsShowingFilter,
        config,
        itemListLabel,
        empty,
        hasInitiallyLoaded: true,
        hasInfiniteScrollHandler: !!infiniteScrollHandler
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "dataviews-picker-wrapper", children: children ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DefaultUI, { search, searchLabel }) })
    }
  );
}
var DataViewsPickerSubComponents = DataViewsPicker;
DataViewsPickerSubComponents.BulkActionToolbar = import_dataviews_picker_footer.DataViewsPickerFooter;
DataViewsPickerSubComponents.Filters = import_dataviews_filters.Filters;
DataViewsPickerSubComponents.FiltersToggled = import_dataviews_filters.FiltersToggled;
DataViewsPickerSubComponents.FiltersToggle = import_dataviews_filters.FiltersToggle;
DataViewsPickerSubComponents.Layout = import_dataviews_layout.default;
DataViewsPickerSubComponents.LayoutSwitcher = import_dataviews_view_config.ViewTypeMenu;
DataViewsPickerSubComponents.Pagination = import_dataviews_pagination.DataViewsPagination;
DataViewsPickerSubComponents.Search = import_dataviews_search.default;
DataViewsPickerSubComponents.ViewConfig = import_dataviews_view_config.DataviewsViewConfigDropdown;
var dataviews_picker_default = DataViewsPickerSubComponents;
