// @wordpress/reusable-blocks@5.42.0/build-module/components/reusable-blocks-menu-items/reusable-block-convert-button.mjs
// packages/reusable-blocks/src/components/reusable-blocks-menu-items/reusable-block-convert-button.js
import { hasBlockSupport, isReusableBlock } from "@wordpress/blocks";
import { store as blockEditorStore } from "@wordpress/block-editor";
import { useCallback, useState } from "@wordpress/element";
import {
  MenuItem,
  Modal,
  Button,
  TextControl,
  __experimentalHStack as HStack,
  __experimentalVStack as VStack,
  ToggleControl
} from "@wordpress/components";
import { symbol } from "@wordpress/icons";
import { useDispatch, useSelect } from "@wordpress/data";
import { __, _x, sprintf } from "@wordpress/i18n";
import { store as noticesStore } from "@wordpress/notices";
import { store as coreStore } from "@wordpress/core-data";
import { store } from "../../store/index.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function ReusableBlockConvertButton({
  clientIds,
  rootClientId,
  onClose
}) {
  const [syncType, setSyncType] = useState(void 0);
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [title, setTitle] = useState("");
  const canConvert = useSelect(
    (select) => {
      const { canUser } = select(coreStore);
      const {
        getBlocksByClientId,
        canInsertBlockType,
        getBlockRootClientId
      } = select(blockEditorStore);
      const rootId = rootClientId || (clientIds.length > 0 ? getBlockRootClientId(clientIds[0]) : void 0);
      const blocks = getBlocksByClientId(clientIds) ?? [];
      const isReusable = blocks.length === 1 && blocks[0] && isReusableBlock(blocks[0]) && !!select(coreStore).getEntityRecord(
        "postType",
        "wp_block",
        blocks[0].attributes.ref
      );
      const _canConvert = (
        // Hide when this is already a reusable block.
        !isReusable && // Hide when reusable blocks are disabled.
        canInsertBlockType("core/block", rootId) && blocks.every(
          (block) => (
            // Guard against the case where a regular block has *just* been converted.
            !!block && // Hide on invalid blocks.
            block.isValid && // Hide when block doesn't support being made reusable.
            hasBlockSupport(block.name, "reusable", true)
          )
        ) && // Hide when current doesn't have permission to do that.
        // Blocks refers to the wp_block post type, this checks the ability to create a post of that type.
        !!canUser("create", {
          kind: "postType",
          name: "wp_block"
        })
      );
      return _canConvert;
    },
    [clientIds, rootClientId]
  );
  const { __experimentalConvertBlocksToReusable: convertBlocksToReusable } = useDispatch(store);
  const { createSuccessNotice, createErrorNotice } = useDispatch(noticesStore);
  const onConvert = useCallback(
    async function(reusableBlockTitle) {
      try {
        await convertBlocksToReusable(
          clientIds,
          reusableBlockTitle,
          syncType
        );
        createSuccessNotice(
          !syncType ? sprintf(
            // translators: %s: the name the user has given to the pattern.
            __("Synced pattern created: %s"),
            reusableBlockTitle
          ) : sprintf(
            // translators: %s: the name the user has given to the pattern.
            __("Unsynced pattern created: %s"),
            reusableBlockTitle
          ),
          {
            type: "snackbar",
            id: "convert-to-reusable-block-success"
          }
        );
      } catch (error) {
        createErrorNotice(error.message, {
          type: "snackbar",
          id: "convert-to-reusable-block-error"
        });
      }
    },
    [
      convertBlocksToReusable,
      clientIds,
      syncType,
      createSuccessNotice,
      createErrorNotice
    ]
  );
  if (!canConvert) {
    return null;
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(MenuItem, { icon: symbol, onClick: () => setIsModalOpen(true), children: __("Create pattern") }),
    isModalOpen && /* @__PURE__ */ jsx(
      Modal,
      {
        title: __("Create pattern"),
        onRequestClose: () => {
          setIsModalOpen(false);
          setTitle("");
        },
        overlayClassName: "reusable-blocks-menu-items__convert-modal",
        children: /* @__PURE__ */ jsx(
          "form",
          {
            onSubmit: (event) => {
              event.preventDefault();
              onConvert(title);
              setIsModalOpen(false);
              setTitle("");
              onClose();
            },
            children: /* @__PURE__ */ jsxs(VStack, { spacing: "5", children: [
              /* @__PURE__ */ jsx(
                TextControl,
                {
                  __next40pxDefaultSize: true,
                  label: __("Name"),
                  value: title,
                  onChange: setTitle,
                  placeholder: __("My pattern")
                }
              ),
              /* @__PURE__ */ jsx(
                ToggleControl,
                {
                  label: _x("Synced", "pattern (singular)"),
                  help: __(
                    "Sync this pattern across multiple locations."
                  ),
                  checked: !syncType,
                  onChange: () => {
                    setSyncType(
                      !syncType ? "unsynced" : void 0
                    );
                  }
                }
              ),
              /* @__PURE__ */ jsxs(HStack, { justify: "right", children: [
                /* @__PURE__ */ jsx(
                  Button,
                  {
                    __next40pxDefaultSize: true,
                    variant: "tertiary",
                    onClick: () => {
                      setIsModalOpen(false);
                      setTitle("");
                    },
                    children: __("Cancel")
                  }
                ),
                /* @__PURE__ */ jsx(
                  Button,
                  {
                    __next40pxDefaultSize: true,
                    variant: "primary",
                    type: "submit",
                    children: __("Create")
                  }
                )
              ] })
            ] })
          }
        )
      }
    )
  ] });
}
export {
  ReusableBlockConvertButton as default
};
