// @wordpress/block-library@9.42.0/build/navigation/edit/deleted-navigation-warning.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/navigation/edit/deleted-navigation-warning.js
var deleted_navigation_warning_exports = {};
__export(deleted_navigation_warning_exports, {
  default: () => deleted_navigation_warning_default
});
module.exports = __toCommonJS(deleted_navigation_warning_exports);
var import_block_editor = require("@wordpress/block-editor");
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_element = require("@wordpress/element");
var import_jsx_runtime = require("react/jsx-runtime");
function DeletedNavigationWarning({ onCreateNew, isNotice = false }) {
  const [isButtonDisabled, setIsButtonDisabled] = (0, import_element.useState)(false);
  const handleButtonClick = () => {
    setIsButtonDisabled(true);
    onCreateNew();
  };
  const message = (0, import_element.createInterpolateElement)(
    (0, import_i18n.__)(
      "Navigation Menu has been deleted or is unavailable. <button>Create a new Menu?</button>"
    ),
    {
      button: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.Button,
        {
          __next40pxDefaultSize: true,
          onClick: handleButtonClick,
          variant: "link",
          disabled: isButtonDisabled,
          accessibleWhenDisabled: true
        }
      )
    }
  );
  return isNotice ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.Notice, { status: "warning", isDismissible: false, children: message }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_block_editor.Warning, { children: message });
}
var deleted_navigation_warning_default = DeletedNavigationWarning;
