// @wordpress/block-library@9.42.0/build/navigation/edit/overlay-menu-preview-button.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/navigation/edit/overlay-menu-preview-button.js
var overlay_menu_preview_button_exports = {};
__export(overlay_menu_preview_button_exports, {
  default: () => OverlayMenuPreviewButton
});
module.exports = __toCommonJS(overlay_menu_preview_button_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var import_overlay_menu_icon = __toESM(require("./overlay-menu-icon.cjs"));
var import_overlay_menu_preview_controls = __toESM(require("./overlay-menu-preview-controls.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function OverlayMenuPreviewButton({
  isResponsive,
  overlayMenuPreview,
  setOverlayMenuPreview,
  hasIcon,
  icon,
  setAttributes,
  overlayMenuPreviewClasses,
  overlayMenuPreviewId,
  containerStyle
}) {
  if (!isResponsive) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_components.Button,
      {
        __next40pxDefaultSize: true,
        className: overlayMenuPreviewClasses,
        onClick: () => setOverlayMenuPreview(!overlayMenuPreview),
        "aria-label": (0, import_i18n.__)("Overlay menu controls"),
        "aria-controls": overlayMenuPreviewId,
        "aria-expanded": overlayMenuPreview,
        children: [
          hasIcon && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_overlay_menu_icon.default, { icon }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icons.Icon, { icon: import_icons.close })
          ] }),
          !hasIcon && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: (0, import_i18n.__)("Menu") }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: (0, import_i18n.__)("Close") })
          ] })
        ]
      }
    ),
    overlayMenuPreview && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.__experimentalVStack,
      {
        id: overlayMenuPreviewId,
        spacing: 4,
        style: containerStyle,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_overlay_menu_preview_controls.default,
          {
            hasIcon,
            icon,
            setAttributes
          }
        )
      }
    )
  ] });
}
