// @wordpress/block-library@9.42.0/build/navigation/edit/use-inner-blocks.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/navigation/edit/use-inner-blocks.js
var use_inner_blocks_exports = {};
__export(use_inner_blocks_exports, {
  useInnerBlocks: () => useInnerBlocks
});
module.exports = __toCommonJS(use_inner_blocks_exports);
var import_data = require("@wordpress/data");
var import_block_editor = require("@wordpress/block-editor");
var EMPTY_ARRAY = [];
function useInnerBlocks(clientId) {
  return (0, import_data.useSelect)(
    (select) => {
      const { getBlock, getBlocks, hasSelectedInnerBlock } = select(import_block_editor.store);
      const _uncontrolledInnerBlocks = getBlock(clientId).innerBlocks;
      const _hasUncontrolledInnerBlocks = !!_uncontrolledInnerBlocks?.length;
      const _controlledInnerBlocks = _hasUncontrolledInnerBlocks ? EMPTY_ARRAY : getBlocks(clientId);
      return {
        innerBlocks: _hasUncontrolledInnerBlocks ? _uncontrolledInnerBlocks : _controlledInnerBlocks,
        hasUncontrolledInnerBlocks: _hasUncontrolledInnerBlocks,
        uncontrolledInnerBlocks: _uncontrolledInnerBlocks,
        controlledInnerBlocks: _controlledInnerBlocks,
        isInnerBlockSelected: hasSelectedInnerBlock(clientId, true)
      };
    },
    [clientId]
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useInnerBlocks
});
