// @wordpress/patterns@2.42.0/build-module/components/pattern-overrides-controls.mjs
// packages/patterns/src/components/pattern-overrides-controls.js
import { useState, useId } from "@wordpress/element";
import {
  InspectorControls,
  useBlockBindingsUtils
} from "@wordpress/block-editor";
import { BaseControl, Button } from "@wordpress/components";
import { __ } from "@wordpress/i18n";
import { PATTERN_OVERRIDES_BINDING_SOURCE } from "../constants.mjs";
import {
  AllowOverridesModal,
  DisallowOverridesModal
} from "./allow-overrides-modal.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function PatternOverridesControls({
  attributes,
  setAttributes,
  name: blockName
}) {
  const controlId = useId();
  const [showAllowOverridesModal, setShowAllowOverridesModal] = useState(false);
  const [showDisallowOverridesModal, setShowDisallowOverridesModal] = useState(false);
  const hasName = !!attributes.metadata?.name;
  const defaultBindings = attributes.metadata?.bindings?.__default;
  const hasOverrides = hasName && defaultBindings?.source === PATTERN_OVERRIDES_BINDING_SOURCE;
  const isConnectedToOtherSources = defaultBindings?.source && defaultBindings.source !== PATTERN_OVERRIDES_BINDING_SOURCE;
  const { updateBlockBindings } = useBlockBindingsUtils();
  function updateBindings(isChecked, customName) {
    if (customName) {
      setAttributes({
        metadata: {
          ...attributes.metadata,
          name: customName
        }
      });
    }
    updateBlockBindings({
      __default: isChecked ? { source: PATTERN_OVERRIDES_BINDING_SOURCE } : void 0
    });
  }
  if (isConnectedToOtherSources) {
    return null;
  }
  const hasUnsupportedImageAttributes = blockName === "core/image" && !!attributes.href?.length;
  const helpText = !hasOverrides && hasUnsupportedImageAttributes ? __(
    `Overrides currently don't support image links. Remove the link first before enabling overrides.`
  ) : __(
    "Allow changes to this block throughout instances of this pattern."
  );
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(InspectorControls, { group: "advanced", children: /* @__PURE__ */ jsx(
      BaseControl,
      {
        id: controlId,
        label: __("Overrides"),
        help: helpText,
        children: /* @__PURE__ */ jsx(
          Button,
          {
            __next40pxDefaultSize: true,
            className: "pattern-overrides-control__allow-overrides-button",
            variant: "secondary",
            "aria-haspopup": "dialog",
            onClick: () => {
              if (hasOverrides) {
                setShowDisallowOverridesModal(true);
              } else {
                setShowAllowOverridesModal(true);
              }
            },
            disabled: !hasOverrides && hasUnsupportedImageAttributes,
            accessibleWhenDisabled: true,
            children: hasOverrides ? __("Disable overrides") : __("Enable overrides")
          }
        )
      }
    ) }),
    showAllowOverridesModal && /* @__PURE__ */ jsx(
      AllowOverridesModal,
      {
        initialName: attributes.metadata?.name,
        onClose: () => setShowAllowOverridesModal(false),
        onSave: (newName) => {
          updateBindings(true, newName);
        }
      }
    ),
    showDisallowOverridesModal && /* @__PURE__ */ jsx(
      DisallowOverridesModal,
      {
        onClose: () => setShowDisallowOverridesModal(false),
        onSave: () => updateBindings(false)
      }
    )
  ] });
}
var pattern_overrides_controls_default = PatternOverridesControls;
export {
  pattern_overrides_controls_default as default
};
