// @wordpress/patterns@2.42.0/build-module/components/rename-pattern-category-modal.mjs
// packages/patterns/src/components/rename-pattern-category-modal.js
import {
  Modal,
  Button,
  TextControl,
  __experimentalHStack as HStack,
  __experimentalVStack as VStack
} from "@wordpress/components";
import { store as coreStore } from "@wordpress/core-data";
import { useDispatch } from "@wordpress/data";
import { useId, useRef, useState } from "@wordpress/element";
import { decodeEntities } from "@wordpress/html-entities";
import { __ } from "@wordpress/i18n";
import { store as noticesStore } from "@wordpress/notices";
import { speak } from "@wordpress/a11y";
import { CATEGORY_SLUG } from "./category-selector.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
function RenamePatternCategoryModal({
  category,
  existingCategories,
  onClose,
  onError,
  onSuccess,
  ...props
}) {
  const id = useId();
  const textControlRef = useRef();
  const [name, setName] = useState(decodeEntities(category.name));
  const [isSaving, setIsSaving] = useState(false);
  const [validationMessage, setValidationMessage] = useState(false);
  const validationMessageId = validationMessage ? `patterns-rename-pattern-category-modal__validation-message-${id}` : void 0;
  const { saveEntityRecord, invalidateResolution } = useDispatch(coreStore);
  const { createErrorNotice, createSuccessNotice } = useDispatch(noticesStore);
  const onChange = (newName) => {
    if (validationMessage) {
      setValidationMessage(void 0);
    }
    setName(newName);
  };
  const onSave = async (event) => {
    event.preventDefault();
    if (isSaving) {
      return;
    }
    if (!name || name === category.name) {
      const message = __("Please enter a new name for this category.");
      speak(message, "assertive");
      setValidationMessage(message);
      textControlRef.current?.focus();
      return;
    }
    if (existingCategories.patternCategories.find((existingCategory) => {
      return existingCategory.id !== category.id && existingCategory.label.toLowerCase() === name.toLowerCase();
    })) {
      const message = __(
        "This category already exists. Please use a different name."
      );
      speak(message, "assertive");
      setValidationMessage(message);
      textControlRef.current?.focus();
      return;
    }
    try {
      setIsSaving(true);
      const savedRecord = await saveEntityRecord(
        "taxonomy",
        CATEGORY_SLUG,
        {
          id: category.id,
          slug: category.slug,
          name
        }
      );
      invalidateResolution("getUserPatternCategories");
      onSuccess?.(savedRecord);
      onClose();
      createSuccessNotice(__("Pattern category renamed."), {
        type: "snackbar",
        id: "pattern-category-update"
      });
    } catch (error) {
      onError?.();
      createErrorNotice(error.message, {
        type: "snackbar",
        id: "pattern-category-update"
      });
    } finally {
      setIsSaving(false);
      setName("");
    }
  };
  const onRequestClose = () => {
    onClose();
    setName("");
  };
  return /* @__PURE__ */ jsx(
    Modal,
    {
      title: __("Rename"),
      onRequestClose,
      ...props,
      children: /* @__PURE__ */ jsx("form", { onSubmit: onSave, children: /* @__PURE__ */ jsxs(VStack, { spacing: "5", children: [
        /* @__PURE__ */ jsxs(VStack, { spacing: "2", children: [
          /* @__PURE__ */ jsx(
            TextControl,
            {
              ref: textControlRef,
              __next40pxDefaultSize: true,
              label: __("Name"),
              value: name,
              onChange,
              "aria-describedby": validationMessageId,
              required: true
            }
          ),
          validationMessage && /* @__PURE__ */ jsx(
            "span",
            {
              className: "patterns-rename-pattern-category-modal__validation-message",
              id: validationMessageId,
              children: validationMessage
            }
          )
        ] }),
        /* @__PURE__ */ jsxs(HStack, { justify: "right", children: [
          /* @__PURE__ */ jsx(
            Button,
            {
              __next40pxDefaultSize: true,
              variant: "tertiary",
              onClick: onRequestClose,
              children: __("Cancel")
            }
          ),
          /* @__PURE__ */ jsx(
            Button,
            {
              __next40pxDefaultSize: true,
              variant: "primary",
              type: "submit",
              "aria-disabled": !name || name === category.name || isSaving,
              isBusy: isSaving,
              children: __("Save")
            }
          )
        ] })
      ] }) })
    }
  );
}
export {
  RenamePatternCategoryModal as default
};
