// @wordpress/block-editor/build/components/inspector-controls-tabs/position-controls-panel.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/inspector-controls-tabs/position-controls-panel.js
var position_controls_panel_exports = {};
__export(position_controls_panel_exports, {
  default: () => position_controls_panel_default
});
module.exports = __toCommonJS(position_controls_panel_exports);
var import_components = require("@wordpress/components");
var import_data = require("@wordpress/data");
var import_i18n = require("@wordpress/i18n");
var import_groups = __toESM(require("../inspector-controls/groups.cjs"));
var import_inspector_controls = __toESM(require("../inspector-controls/index.cjs"));
var import_store = require("../../store/index.cjs");
var import_utils = require("../global-styles/utils.cjs");
var import_utils2 = require("../../hooks/utils.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var PositionControlsPanel = () => {
  const { selectedClientIds, selectedBlocks, hasPositionAttribute } = (0, import_data.useSelect)((select) => {
    const { getBlocksByClientId, getSelectedBlockClientIds } = select(import_store.store);
    const selectedBlockClientIds = getSelectedBlockClientIds();
    const _selectedBlocks = getBlocksByClientId(
      selectedBlockClientIds
    );
    return {
      selectedClientIds: selectedBlockClientIds,
      selectedBlocks: _selectedBlocks,
      hasPositionAttribute: _selectedBlocks?.some(
        ({ attributes }) => !!attributes?.style?.position?.type
      )
    };
  }, []);
  const { updateBlockAttributes } = (0, import_data.useDispatch)(import_store.store);
  const dropdownMenuProps = (0, import_utils.useToolsPanelDropdownMenuProps)();
  function resetPosition() {
    if (!selectedClientIds?.length || !selectedBlocks?.length) {
      return;
    }
    const attributesByClientId = Object.fromEntries(
      selectedBlocks?.map(({ clientId, attributes }) => [
        clientId,
        {
          style: (0, import_utils2.cleanEmptyObject)({
            ...attributes?.style,
            position: {
              ...attributes?.style?.position,
              type: void 0,
              top: void 0,
              right: void 0,
              bottom: void 0,
              left: void 0
            }
          })
        }
      ])
    );
    updateBlockAttributes(selectedClientIds, attributesByClientId, true);
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.__experimentalToolsPanel,
    {
      className: "block-editor-block-inspector__position",
      label: (0, import_i18n.__)("Position"),
      resetAll: resetPosition,
      dropdownMenuProps,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.__experimentalToolsPanelItem,
        {
          isShownByDefault: hasPositionAttribute,
          label: (0, import_i18n.__)("Position"),
          hasValue: () => hasPositionAttribute,
          onDeselect: resetPosition,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_inspector_controls.default.Slot, { group: "position" })
        }
      )
    }
  );
};
var PositionControls = () => {
  const fills = (0, import_components.__experimentalUseSlotFills)(import_groups.default.position.name);
  const hasFills = Boolean(fills && fills.length);
  if (!hasFills) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(PositionControlsPanel, {});
};
var position_controls_panel_default = PositionControls;
