// @wordpress/core-data@7.42.0/build/hooks/use-entity-records.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/hooks/use-entity-records.ts
var use_entity_records_exports = {};
__export(use_entity_records_exports, {
  __experimentalUseEntityRecords: () => __experimentalUseEntityRecords,
  default: () => useEntityRecords,
  useEntityRecordsWithPermissions: () => useEntityRecordsWithPermissions
});
module.exports = __toCommonJS(use_entity_records_exports);
var import_url = require("@wordpress/url");
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var import_data = require("@wordpress/data");
var import_element = require("@wordpress/element");
var import_use_query_select = __toESM(require("./use-query-select.cjs"));
var import__ = require("../index.cjs");
var import_lock_unlock = require("../lock-unlock.cjs");
var import_utils = require("../utils/index.cjs");
var EMPTY_ARRAY = [];
function useEntityRecords(kind, name, queryArgs = {}, options = { enabled: true }) {
  const queryAsString = (0, import_url.addQueryArgs)("", queryArgs);
  const { data: records, ...rest } = (0, import_use_query_select.default)(
    (query) => {
      if (!options.enabled) {
        return {
          // Avoiding returning a new reference on every execution.
          data: EMPTY_ARRAY
        };
      }
      return query(import__.store).getEntityRecords(kind, name, queryArgs);
    },
    [kind, name, queryAsString, options.enabled]
  );
  const { totalItems, totalPages } = (0, import_data.useSelect)(
    (select) => {
      if (!options.enabled) {
        return {
          totalItems: null,
          totalPages: null
        };
      }
      return {
        totalItems: select(import__.store).getEntityRecordsTotalItems(
          kind,
          name,
          queryArgs
        ),
        totalPages: select(import__.store).getEntityRecordsTotalPages(
          kind,
          name,
          queryArgs
        )
      };
    },
    [kind, name, queryAsString, options.enabled]
  );
  return {
    records,
    totalItems,
    totalPages,
    ...rest
  };
}
function __experimentalUseEntityRecords(kind, name, queryArgs, options) {
  (0, import_deprecated.default)(`wp.data.__experimentalUseEntityRecords`, {
    alternative: "wp.data.useEntityRecords",
    since: "6.1"
  });
  return useEntityRecords(kind, name, queryArgs, options);
}
function useEntityRecordsWithPermissions(kind, name, queryArgs = {}, options = { enabled: true }) {
  const entityConfig = (0, import_data.useSelect)(
    (select) => select(import__.store).getEntityConfig(kind, name),
    [kind, name]
  );
  const { records: data, ...ret } = useEntityRecords(
    kind,
    name,
    {
      ...queryArgs,
      // If _fields is provided, we need to include _links in the request for permission caching to work.
      ...queryArgs._fields ? {
        _fields: [
          .../* @__PURE__ */ new Set([
            ...(0, import_utils.getNormalizedCommaSeparable)(
              queryArgs._fields
            ) || [],
            "_links"
          ])
        ].join()
      } : {}
    },
    options
  );
  const ids = (0, import_element.useMemo)(
    () => data?.map(
      // @ts-ignore
      (record) => record[entityConfig?.key ?? "id"]
    ) ?? [],
    [data, entityConfig?.key]
  );
  const permissions = (0, import_data.useSelect)(
    (select) => {
      const { getEntityRecordsPermissions } = (0, import_lock_unlock.unlock)(
        select(import__.store)
      );
      return getEntityRecordsPermissions(kind, name, ids);
    },
    [ids, kind, name]
  );
  const dataWithPermissions = (0, import_element.useMemo)(
    () => data?.map((record, index) => ({
      // @ts-ignore
      ...record,
      permissions: permissions[index]
    })) ?? [],
    [data, permissions]
  );
  return { records: dataWithPermissions, ...ret };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  __experimentalUseEntityRecords,
  useEntityRecordsWithPermissions
});
