// @wordpress/reusable-blocks/build-module/store/actions.mjs
// packages/reusable-blocks/src/store/actions.js
import { store as blockEditorStore } from "@wordpress/block-editor";
import {
  createBlock,
  isReusableBlock,
  parse,
  serialize
} from "@wordpress/blocks";
import { __ } from "@wordpress/i18n";
var __experimentalConvertBlockToStatic = (clientId) => ({ registry }) => {
  const oldBlock = registry.select(blockEditorStore).getBlock(clientId);
  const reusableBlock = registry.select("core").getEditedEntityRecord(
    "postType",
    "wp_block",
    oldBlock.attributes.ref
  );
  const newBlocks = parse(
    typeof reusableBlock.content === "function" ? reusableBlock.content(reusableBlock) : reusableBlock.content
  );
  registry.dispatch(blockEditorStore).replaceBlocks(oldBlock.clientId, newBlocks);
};
var __experimentalConvertBlocksToReusable = (clientIds, title, syncType) => async ({ registry, dispatch }) => {
  const meta = syncType === "unsynced" ? {
    wp_pattern_sync_status: syncType
  } : void 0;
  const reusableBlock = {
    title: title || __("Untitled pattern block"),
    content: serialize(
      registry.select(blockEditorStore).getBlocksByClientId(clientIds)
    ),
    status: "publish",
    meta
  };
  const updatedRecord = await registry.dispatch("core").saveEntityRecord("postType", "wp_block", reusableBlock);
  if (syncType === "unsynced") {
    return;
  }
  const newBlock = createBlock("core/block", {
    ref: updatedRecord.id
  });
  registry.dispatch(blockEditorStore).replaceBlocks(clientIds, newBlock);
  dispatch.__experimentalSetEditingReusableBlock(
    newBlock.clientId,
    true
  );
};
var __experimentalDeleteReusableBlock = (id) => async ({ registry }) => {
  const reusableBlock = registry.select("core").getEditedEntityRecord("postType", "wp_block", id);
  if (!reusableBlock) {
    return;
  }
  const allBlocks = registry.select(blockEditorStore).getBlocks();
  const associatedBlocks = allBlocks.filter(
    (block) => isReusableBlock(block) && block.attributes.ref === id
  );
  const associatedBlockClientIds = associatedBlocks.map(
    (block) => block.clientId
  );
  if (associatedBlockClientIds.length) {
    registry.dispatch(blockEditorStore).removeBlocks(associatedBlockClientIds);
  }
  await registry.dispatch("core").deleteEntityRecord("postType", "wp_block", id);
};
function __experimentalSetEditingReusableBlock(clientId, isEditing) {
  return {
    type: "SET_EDITING_REUSABLE_BLOCK",
    clientId,
    isEditing
  };
}
export {
  __experimentalConvertBlockToStatic,
  __experimentalConvertBlocksToReusable,
  __experimentalDeleteReusableBlock,
  __experimentalSetEditingReusableBlock
};
