// @wordpress/patterns/build-module/components/patterns-manage-button.mjs
// packages/patterns/src/components/patterns-manage-button.js
import {
  MenuItem,
  __experimentalConfirmDialog as ConfirmDialog
} from "@wordpress/components";
import { __ } from "@wordpress/i18n";
import { isReusableBlock } from "@wordpress/blocks";
import { useSelect, useDispatch } from "@wordpress/data";
import { useState } from "@wordpress/element";
import { store as blockEditorStore } from "@wordpress/block-editor";
import { addQueryArgs } from "@wordpress/url";
import { store as coreStore } from "@wordpress/core-data";
import { store as patternsStore } from "../store/index.mjs";
import { unlock } from "../lock-unlock.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function PatternsManageButton({ clientId, onClose }) {
  const [showConfirmDialog, setShowConfirmDialog] = useState(false);
  const {
    attributes,
    canDetach,
    isVisible,
    managePatternsUrl,
    isSyncedPattern,
    isUnsyncedPattern,
    canEdit
  } = useSelect(
    (select) => {
      const { canRemoveBlock, getBlock, canEditBlock } = select(blockEditorStore);
      const { canUser } = select(coreStore);
      const block = getBlock(clientId);
      const _isUnsyncedPattern = !!block?.attributes?.metadata?.patternName;
      const _isSyncedPattern = !!block && isReusableBlock(block) && !!canUser("update", {
        kind: "postType",
        name: "wp_block",
        id: block.attributes.ref
      });
      return {
        attributes: block.attributes,
        canEdit: canEditBlock(clientId),
        // For unsynced patterns, detaching is simply removing the `patternName` attribute.
        // For synced patterns, the `core:block` block is replaced with its inner blocks,
        // so checking whether `canRemoveBlock` is possible is required.
        canDetach: _isUnsyncedPattern || _isSyncedPattern && canRemoveBlock(clientId),
        isUnsyncedPattern: _isUnsyncedPattern,
        isSyncedPattern: _isSyncedPattern,
        isVisible: _isUnsyncedPattern || _isSyncedPattern,
        // The site editor and templates both check whether the user
        // has edit_theme_options capabilities. We can leverage that here
        // and omit the manage patterns link if the user can't access it.
        managePatternsUrl: canUser("create", {
          kind: "postType",
          name: "wp_template"
        }) ? addQueryArgs("site-editor.php", {
          p: "/pattern"
        }) : addQueryArgs("edit.php", {
          post_type: "wp_block"
        })
      };
    },
    [clientId]
  );
  const { updateBlockAttributes } = useDispatch(blockEditorStore);
  const { convertSyncedPatternToStatic } = unlock(
    useDispatch(patternsStore)
  );
  if (!isVisible || !canEdit) {
    return null;
  }
  const handleDetach = () => {
    if (isSyncedPattern) {
      convertSyncedPatternToStatic(clientId);
    }
    if (isUnsyncedPattern) {
      const { patternName, ...attributesWithoutPatternName } = attributes?.metadata ?? {};
      updateBlockAttributes(clientId, {
        metadata: attributesWithoutPatternName
      });
    }
    onClose?.();
    setShowConfirmDialog(false);
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    canDetach && /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(MenuItem, { onClick: () => setShowConfirmDialog(true), children: isSyncedPattern ? __("Disconnect pattern") : __("Detach pattern") }),
      /* @__PURE__ */ jsx(
        ConfirmDialog,
        {
          isOpen: showConfirmDialog,
          onConfirm: handleDetach,
          onCancel: () => setShowConfirmDialog(false),
          confirmButtonText: isSyncedPattern ? __("Disconnect") : __("Detach"),
          size: "medium",
          title: isSyncedPattern ? __("Disconnect pattern?") : __("Detach pattern?"),
          __experimentalHideHeader: false,
          children: isSyncedPattern ? __(
            "The blocks will be separated from the original pattern and will be fully editable. Future changes to the pattern will not apply here."
          ) : __(
            "Blocks will no longer be associated with this pattern and will be fully editable."
          )
        }
      )
    ] }),
    /* @__PURE__ */ jsx(MenuItem, { href: managePatternsUrl, children: __("Manage patterns") })
  ] });
}
var patterns_manage_button_default = PatternsManageButton;
export {
  patterns_manage_button_default as default
};
