// @wordpress/block-editor/build/components/provider/use-media-upload-settings.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/provider/use-media-upload-settings.js
var use_media_upload_settings_exports = {};
__export(use_media_upload_settings_exports, {
  default: () => use_media_upload_settings_default
});
module.exports = __toCommonJS(use_media_upload_settings_exports);
var import_element = require("@wordpress/element");
function useMediaUploadSettings(settings = {}) {
  return (0, import_element.useMemo)(
    () => ({
      mediaUpload: settings.mediaUpload,
      mediaSideload: settings.mediaSideload,
      mediaFinalize: settings.mediaFinalize,
      maxUploadFileSize: settings.maxUploadFileSize,
      allowedMimeTypes: settings.allowedMimeTypes,
      allImageSizes: settings.allImageSizes,
      bigImageSizeThreshold: settings.bigImageSizeThreshold
    }),
    [settings]
  );
}
var use_media_upload_settings_default = useMediaUploadSettings;
