// @wordpress/block-library@9.42.0/build/navigation-link/link-ui/block-inserter.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/navigation-link/link-ui/block-inserter.js
var block_inserter_exports = {};
__export(block_inserter_exports, {
  default: () => block_inserter_default
});
module.exports = __toCommonJS(block_inserter_exports);
var import_i18n = require("@wordpress/i18n");
var import_data = require("@wordpress/data");
var import_block_editor = require("@wordpress/block-editor");
var import_dialog_wrapper = __toESM(require("./dialog-wrapper.cjs"));
var import_lock_unlock = require("../../lock-unlock.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var { PrivateQuickInserter: QuickInserter } = (0, import_lock_unlock.unlock)(
  import_block_editor.privateApis
);
function LinkUIBlockInserter({ clientId, onBack, onBlockInsert }) {
  const { rootBlockClientId } = (0, import_data.useSelect)(
    (select) => {
      const { getBlockRootClientId } = select(import_block_editor.store);
      return {
        rootBlockClientId: getBlockRootClientId(clientId)
      };
    },
    [clientId]
  );
  if (!clientId) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_dialog_wrapper.default,
    {
      className: "link-ui-block-inserter",
      title: (0, import_i18n.__)("Add block"),
      description: (0, import_i18n.__)("Choose a block to add to your Navigation."),
      onBack,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        QuickInserter,
        {
          rootClientId: rootBlockClientId,
          clientId,
          isAppender: false,
          prioritizePatterns: false,
          selectBlockOnInsert: !onBlockInsert,
          onSelect: onBlockInsert ? onBlockInsert : void 0,
          hasSearch: false
        }
      )
    }
  );
}
var block_inserter_default = LinkUIBlockInserter;
