// @wordpress/block-library@9.42.0/build/navigation-link/link-ui/dialog-wrapper.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/navigation-link/link-ui/dialog-wrapper.js
var dialog_wrapper_exports = {};
__export(dialog_wrapper_exports, {
  default: () => dialog_wrapper_default
});
module.exports = __toCommonJS(dialog_wrapper_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var import_compose = require("@wordpress/compose");
var import_jsx_runtime = require("react/jsx-runtime");
function BackButton({ className, onBack }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Button,
    {
      className,
      icon: (0, import_i18n.isRTL)() ? import_icons.chevronRightSmall : import_icons.chevronLeftSmall,
      onClick: (e) => {
        e.preventDefault();
        onBack();
      },
      size: "small",
      children: (0, import_i18n.__)("Back")
    }
  );
}
function DialogWrapper({ className, title, description, onBack, children }) {
  const dialogTitleId = (0, import_compose.useInstanceId)(
    DialogWrapper,
    "link-ui-dialog-title"
  );
  const dialogDescriptionId = (0, import_compose.useInstanceId)(
    DialogWrapper,
    "link-ui-dialog-description"
  );
  const focusOnMountRef = (0, import_compose.useFocusOnMount)("firstElement");
  const backButtonClassName = `${className}__back`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      className,
      role: "dialog",
      "aria-labelledby": dialogTitleId,
      "aria-describedby": dialogDescriptionId,
      ref: focusOnMountRef,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.VisuallyHidden, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h2", { id: dialogTitleId, children: title }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("p", { id: dialogDescriptionId, children: description })
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(BackButton, { className: backButtonClassName, onBack }),
        children
      ]
    }
  );
}
var dialog_wrapper_default = DialogWrapper;
