// @wordpress/block-library@9.42.0/build/navigation-link/link-ui/page-creator.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/navigation-link/link-ui/page-creator.js
var page_creator_exports = {};
__export(page_creator_exports, {
  LinkUIPageCreator: () => LinkUIPageCreator
});
module.exports = __toCommonJS(page_creator_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_data = require("@wordpress/data");
var import_core_data = require("@wordpress/core-data");
var import_notices = require("@wordpress/notices");
var import_html_entities = require("@wordpress/html-entities");
var import_element = require("@wordpress/element");
var import_dialog_wrapper = __toESM(require("./dialog-wrapper.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function LinkUIPageCreator({
  postType,
  onBack,
  onPageCreated,
  initialTitle = ""
}) {
  const [title, setTitle] = (0, import_element.useState)(initialTitle);
  const [shouldPublish, setShouldPublish] = (0, import_element.useState)(true);
  const isTitleValid = title.trim().length > 0;
  const { lastError, isSaving } = (0, import_data.useSelect)(
    (select) => ({
      lastError: select(import_core_data.store).getLastEntitySaveError(
        "postType",
        postType
      ),
      isSaving: select(import_core_data.store).isSavingEntityRecord(
        "postType",
        postType
      )
    }),
    [postType]
  );
  const { saveEntityRecord } = (0, import_data.useDispatch)(import_core_data.store);
  const { createSuccessNotice, createErrorNotice } = (0, import_data.useDispatch)(import_notices.store);
  async function createPage(event) {
    event.preventDefault();
    if (isSaving || !isTitleValid) {
      return;
    }
    try {
      const savedRecord = await saveEntityRecord(
        "postType",
        postType,
        {
          title,
          status: shouldPublish ? "publish" : "draft"
        },
        { throwOnError: true }
      );
      if (savedRecord) {
        const pageLink = {
          id: savedRecord.id,
          type: postType,
          title: (0, import_html_entities.decodeEntities)(savedRecord.title.rendered),
          url: savedRecord.link,
          kind: "post-type"
        };
        createSuccessNotice(
          (0, import_i18n.sprintf)(
            // translators: %s: the name of the new page being created.
            (0, import_i18n.__)("%s page created successfully."),
            (0, import_html_entities.decodeEntities)(savedRecord.title.rendered)
          ),
          {
            type: "snackbar",
            id: "page-created-success"
          }
        );
        onPageCreated(pageLink);
      }
    } catch (error) {
      createErrorNotice(
        (0, import_i18n.__)("Failed to create page. Please try again."),
        {
          type: "snackbar",
          id: "page-created-error"
        }
      );
    }
  }
  const isSubmitDisabled = isSaving || !isTitleValid;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_dialog_wrapper.default,
    {
      className: "link-ui-page-creator",
      title: (0, import_i18n.__)("Create page"),
      description: (0, import_i18n.__)("Create a new page to add to your Navigation."),
      onBack,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.__experimentalVStack, { className: "link-ui-page-creator__inner", spacing: 4, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("form", { onSubmit: createPage, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalVStack, { spacing: 4, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.TextControl,
          {
            __next40pxDefaultSize: true,
            label: (0, import_i18n.__)("Title"),
            onChange: setTitle,
            placeholder: (0, import_i18n.__)("No title"),
            value: title
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.CheckboxControl,
          {
            label: (0, import_i18n.__)("Publish"),
            help: (0, import_i18n.__)(
              "Turn off to save as a draft. Drafts won't appear on your site until published."
            ),
            checked: shouldPublish,
            onChange: setShouldPublish
          }
        ),
        lastError && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.Notice, { status: "error", isDismissible: false, children: lastError.message }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalHStack, { spacing: 2, justify: "flex-end", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              __next40pxDefaultSize: true,
              variant: "tertiary",
              onClick: onBack,
              disabled: isSaving,
              accessibleWhenDisabled: true,
              children: (0, import_i18n.__)("Cancel")
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              __next40pxDefaultSize: true,
              variant: "primary",
              type: "submit",
              isBusy: isSaving,
              "aria-disabled": isSubmitDisabled,
              children: (0, import_i18n.__)("Create page")
            }
          )
        ] })
      ] }) }) })
    }
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LinkUIPageCreator
});
