// @wordpress/block-editor/build/components/inserter/block-patterns-explorer/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/inserter/block-patterns-explorer/index.js
var block_patterns_explorer_exports = {};
__export(block_patterns_explorer_exports, {
  default: () => block_patterns_explorer_default
});
module.exports = __toCommonJS(block_patterns_explorer_exports);
var import_components = require("@wordpress/components");
var import_element = require("@wordpress/element");
var import_i18n = require("@wordpress/i18n");
var import_pattern_explorer_sidebar = __toESM(require("./pattern-explorer-sidebar.cjs"));
var import_pattern_list = __toESM(require("./pattern-list.cjs"));
var import_use_pattern_categories = require("../block-patterns-tab/use-pattern-categories.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function PatternsExplorer({ initialCategory, rootClientId, onModalClose }) {
  const [searchValue, setSearchValue] = (0, import_element.useState)("");
  const [selectedCategory, setSelectedCategory] = (0, import_element.useState)(
    initialCategory?.name
  );
  const patternCategories = (0, import_use_pattern_categories.usePatternCategories)(rootClientId);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "block-editor-block-patterns-explorer", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_pattern_explorer_sidebar.default,
      {
        selectedCategory,
        patternCategories,
        onClickCategory: setSelectedCategory,
        searchValue,
        setSearchValue
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_pattern_list.default,
      {
        searchValue,
        selectedCategory,
        patternCategories,
        rootClientId,
        onModalClose
      }
    )
  ] });
}
function PatternsExplorerModal({ onModalClose, ...restProps }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Modal,
    {
      title: (0, import_i18n.__)("Patterns"),
      onRequestClose: onModalClose,
      isFullScreen: true,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(PatternsExplorer, { onModalClose, ...restProps })
    }
  );
}
var block_patterns_explorer_default = PatternsExplorerModal;
