// @wordpress/block-library@9.42.0/build/terms-query/edit/terms-query-placeholder.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/terms-query/edit/terms-query-placeholder.js
var terms_query_placeholder_exports = {};
__export(terms_query_placeholder_exports, {
  default: () => TermsQueryPlaceholder
});
module.exports = __toCommonJS(terms_query_placeholder_exports);
var import_data = require("@wordpress/data");
var import_blocks = require("@wordpress/blocks");
var import_block_editor = require("@wordpress/block-editor");
var import_jsx_runtime = require("react/jsx-runtime");
function TermsQueryPlaceholder({
  attributes,
  clientId,
  name
}) {
  const { blockType, activeBlockVariation, scopeVariations } = (0, import_data.useSelect)(
    (select) => {
      const {
        getActiveBlockVariation,
        getBlockType,
        getBlockVariations
      } = select(import_blocks.store);
      return {
        blockType: getBlockType(name),
        activeBlockVariation: getActiveBlockVariation(
          name,
          attributes
        ),
        scopeVariations: getBlockVariations(name, "block")
      };
    },
    [name, attributes]
  );
  const icon = activeBlockVariation?.icon?.src || activeBlockVariation?.icon || blockType?.icon?.src;
  const label = activeBlockVariation?.title || blockType?.title;
  const { replaceInnerBlocks } = (0, import_data.useDispatch)(import_block_editor.store);
  const blockProps = (0, import_block_editor.useBlockProps)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ...blockProps, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_block_editor.__experimentalBlockVariationPicker,
    {
      icon,
      label,
      variations: scopeVariations,
      onSelect: (variation) => {
        if (variation.innerBlocks) {
          replaceInnerBlocks(
            clientId,
            (0, import_blocks.createBlocksFromInnerBlocksTemplate)(
              variation.innerBlocks
            ),
            false
          );
        }
      }
    }
  ) });
}
