// @wordpress/block-editor/build/components/recursion-provider/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/recursion-provider/index.js
var recursion_provider_exports = {};
__export(recursion_provider_exports, {
  DeprecatedExperimentalRecursionProvider: () => DeprecatedExperimentalRecursionProvider,
  DeprecatedExperimentalUseHasRecursion: () => DeprecatedExperimentalUseHasRecursion,
  RecursionProvider: () => RecursionProvider,
  useHasRecursion: () => useHasRecursion
});
module.exports = __toCommonJS(recursion_provider_exports);
var import_element = require("@wordpress/element");
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var import_context = require("../block-edit/context.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var RenderedRefsContext = (0, import_element.createContext)({});
RenderedRefsContext.displayName = "RenderedRefsContext";
function addToBlockType(renderedBlocks, blockName, uniqueId) {
  const result = {
    ...renderedBlocks,
    [blockName]: renderedBlocks[blockName] ? new Set(renderedBlocks[blockName]) : /* @__PURE__ */ new Set()
  };
  result[blockName].add(uniqueId);
  return result;
}
function RecursionProvider({ children, uniqueId, blockName = "" }) {
  const previouslyRenderedBlocks = (0, import_element.useContext)(RenderedRefsContext);
  const { name } = (0, import_context.useBlockEditContext)();
  blockName = blockName || name;
  const newRenderedBlocks = (0, import_element.useMemo)(
    () => addToBlockType(previouslyRenderedBlocks, blockName, uniqueId),
    [previouslyRenderedBlocks, blockName, uniqueId]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(RenderedRefsContext.Provider, { value: newRenderedBlocks, children });
}
function useHasRecursion(uniqueId, blockName = "") {
  const previouslyRenderedBlocks = (0, import_element.useContext)(RenderedRefsContext);
  const { name } = (0, import_context.useBlockEditContext)();
  blockName = blockName || name;
  return Boolean(previouslyRenderedBlocks[blockName]?.has(uniqueId));
}
var DeprecatedExperimentalRecursionProvider = (props) => {
  (0, import_deprecated.default)("wp.blockEditor.__experimentalRecursionProvider", {
    since: "6.5",
    alternative: "wp.blockEditor.RecursionProvider"
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(RecursionProvider, { ...props });
};
var DeprecatedExperimentalUseHasRecursion = (...args) => {
  (0, import_deprecated.default)("wp.blockEditor.__experimentalUseHasRecursion", {
    since: "6.5",
    alternative: "wp.blockEditor.useHasRecursion"
  });
  return useHasRecursion(...args);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DeprecatedExperimentalRecursionProvider,
  DeprecatedExperimentalUseHasRecursion,
  RecursionProvider,
  useHasRecursion
});
