// @wordpress/block-editor@15.15.0/build/components/publish-date-time-picker/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/publish-date-time-picker/index.js
var publish_date_time_picker_exports = {};
__export(publish_date_time_picker_exports, {
  PrivatePublishDateTimePicker: () => PrivatePublishDateTimePicker,
  PublishDateTimePicker: () => PublishDateTimePicker,
  default: () => publish_date_time_picker_default
});
module.exports = __toCommonJS(publish_date_time_picker_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_element = require("@wordpress/element");
var import_date = require("@wordpress/date");
var import_inspector_popover_header = __toESM(require("../inspector-popover-header/index.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function PublishDateTimePicker({
  onClose,
  onChange,
  showPopoverHeaderActions,
  isCompact,
  currentDate,
  title,
  ...additionalProps
}, ref) {
  const datePickerProps = {
    startOfWeek: (0, import_date.getSettings)().l10n.startOfWeek,
    onChange,
    currentDate: isCompact ? void 0 : currentDate,
    currentTime: isCompact ? currentDate : void 0,
    ...additionalProps
  };
  const DatePickerComponent = isCompact ? import_components.TimePicker : import_components.DateTimePicker;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { ref, className: "block-editor-publish-date-time-picker", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_inspector_popover_header.default,
      {
        title: title || (0, import_i18n.__)("Publish"),
        actions: showPopoverHeaderActions ? [
          {
            label: (0, import_i18n.__)("Now"),
            onClick: () => onChange?.(null)
          }
        ] : void 0,
        onClose
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DatePickerComponent, { ...datePickerProps })
  ] });
}
var PrivatePublishDateTimePicker = (0, import_element.forwardRef)(PublishDateTimePicker);
function PublicPublishDateTimePicker(props, ref) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    PrivatePublishDateTimePicker,
    {
      ...props,
      showPopoverHeaderActions: true,
      isCompact: false,
      ref
    }
  );
}
var publish_date_time_picker_default = (0, import_element.forwardRef)(PublicPublishDateTimePicker);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PrivatePublishDateTimePicker,
  PublishDateTimePicker
});
