// @wordpress/dataviews@13.1.0/build-module/components/dataviews-view-config/index.mjs
// packages/dataviews/src/components/dataviews-view-config/index.tsx
import {
  Button,
  __experimentalDropdownContentWrapper as DropdownContentWrapper,
  Dropdown,
  __experimentalToggleGroupControl as ToggleGroupControl,
  __experimentalToggleGroupControlOption as ToggleGroupControlOption,
  __experimentalToggleGroupControlOptionIcon as ToggleGroupControlOptionIcon,
  SelectControl,
  __experimentalHeading as Heading,
  privateApis as componentsPrivateApis
} from "@wordpress/components";
import { __, _x } from "@wordpress/i18n";
import { memo, useContext, useMemo } from "@wordpress/element";
import { cog } from "@wordpress/icons";
import warning from "@wordpress/warning";
import { useInstanceId } from "@wordpress/compose";
import { Stack } from "@wordpress/ui";
import { SORTING_DIRECTIONS, sortIcons, sortLabels } from "../../constants.mjs";
import { VIEW_LAYOUTS } from "../dataviews-layouts/index.mjs";
import DataViewsContext from "../dataviews-context/index.mjs";
import InfiniteScrollToggle from "./infinite-scroll-toggle.mjs";
import { PropertiesSection } from "./properties-section.mjs";
import { unlock } from "../../lock-unlock.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var { Menu } = unlock(componentsPrivateApis);
var DATAVIEWS_CONFIG_POPOVER_PROPS = {
  className: "dataviews-config__popover",
  placement: "bottom-end",
  offset: 9
};
function ViewTypeMenu() {
  const { view, onChangeView, defaultLayouts } = useContext(DataViewsContext);
  const availableLayouts = Object.keys(defaultLayouts);
  if (availableLayouts.length <= 1) {
    return null;
  }
  const activeView = VIEW_LAYOUTS.find((v) => view.type === v.type);
  return /* @__PURE__ */ jsxs(Menu, { children: [
    /* @__PURE__ */ jsx(
      Menu.TriggerButton,
      {
        render: /* @__PURE__ */ jsx(
          Button,
          {
            size: "compact",
            icon: activeView?.icon,
            label: __("Layout")
          }
        )
      }
    ),
    /* @__PURE__ */ jsx(Menu.Popover, { children: availableLayouts.map((layout) => {
      const config = VIEW_LAYOUTS.find(
        (v) => v.type === layout
      );
      if (!config) {
        return null;
      }
      return /* @__PURE__ */ jsx(
        Menu.RadioItem,
        {
          value: layout,
          name: "view-actions-available-view",
          checked: layout === view.type,
          hideOnClick: true,
          onChange: (e) => {
            switch (e.target.value) {
              case "list":
              case "grid":
              case "table":
              case "pickerGrid":
              case "pickerTable":
              case "activity":
                const viewWithoutLayout = { ...view };
                if ("layout" in viewWithoutLayout) {
                  delete viewWithoutLayout.layout;
                }
                return onChangeView({
                  ...viewWithoutLayout,
                  type: e.target.value,
                  ...defaultLayouts[e.target.value]
                });
            }
            warning("Invalid dataview");
          },
          children: /* @__PURE__ */ jsx(Menu.ItemLabel, { children: config.label })
        },
        layout
      );
    }) })
  ] });
}
function SortFieldControl() {
  const { view, fields, onChangeView } = useContext(DataViewsContext);
  const orderOptions = useMemo(() => {
    const sortableFields = fields.filter(
      (field) => field.enableSorting !== false
    );
    return sortableFields.map((field) => {
      return {
        label: field.label,
        value: field.id
      };
    });
  }, [fields]);
  return /* @__PURE__ */ jsx(
    SelectControl,
    {
      __next40pxDefaultSize: true,
      label: __("Sort by"),
      value: view.sort?.field,
      options: orderOptions,
      onChange: (value) => {
        onChangeView({
          ...view,
          sort: {
            direction: view?.sort?.direction || "desc",
            field: value
          },
          showLevels: false
        });
      }
    }
  );
}
function SortDirectionControl() {
  const { view, fields, onChangeView } = useContext(DataViewsContext);
  const sortableFields = fields.filter(
    (field) => field.enableSorting !== false
  );
  if (sortableFields.length === 0) {
    return null;
  }
  let value = view.sort?.direction;
  if (!value && view.sort?.field) {
    value = "desc";
  }
  return /* @__PURE__ */ jsx(
    ToggleGroupControl,
    {
      className: "dataviews-view-config__sort-direction",
      __next40pxDefaultSize: true,
      isBlock: true,
      label: __("Order"),
      value,
      onChange: (newDirection) => {
        if (newDirection === "asc" || newDirection === "desc") {
          onChangeView({
            ...view,
            sort: {
              direction: newDirection,
              field: view.sort?.field || // If there is no field assigned as the sorting field assign the first sortable field.
              fields.find(
                (field) => field.enableSorting !== false
              )?.id || ""
            },
            showLevels: false
          });
          return;
        }
        warning("Invalid direction");
      },
      children: SORTING_DIRECTIONS.map((direction) => {
        return /* @__PURE__ */ jsx(
          ToggleGroupControlOptionIcon,
          {
            value: direction,
            icon: sortIcons[direction],
            label: sortLabels[direction]
          },
          direction
        );
      })
    }
  );
}
function ItemsPerPageControl() {
  const { view, config, onChangeView } = useContext(DataViewsContext);
  const { infiniteScrollEnabled } = view;
  if (!config || !config.perPageSizes || config.perPageSizes.length < 2 || config.perPageSizes.length > 6 || infiniteScrollEnabled) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    ToggleGroupControl,
    {
      __next40pxDefaultSize: true,
      isBlock: true,
      label: __("Items per page"),
      value: view.perPage || 10,
      disabled: !view?.sort?.field,
      onChange: (newItemsPerPage) => {
        const newItemsPerPageNumber = typeof newItemsPerPage === "number" || newItemsPerPage === void 0 ? newItemsPerPage : parseInt(newItemsPerPage, 10);
        onChangeView({
          ...view,
          perPage: newItemsPerPageNumber,
          page: 1
        });
      },
      children: config.perPageSizes.map((value) => {
        return /* @__PURE__ */ jsx(
          ToggleGroupControlOption,
          {
            value,
            label: value.toString()
          },
          value
        );
      })
    }
  );
}
function ResetViewButton() {
  const { onReset } = useContext(DataViewsContext);
  if (onReset === void 0) {
    return null;
  }
  const isDisabled = onReset === false;
  return /* @__PURE__ */ jsx(
    Button,
    {
      variant: "tertiary",
      size: "compact",
      disabled: isDisabled,
      accessibleWhenDisabled: true,
      className: "dataviews-view-config__reset-button",
      onClick: () => {
        if (typeof onReset === "function") {
          onReset();
        }
      },
      children: __("Reset view")
    }
  );
}
function DataviewsViewConfigDropdown() {
  const { view, onReset } = useContext(DataViewsContext);
  const popoverId = useInstanceId(
    _DataViewsViewConfig,
    "dataviews-view-config-dropdown"
  );
  const activeLayout = VIEW_LAYOUTS.find(
    (layout) => layout.type === view.type
  );
  const isModified = typeof onReset === "function";
  return /* @__PURE__ */ jsx(
    Dropdown,
    {
      expandOnMobile: true,
      popoverProps: {
        ...DATAVIEWS_CONFIG_POPOVER_PROPS,
        id: popoverId
      },
      renderToggle: ({ onToggle, isOpen }) => {
        return /* @__PURE__ */ jsxs("div", { className: "dataviews-view-config__toggle-wrapper", children: [
          /* @__PURE__ */ jsx(
            Button,
            {
              size: "compact",
              icon: cog,
              label: _x(
                "View options",
                "View is used as a noun"
              ),
              onClick: onToggle,
              "aria-expanded": isOpen ? "true" : "false",
              "aria-controls": popoverId
            }
          ),
          isModified && /* @__PURE__ */ jsx("span", { className: "dataviews-view-config__modified-indicator" })
        ] });
      },
      renderContent: () => /* @__PURE__ */ jsx(
        DropdownContentWrapper,
        {
          paddingSize: "medium",
          className: "dataviews-config__popover-content-wrapper",
          children: /* @__PURE__ */ jsxs(
            Stack,
            {
              direction: "column",
              className: "dataviews-view-config",
              gap: "xl",
              children: [
                /* @__PURE__ */ jsxs(
                  Stack,
                  {
                    direction: "row",
                    justify: "space-between",
                    align: "center",
                    className: "dataviews-view-config__header",
                    children: [
                      /* @__PURE__ */ jsx(
                        Heading,
                        {
                          level: 2,
                          className: "dataviews-settings-section__title",
                          children: __("Appearance")
                        }
                      ),
                      /* @__PURE__ */ jsx(ResetViewButton, {})
                    ]
                  }
                ),
                /* @__PURE__ */ jsxs(Stack, { direction: "column", gap: "lg", children: [
                  /* @__PURE__ */ jsxs(
                    Stack,
                    {
                      direction: "row",
                      gap: "sm",
                      className: "dataviews-view-config__sort-controls",
                      children: [
                        /* @__PURE__ */ jsx(SortFieldControl, {}),
                        /* @__PURE__ */ jsx(SortDirectionControl, {})
                      ]
                    }
                  ),
                  !!activeLayout?.viewConfigOptions && /* @__PURE__ */ jsx(activeLayout.viewConfigOptions, {}),
                  /* @__PURE__ */ jsx(InfiniteScrollToggle, {}),
                  /* @__PURE__ */ jsx(ItemsPerPageControl, {}),
                  /* @__PURE__ */ jsx(PropertiesSection, {})
                ] })
              ]
            }
          )
        }
      )
    }
  );
}
function _DataViewsViewConfig() {
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(ViewTypeMenu, {}),
    /* @__PURE__ */ jsx(DataviewsViewConfigDropdown, {})
  ] });
}
var DataViewsViewConfig = memo(_DataViewsViewConfig);
var dataviews_view_config_default = DataViewsViewConfig;
export {
  DataviewsViewConfigDropdown,
  ViewTypeMenu,
  dataviews_view_config_default as default
};
