// @wordpress/global-styles-engine@1.9.0/build/settings/get-palette.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/global-styles-engine/src/settings/get-palette.ts
var get_palette_exports = {};
__export(get_palette_exports, {
  default: () => getPalettes
});
module.exports = __toCommonJS(get_palette_exports);
var import_i18n = require("@wordpress/i18n");
var import_get_setting = require("./get-setting.cjs");
function getPalettes(settings) {
  if (!settings) {
    return {
      disableCustomColors: true,
      disableCustomGradients: true,
      colors: [],
      gradients: [],
      duotones: [],
      hasColorsOrGradients: false
    };
  }
  const [
    enableCustomColors,
    customColors,
    themeColors,
    defaultColors,
    shouldDisplayDefaultColors,
    enableCustomGradients,
    customGradients,
    themeGradients,
    defaultGradients,
    shouldDisplayDefaultGradients,
    shouldDisplayDefaultDuotones,
    customDuotones,
    themeDuotones,
    defaultDuotones
  ] = [
    "color.custom",
    "color.palette.custom",
    "color.palette.theme",
    "color.palette.default",
    "color.defaultPalette",
    "color.customGradient",
    "color.gradients.custom",
    "color.gradients.theme",
    "color.gradients.default",
    "color.defaultGradients",
    "color.defaultDuotone",
    "color.duotone.custom",
    "color.duotone.theme",
    "color.duotone.default"
  ].map((path) => (0, import_get_setting.getSetting)(settings, path));
  const palettes = {
    disableCustomColors: !enableCustomColors,
    disableCustomGradients: !enableCustomGradients,
    colors: [],
    gradients: [],
    duotones: [],
    hasColorsOrGradients: false
  };
  if (themeColors && themeColors.length) {
    palettes.colors?.push({
      name: (0, import_i18n._x)("Theme", "Indicates this palette comes from the theme."),
      slug: "theme",
      colors: themeColors
    });
  }
  if (shouldDisplayDefaultColors && defaultColors && defaultColors.length) {
    palettes.colors?.push({
      name: (0, import_i18n._x)(
        "Default",
        "Indicates this palette comes from WordPress."
      ),
      slug: "default",
      colors: defaultColors
    });
  }
  if (customColors && customColors.length) {
    palettes.colors?.push({
      name: (0, import_i18n._x)(
        "Custom",
        "Indicates this palette is created by the user."
      ),
      slug: "custom",
      colors: customColors
    });
  }
  if (themeGradients && themeGradients.length) {
    palettes.gradients?.push({
      name: (0, import_i18n._x)("Theme", "Indicates this palette comes from the theme."),
      slug: "theme",
      gradients: themeGradients
    });
  }
  if (shouldDisplayDefaultGradients && defaultGradients && defaultGradients.length) {
    palettes.gradients?.push({
      name: (0, import_i18n._x)(
        "Default",
        "Indicates this palette comes from WordPress."
      ),
      slug: "default",
      gradients: defaultGradients
    });
  }
  if (customGradients && customGradients.length) {
    palettes.gradients?.push({
      name: (0, import_i18n._x)(
        "Custom",
        "Indicates this palette is created by the user."
      ),
      slug: "custom",
      gradients: customGradients
    });
  }
  if (themeDuotones && themeDuotones.length) {
    palettes.duotones?.push({
      name: (0, import_i18n._x)(
        "Theme",
        "Indicates these duotone filters come from the theme."
      ),
      slug: "theme",
      duotones: themeDuotones
    });
  }
  if (shouldDisplayDefaultDuotones && defaultDuotones && defaultDuotones.length) {
    palettes.duotones?.push({
      name: (0, import_i18n._x)(
        "Default",
        "Indicates these duotone filters come from WordPress."
      ),
      slug: "default",
      duotones: defaultDuotones
    });
  }
  if (customDuotones && customDuotones.length) {
    palettes.duotones?.push({
      name: (0, import_i18n._x)(
        "Custom",
        "Indicates these doutone filters are created by the user."
      ),
      slug: "custom",
      duotones: customDuotones
    });
  }
  palettes.hasColorsOrGradients = !!palettes.colors?.length || !!palettes.gradients?.length;
  return palettes;
}
