// @wordpress/block-editor/build/components/block-variation-picker/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-variation-picker/index.js
var block_variation_picker_exports = {};
__export(block_variation_picker_exports, {
  default: () => block_variation_picker_default
});
module.exports = __toCommonJS(block_variation_picker_exports);
var import_clsx = __toESM(require("clsx"));
var import_i18n = require("@wordpress/i18n");
var import_components = require("@wordpress/components");
var import_icons = require("@wordpress/icons");
var import_jsx_runtime = require("react/jsx-runtime");
function BlockVariationPicker({
  icon = import_icons.layout,
  label = (0, import_i18n.__)("Choose variation"),
  instructions = (0, import_i18n.__)("Select a variation to start with:"),
  variations,
  onSelect,
  allowSkip
}) {
  const classes = (0, import_clsx.default)("block-editor-block-variation-picker", {
    "has-many-variations": variations.length > 4
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_components.Placeholder,
    {
      icon,
      label,
      instructions,
      className: classes,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "ul",
          {
            className: "block-editor-block-variation-picker__variations",
            role: "list",
            "aria-label": (0, import_i18n.__)("Block variations"),
            children: variations.map((variation) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("li", { children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_components.Button,
                {
                  __next40pxDefaultSize: true,
                  variant: "tertiary",
                  icon: variation.icon && variation.icon.src ? variation.icon.src : variation.icon,
                  iconSize: 48,
                  onClick: () => onSelect(variation),
                  className: "block-editor-block-variation-picker__variation",
                  label: variation.description || variation.title
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: "block-editor-block-variation-picker__variation-label", children: variation.title })
            ] }, variation.name))
          }
        ),
        allowSkip && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "block-editor-block-variation-picker__skip", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.Button,
          {
            __next40pxDefaultSize: true,
            variant: "link",
            onClick: () => onSelect(),
            children: (0, import_i18n.__)("Skip")
          }
        ) })
      ]
    }
  );
}
var block_variation_picker_default = BlockVariationPicker;
