// @wordpress/block-editor/build/components/global-styles/advanced-panel.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/global-styles/advanced-panel.js
var advanced_panel_exports = {};
__export(advanced_panel_exports, {
  default: () => AdvancedPanel,
  validateCSS: () => validateCSS
});
module.exports = __toCommonJS(advanced_panel_exports);
var import_components = require("@wordpress/components");
var import_element = require("@wordpress/element");
var import_i18n = require("@wordpress/i18n");
var import_transform_styles = __toESM(require("../../utils/transform-styles/index.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function validateCSS(css) {
  if (typeof css === "string" && /<\/?\w/.test(css)) {
    return false;
  }
  return true;
}
function AdvancedPanel({
  value,
  onChange,
  inheritedValue = value,
  help
}) {
  const [cssError, setCSSError] = (0, import_element.useState)(null);
  const customCSS = inheritedValue?.css;
  function handleOnChange(newValue) {
    onChange({
      ...value,
      css: newValue
    });
    if (!validateCSS(newValue)) {
      setCSSError(
        (0, import_i18n.__)("The custom CSS is invalid. Do not use <> markup.")
      );
      return;
    }
    if (cssError) {
      setCSSError(null);
    }
  }
  function handleOnBlur(event) {
    const cssValue = event?.target?.value;
    if (!cssValue || !validateCSS(cssValue)) {
      return;
    }
    const [transformed] = (0, import_transform_styles.default)(
      [{ css: cssValue }],
      ".for-validation-only"
    );
    setCSSError(
      transformed === null ? (0, import_i18n.__)("There is an error with your CSS structure.") : null
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalVStack, { spacing: 3, children: [
    cssError && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.Notice, { status: "error", onRemove: () => setCSSError(null), children: cssError }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.TextareaControl,
      {
        label: (0, import_i18n.__)("Additional CSS"),
        value: customCSS,
        onChange: (newValue) => handleOnChange(newValue),
        onBlur: handleOnBlur,
        className: "block-editor-global-styles-advanced-panel__custom-css-input",
        spellCheck: false,
        help
      }
    )
  ] });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  validateCSS
});
