// @wordpress/block-editor/build/components/global-styles/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/global-styles/index.js
var global_styles_exports = {};
__export(global_styles_exports, {
  AdvancedPanel: () => import_advanced_panel.default,
  BackgroundPanel: () => import_background_panel.default,
  BorderPanel: () => import_border_panel.default,
  ColorPanel: () => import_color_panel.default,
  DimensionsPanel: () => import_dimensions_panel.default,
  FiltersPanel: () => import_filters_panel.default,
  ImageSettingsPanel: () => import_image_settings_panel.default,
  StateControl: () => import_state_control.default,
  TypographyPanel: () => import_typography_panel.default,
  useHasBackgroundPanel: () => import_background_panel.useHasBackgroundPanel,
  useHasBorderPanel: () => import_border_panel.useHasBorderPanel,
  useHasBorderPanelControls: () => import_border_panel.useHasBorderPanelControls,
  useHasColorPanel: () => import_color_panel.useHasColorPanel,
  useHasDimensionsPanel: () => import_dimensions_panel.useHasDimensionsPanel,
  useHasFiltersPanel: () => import_filters_panel.useHasFiltersPanel,
  useHasImageSettingsPanel: () => import_image_settings_panel.useHasImageSettingsPanel,
  useHasTypographyPanel: () => import_typography_panel.useHasTypographyPanel,
  useSettingsForBlockElement: () => import_hooks.useSettingsForBlockElement
});
module.exports = __toCommonJS(global_styles_exports);
var import_hooks = require("./hooks.cjs");
var import_typography_panel = __toESM(require("./typography-panel.cjs"));
var import_dimensions_panel = __toESM(require("./dimensions-panel.cjs"));
var import_border_panel = __toESM(require("./border-panel.cjs"));
var import_color_panel = __toESM(require("./color-panel.cjs"));
var import_filters_panel = __toESM(require("./filters-panel.cjs"));
var import_image_settings_panel = __toESM(require("./image-settings-panel.cjs"));
var import_advanced_panel = __toESM(require("./advanced-panel.cjs"));
var import_background_panel = __toESM(require("./background-panel.cjs"));
var import_state_control = __toESM(require("./state-control.cjs"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AdvancedPanel,
  BackgroundPanel,
  BorderPanel,
  ColorPanel,
  DimensionsPanel,
  FiltersPanel,
  ImageSettingsPanel,
  StateControl,
  TypographyPanel,
  useHasBackgroundPanel,
  useHasBorderPanel,
  useHasBorderPanelControls,
  useHasColorPanel,
  useHasDimensionsPanel,
  useHasFiltersPanel,
  useHasImageSettingsPanel,
  useHasTypographyPanel,
  useSettingsForBlockElement
});
