// .ignored/@wordpress/core-data/build/utils/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/utils/index.js
var utils_exports = {};
__export(utils_exports, {
  ALLOWED_RESOURCE_ACTIONS: () => import_user_permissions.ALLOWED_RESOURCE_ACTIONS,
  RECEIVE_INTERMEDIATE_RESULTS: () => import_receive_intermediate_results.RECEIVE_INTERMEDIATE_RESULTS,
  conservativeMapItem: () => import_conservative_map_item.default,
  forwardResolver: () => import_forward_resolver.default,
  getNestedValue: () => import_get_nested_value.default,
  getNormalizedCommaSeparable: () => import_get_normalized_comma_separable.default,
  getUserPermissionCacheKey: () => import_user_permissions.getUserPermissionCacheKey,
  getUserPermissionsFromAllowHeader: () => import_user_permissions.getUserPermissionsFromAllowHeader,
  ifMatchingAction: () => import_if_matching_action.default,
  isNumericID: () => import_is_numeric_id.default,
  isRawAttribute: () => import_is_raw_attribute.default,
  normalizeQueryForResolution: () => import_normalize_query_for_resolution.default,
  onSubKey: () => import_on_sub_key.default,
  replaceAction: () => import_replace_action.default,
  setNestedValue: () => import_set_nested_value.default,
  withWeakMapCache: () => import_with_weak_map_cache.default
});
module.exports = __toCommonJS(utils_exports);
var import_conservative_map_item = __toESM(require("./conservative-map-item.cjs"));
var import_get_normalized_comma_separable = __toESM(require("./get-normalized-comma-separable.cjs"));
var import_if_matching_action = __toESM(require("./if-matching-action.cjs"));
var import_forward_resolver = __toESM(require("./forward-resolver.cjs"));
var import_on_sub_key = __toESM(require("./on-sub-key.cjs"));
var import_replace_action = __toESM(require("./replace-action.cjs"));
var import_with_weak_map_cache = __toESM(require("./with-weak-map-cache.cjs"));
var import_is_raw_attribute = __toESM(require("./is-raw-attribute.cjs"));
var import_set_nested_value = __toESM(require("./set-nested-value.cjs"));
var import_get_nested_value = __toESM(require("./get-nested-value.cjs"));
var import_is_numeric_id = __toESM(require("./is-numeric-id.cjs"));
var import_user_permissions = require("./user-permissions.cjs");
var import_receive_intermediate_results = require("./receive-intermediate-results.cjs");
var import_normalize_query_for_resolution = __toESM(require("./normalize-query-for-resolution.cjs"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ALLOWED_RESOURCE_ACTIONS,
  RECEIVE_INTERMEDIATE_RESULTS,
  conservativeMapItem,
  forwardResolver,
  getNestedValue,
  getNormalizedCommaSeparable,
  getUserPermissionCacheKey,
  getUserPermissionsFromAllowHeader,
  ifMatchingAction,
  isNumericID,
  isRawAttribute,
  normalizeQueryForResolution,
  onSubKey,
  replaceAction,
  setNestedValue,
  withWeakMapCache
});
