// @wordpress/block-editor/build/components/url-input/button.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/url-input/button.js
var button_exports = {};
__export(button_exports, {
  default: () => button_default
});
module.exports = __toCommonJS(button_exports);
var import_i18n = require("@wordpress/i18n");
var import_element = require("@wordpress/element");
var import_components = require("@wordpress/components");
var import_icons = require("@wordpress/icons");
var import__ = __toESM(require("./index.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function URLInputButton({ url, onChange }) {
  const [expanded, toggleExpanded] = (0, import_element.useReducer)(
    (isExpanded) => !isExpanded,
    false
  );
  const submitLink = (event) => {
    event.preventDefault();
    toggleExpanded();
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "block-editor-url-input__button", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.Button,
      {
        size: "compact",
        icon: import_icons.link,
        label: url ? (0, import_i18n.__)("Edit link") : (0, import_i18n.__)("Insert link"),
        onClick: toggleExpanded,
        className: "components-toolbar__control",
        isPressed: !!url
      }
    ),
    expanded && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "form",
      {
        className: "block-editor-url-input__button-modal",
        onSubmit: submitLink,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "block-editor-url-input__button-modal-line", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              __next40pxDefaultSize: true,
              className: "block-editor-url-input__back",
              icon: import_icons.arrowLeft,
              label: (0, import_i18n.__)("Close"),
              onClick: toggleExpanded
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import__.default,
            {
              value: url || "",
              onChange,
              suffix: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.__experimentalInputControlSuffixWrapper, { variant: "control", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_components.Button,
                {
                  size: "small",
                  icon: import_icons.keyboardReturn,
                  label: (0, import_i18n.__)("Submit"),
                  type: "submit"
                }
              ) })
            }
          )
        ] })
      }
    )
  ] });
}
var button_default = URLInputButton;
