// @wordpress/block-editor@15.15.0/build/components/block-list-appender/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-list-appender/index.js
var block_list_appender_exports = {};
__export(block_list_appender_exports, {
  default: () => BlockListAppender
});
module.exports = __toCommonJS(block_list_appender_exports);
var import_clsx = __toESM(require("clsx"));
var import_data = require("@wordpress/data");
var import_blocks = require("@wordpress/blocks");
var import_default_block_appender = __toESM(require("../default-block-appender/index.cjs"));
var import_button_block_appender = __toESM(require("../button-block-appender/index.cjs"));
var import_store = require("../../store/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function DefaultAppender({ rootClientId }) {
  const canInsertDefaultBlock = (0, import_data.useSelect)(
    (select) => select(import_store.store).canInsertBlockType(
      (0, import_blocks.getDefaultBlockName)(),
      rootClientId
    )
  );
  if (canInsertDefaultBlock) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_default_block_appender.default, { rootClientId });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_button_block_appender.default,
    {
      rootClientId,
      className: "block-list-appender__toggle"
    }
  );
}
function BlockListAppender({
  rootClientId,
  CustomAppender,
  className,
  tagName: TagName = "div"
}) {
  const isDragOver = (0, import_data.useSelect)(
    (select) => {
      const {
        getBlockInsertionPoint,
        isBlockInsertionPointVisible,
        getBlockCount
      } = select(import_store.store);
      const insertionPoint = getBlockInsertionPoint();
      return isBlockInsertionPointVisible() && rootClientId === insertionPoint?.rootClientId && getBlockCount(rootClientId) === 0;
    },
    [rootClientId]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    TagName,
    {
      tabIndex: -1,
      className: (0, import_clsx.default)("block-list-appender wp-block", className, {
        "is-drag-over": isDragOver
      }),
      contentEditable: false,
      "data-block": true,
      children: CustomAppender ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CustomAppender, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DefaultAppender, { rootClientId })
    }
  );
}
