// @wordpress/global-styles-ui/build-module/screen-header.mjs
// packages/global-styles-ui/src/screen-header.tsx
import {
  __experimentalHStack as HStack,
  __experimentalVStack as VStack,
  __experimentalSpacer as Spacer,
  __experimentalHeading as Heading,
  __experimentalView as View,
  __experimentalText as Text,
  Navigator
} from "@wordpress/components";
import { isRTL, __ } from "@wordpress/i18n";
import { chevronRight, chevronLeft } from "@wordpress/icons";
import { privateApis as blockEditorPrivateApis } from "@wordpress/block-editor";
import { unlock } from "./lock-unlock.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
var { StateControl } = unlock(blockEditorPrivateApis);
function ScreenHeader({
  title,
  description,
  onBack,
  states,
  selectedState = "default",
  onChangeState
}) {
  return /* @__PURE__ */ jsx(VStack, { spacing: 0, children: /* @__PURE__ */ jsx(View, { children: /* @__PURE__ */ jsx(Spacer, { marginBottom: 0, paddingX: 4, paddingY: 3, children: /* @__PURE__ */ jsxs(VStack, { spacing: 2, children: [
    /* @__PURE__ */ jsxs(HStack, { spacing: 2, children: [
      /* @__PURE__ */ jsx(
        Navigator.BackButton,
        {
          icon: isRTL() ? chevronRight : chevronLeft,
          size: "small",
          label: __("Back"),
          onClick: onBack
        }
      ),
      /* @__PURE__ */ jsx(Spacer, { children: /* @__PURE__ */ jsxs(
        HStack,
        {
          justify: "space-between",
          alignment: "center",
          children: [
            /* @__PURE__ */ jsx(
              Heading,
              {
                className: "global-styles-ui-header",
                level: 2,
                size: 13,
                children: title
              }
            ),
            /* @__PURE__ */ jsx(
              StateControl,
              {
                states,
                value: selectedState,
                onChange: onChangeState
              }
            )
          ]
        }
      ) })
    ] }),
    description && /* @__PURE__ */ jsx(Text, { className: "global-styles-ui-header__description", children: description })
  ] }) }) }) });
}
export {
  ScreenHeader
};
