// @wordpress/dataviews/build/utils/filter-sort-and-paginate.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/dataviews/src/utils/filter-sort-and-paginate.ts
var filter_sort_and_paginate_exports = {};
__export(filter_sort_and_paginate_exports, {
  default: () => filterSortAndPaginate
});
module.exports = __toCommonJS(filter_sort_and_paginate_exports);
var import_remove_accents = __toESM(require("remove-accents"));
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var import_constants = require("../constants.cjs");
var import_field_types = __toESM(require("../field-types/index.cjs"));
function normalizeSearchInput(input = "") {
  return (0, import_remove_accents.default)(input.trim().toLowerCase());
}
var EMPTY_ARRAY = [];
function filterSortAndPaginate(data, view, fields) {
  if (!data) {
    return {
      data: EMPTY_ARRAY,
      paginationInfo: { totalItems: 0, totalPages: 0 }
    };
  }
  const _fields = (0, import_field_types.default)(fields);
  let filteredData = [...data];
  if (view.search) {
    const normalizedSearch = normalizeSearchInput(view.search);
    filteredData = filteredData.filter((item) => {
      return _fields.filter((field) => field.enableGlobalSearch).some((field) => {
        const fieldValue = field.getValue({ item });
        const values = Array.isArray(fieldValue) ? fieldValue : [fieldValue];
        return values.some(
          (value) => normalizeSearchInput(String(value)).includes(
            normalizedSearch
          )
        );
      });
    });
  }
  if (view.filters && view.filters?.length > 0) {
    view.filters.forEach((filter) => {
      const field = _fields.find(
        (_field) => _field.id === filter.field
      );
      if (field) {
        if (filter.operator === import_constants.OPERATOR_IS_NOT_ALL) {
          (0, import_deprecated.default)("The 'isNotAll' filter operator", {
            since: "7.0",
            alternative: "'isNone'"
          });
        }
        const handler = field.filter[filter.operator];
        if (handler) {
          filteredData = filteredData.filter(
            (item) => handler(item, field, filter.value)
          );
        }
      }
    });
  }
  const sortByField = view.sort?.field ? _fields.find((field) => {
    return field.enableSorting !== false && field.id === view.sort?.field;
  }) : null;
  const groupByField = view.groupBy?.field ? _fields.find((field) => {
    return field.enableSorting !== false && field.id === view.groupBy?.field;
  }) : null;
  if (sortByField || groupByField) {
    filteredData.sort((a, b) => {
      if (groupByField) {
        const groupCompare = groupByField.sort(
          a,
          b,
          view.groupBy?.direction ?? "asc"
        );
        if (groupCompare !== 0) {
          return groupCompare;
        }
      }
      if (sortByField) {
        return sortByField.sort(a, b, view.sort?.direction ?? "desc");
      }
      return 0;
    });
  }
  let totalItems = filteredData.length;
  let totalPages = 1;
  if (view.page !== void 0 && view.perPage !== void 0) {
    const start = (view.page - 1) * view.perPage;
    totalItems = filteredData?.length || 0;
    totalPages = Math.ceil(totalItems / view.perPage);
    filteredData = filteredData?.slice(start, start + view.perPage);
  }
  return {
    data: filteredData,
    paginationInfo: {
      totalItems,
      totalPages
    }
  };
}
