// @wordpress/block-editor/build/components/block-alignment-matrix-control/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-alignment-matrix-control/index.js
var block_alignment_matrix_control_exports = {};
__export(block_alignment_matrix_control_exports, {
  default: () => block_alignment_matrix_control_default
});
module.exports = __toCommonJS(block_alignment_matrix_control_exports);
var import_i18n = require("@wordpress/i18n");
var import_keycodes = require("@wordpress/keycodes");
var import_components = require("@wordpress/components");
var import_jsx_runtime = require("react/jsx-runtime");
var noop = () => {
};
function BlockAlignmentMatrixControl(props) {
  const {
    label = (0, import_i18n.__)("Change matrix alignment"),
    onChange = noop,
    value = "center",
    isDisabled
  } = props;
  const icon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.AlignmentMatrixControl.Icon, { value });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Dropdown,
    {
      popoverProps: { placement: "bottom-start" },
      renderToggle: ({ onToggle, isOpen }) => {
        const openOnArrowDown = (event) => {
          if (!isOpen && event.keyCode === import_keycodes.DOWN) {
            event.preventDefault();
            onToggle();
          }
        };
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.ToolbarButton,
          {
            onClick: onToggle,
            "aria-haspopup": "true",
            "aria-expanded": isOpen,
            onKeyDown: openOnArrowDown,
            label,
            icon,
            showTooltip: true,
            disabled: isDisabled
          }
        );
      },
      renderContent: () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.AlignmentMatrixControl, { onChange, value })
    }
  );
}
var block_alignment_matrix_control_default = BlockAlignmentMatrixControl;
