// @wordpress/block-library@9.42.0/build/comments/edit/template.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/comments/edit/template.js
var template_exports = {};
__export(template_exports, {
  default: () => template_default
});
module.exports = __toCommonJS(template_exports);
var TEMPLATE = [
  ["core/comments-title"],
  [
    "core/comment-template",
    {},
    [
      [
        "core/columns",
        {},
        [
          [
            "core/column",
            { width: "40px" },
            [
              [
                "core/avatar",
                {
                  size: 40,
                  style: {
                    border: { radius: "20px" }
                  }
                }
              ]
            ]
          ],
          [
            "core/column",
            {},
            [
              [
                "core/comment-author-name",
                {
                  fontSize: "small"
                }
              ],
              [
                "core/group",
                {
                  layout: { type: "flex" },
                  style: {
                    spacing: {
                      margin: {
                        top: "0px",
                        bottom: "0px"
                      }
                    }
                  }
                },
                [
                  [
                    "core/comment-date",
                    {
                      fontSize: "small"
                    }
                  ],
                  [
                    "core/comment-edit-link",
                    {
                      fontSize: "small"
                    }
                  ]
                ]
              ],
              ["core/comment-content"],
              [
                "core/comment-reply-link",
                {
                  fontSize: "small"
                }
              ]
            ]
          ]
        ]
      ]
    ]
  ],
  ["core/comments-pagination"],
  ["core/post-comments-form"]
];
var template_default = TEMPLATE;
