// @wordpress/block-library/build-module/tabs-menu/edit.mjs
// packages/block-library/src/tabs-menu/edit.js
import clsx from "clsx";
import {
  useBlockProps,
  useInnerBlocksProps,
  store as blockEditorStore
} from "@wordpress/block-editor";
import { useSelect } from "@wordpress/data";
import AddTabToolbarControl from "../tab/add-tab-toolbar-control.mjs";
import RemoveTabToolbarControl from "../tab/remove-tab-toolbar-control.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function Edit({ clientId, __unstableLayoutClassNames: layoutClassNames }) {
  const { tabsClientId } = useSelect(
    (select) => ({
      tabsClientId: select(blockEditorStore).getBlockRootClientId(clientId)
    }),
    [clientId]
  );
  const blockProps = useBlockProps({
    className: clsx(layoutClassNames),
    role: "tablist"
  });
  const innerBlocksProps = useInnerBlocksProps(blockProps, {
    allowedBlocks: ["core/tabs-menu-item"],
    orientation: "horizontal",
    renderAppender: false
  });
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(AddTabToolbarControl, { tabsClientId }),
    /* @__PURE__ */ jsx(RemoveTabToolbarControl, { tabsClientId }),
    /* @__PURE__ */ jsx("div", { ...innerBlocksProps })
  ] });
}
var edit_default = Edit;
export {
  edit_default as default
};
