// @wordpress/fields/build-module/fields/template/template-edit.mjs
// packages/fields/src/fields/template/template-edit.tsx
import { useCallback, useMemo } from "@wordpress/element";
import { store as coreStore } from "@wordpress/core-data";
import { SelectControl } from "@wordpress/components";
import { useSelect } from "@wordpress/data";
import { __ } from "@wordpress/i18n";
import { getItemTitle } from "../../actions/utils.mjs";
import { useDefaultTemplateLabel, useTemplateFieldMode } from "./hooks.mjs";
import { unlock } from "../../lock-unlock.mjs";
import { jsx } from "react/jsx-runtime";
var EMPTY_ARRAY = [];
function ClassicTemplateEdit({
  data,
  field,
  onChange
}) {
  const postId = typeof data.id === "number" ? data.id : parseInt(data.id, 10);
  const value = field.getValue({ item: data });
  const options = useMemo(
    () => Object.entries(
      data?.available_templates ?? {}
    ).map(([templateSlug, title]) => ({
      label: title,
      value: templateSlug
    })),
    [data]
  );
  const canSwitchTemplate = useSelect(
    (select) => {
      const { getHomePage, getPostsPageId } = unlock(
        select(coreStore)
      );
      const singlePostId = String(postId);
      const isPostsPage = getPostsPageId() === singlePostId;
      const isFrontPage = data.type === "page" && getHomePage()?.postId === singlePostId;
      return !isPostsPage && !isFrontPage;
    },
    [postId, data.type]
  );
  return /* @__PURE__ */ jsx(
    SelectControl,
    {
      __next40pxDefaultSize: true,
      label: __("Template"),
      hideLabelFromVision: true,
      value,
      options,
      onChange,
      disabled: !canSwitchTemplate
    }
  );
}
function BlockThemeTemplateEdit({
  data,
  field,
  onChange
}) {
  const postType = data.type;
  const postId = typeof data.id === "number" ? data.id : parseInt(data.id, 10);
  const slug = data.slug;
  const { templates, canSwitchTemplate } = useSelect(
    (select) => {
      const allTemplates = select(coreStore).getEntityRecords(
        "postType",
        "wp_template",
        {
          per_page: -1,
          post_type: postType
        }
      ) ?? EMPTY_ARRAY;
      const { getHomePage, getPostsPageId } = unlock(
        select(coreStore)
      );
      const singlePostId = String(postId);
      const isPostsPage = getPostsPageId() === singlePostId;
      const isFrontPage = postType === "page" && getHomePage()?.postId === singlePostId;
      return {
        templates: allTemplates,
        canSwitchTemplate: !isPostsPage && !isFrontPage
      };
    },
    [postId, postType]
  );
  const defaultTemplateLabel = useDefaultTemplateLabel(
    postType,
    postId,
    slug
  );
  const value = field.getValue({ item: data });
  const options = useMemo(() => {
    const templateOptions = templates.map((template) => ({
      label: getItemTitle(template),
      value: template.slug
    }));
    return [
      { label: defaultTemplateLabel, value: "" },
      ...templateOptions
    ];
  }, [templates, defaultTemplateLabel]);
  return /* @__PURE__ */ jsx(
    SelectControl,
    {
      __next40pxDefaultSize: true,
      label: __("Template"),
      hideLabelFromVision: true,
      value,
      options,
      onChange,
      disabled: !canSwitchTemplate
    }
  );
}
var TemplateEdit = ({
  data,
  field,
  onChange
}) => {
  const onChangeControl = useCallback(
    (newValue) => onChange({
      [field.id]: newValue
    }),
    [field.id, onChange]
  );
  const mode = useTemplateFieldMode(data);
  if (!mode || !["block-theme", "classic"].includes(mode)) {
    return null;
  }
  const Edit = mode === "classic" ? ClassicTemplateEdit : BlockThemeTemplateEdit;
  return /* @__PURE__ */ jsx(Edit, { data, field, onChange: onChangeControl });
};
export {
  TemplateEdit
};
