// @wordpress/global-styles-engine/build-module/utils/get-global-styles-changes.mjs
// packages/global-styles-engine/src/utils/get-global-styles-changes.ts
import memoize from "memize";
import { __, _n, sprintf } from "@wordpress/i18n";
import { getBlockTypes } from "@wordpress/blocks";
var globalStylesChangesCache = /* @__PURE__ */ new Map();
var EMPTY_ARRAY = [];
var translationMap = {
  caption: __("Caption"),
  link: __("Link"),
  button: __("Button"),
  heading: __("Heading"),
  h1: __("H1"),
  h2: __("H2"),
  h3: __("H3"),
  h4: __("H4"),
  h5: __("H5"),
  h6: __("H6"),
  "settings.color": __("Color"),
  "settings.typography": __("Typography"),
  "settings.shadow": __("Shadow"),
  "settings.layout": __("Layout"),
  "styles.color": __("Colors"),
  "styles.spacing": __("Spacing"),
  "styles.background": __("Background"),
  "styles.typography": __("Typography")
};
var getBlockNames = memoize(
  () => getBlockTypes().reduce(
    (accumulator, {
      name,
      title
    }) => {
      accumulator[name] = title;
      return accumulator;
    },
    {}
  )
);
var isObject = (obj) => obj !== null && typeof obj === "object";
function getTranslation(key) {
  if (translationMap[key]) {
    return translationMap[key];
  }
  const keyArray = key.split(".");
  if (keyArray?.[0] === "blocks") {
    const blockName = getBlockNames()?.[keyArray[1]];
    return blockName || keyArray[1];
  }
  if (keyArray?.[0] === "elements") {
    return translationMap[keyArray[1]] || keyArray[1];
  }
  return void 0;
}
function deepCompare(changedObject, originalObject, parentPath = "") {
  if (!isObject(changedObject) && !isObject(originalObject)) {
    return changedObject !== originalObject ? parentPath.split(".").slice(0, 2).join(".") : void 0;
  }
  changedObject = isObject(changedObject) ? changedObject : {};
  originalObject = isObject(originalObject) ? originalObject : {};
  const allKeys = /* @__PURE__ */ new Set([
    ...Object.keys(changedObject),
    ...Object.keys(originalObject)
  ]);
  let diffs = [];
  for (const key of allKeys) {
    const path = parentPath ? parentPath + "." + key : key;
    const changedPath = deepCompare(
      changedObject[key],
      originalObject[key],
      path
    );
    if (changedPath) {
      diffs = diffs.concat(changedPath);
    }
  }
  return diffs;
}
function getGlobalStylesChangelist(next, previous) {
  const cacheKey = JSON.stringify({ next, previous });
  if (globalStylesChangesCache.has(cacheKey)) {
    return globalStylesChangesCache.get(cacheKey);
  }
  const changedValueTree = deepCompare(
    {
      styles: {
        background: next?.styles?.background,
        color: next?.styles?.color,
        typography: next?.styles?.typography,
        spacing: next?.styles?.spacing
      },
      blocks: next?.styles?.blocks,
      elements: next?.styles?.elements,
      settings: next?.settings
    },
    {
      styles: {
        background: previous?.styles?.background,
        color: previous?.styles?.color,
        typography: previous?.styles?.typography,
        spacing: previous?.styles?.spacing
      },
      blocks: previous?.styles?.blocks,
      elements: previous?.styles?.elements,
      settings: previous?.settings
    }
  );
  if (!changedValueTree || Array.isArray(changedValueTree) && !changedValueTree.length) {
    globalStylesChangesCache.set(cacheKey, []);
    return [];
  }
  const changedValueArray = Array.isArray(changedValueTree) ? changedValueTree : [changedValueTree];
  const result = [...new Set(changedValueArray)].reduce((acc, curr) => {
    const translation = getTranslation(curr);
    if (translation) {
      acc.push([curr.split(".")[0], translation]);
    }
    return acc;
  }, []);
  globalStylesChangesCache.set(cacheKey, result);
  return result;
}
function getGlobalStylesChanges(next, previous, options = {}) {
  let changeList = getGlobalStylesChangelist(next, previous);
  const changesLength = changeList.length;
  const { maxResults } = options;
  if (changesLength) {
    if (!!maxResults && changesLength > maxResults) {
      changeList = changeList.slice(0, maxResults);
    }
    return Object.entries(
      changeList.reduce((acc, curr) => {
        const group = acc[curr[0]] || [];
        if (!group.includes(curr[1])) {
          acc[curr[0]] = [...group, curr[1]];
        }
        return acc;
      }, {})
    ).map(([key, changeValues]) => {
      const changeValuesLength = changeValues.length;
      const joinedChangesValue = changeValues.join(
        /* translators: Used between list items, there is a space after the comma. */
        __(", ")
        // eslint-disable-line @wordpress/i18n-no-flanking-whitespace
      );
      switch (key) {
        case "blocks": {
          return sprintf(
            // translators: %s: a list of block names separated by a comma.
            _n("%s block.", "%s blocks.", changeValuesLength),
            joinedChangesValue
          );
        }
        case "elements": {
          return sprintf(
            // translators: %s: a list of element names separated by a comma.
            _n("%s element.", "%s elements.", changeValuesLength),
            joinedChangesValue
          );
        }
        case "settings": {
          return sprintf(
            // translators: %s: a list of theme.json setting labels separated by a comma.
            __("%s settings."),
            joinedChangesValue
          );
        }
        case "styles": {
          return sprintf(
            // translators: %s: a list of theme.json top-level styles labels separated by a comma.
            __("%s styles."),
            joinedChangesValue
          );
        }
        default: {
          return sprintf(
            // translators: %s: a list of global styles changes separated by a comma.
            __("%s."),
            joinedChangesValue
          );
        }
      }
    });
  }
  return EMPTY_ARRAY;
}
export {
  getGlobalStylesChanges as default,
  getGlobalStylesChangelist
};
