// @wordpress/core-data@7.42.0/build/locks/engine.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/locks/engine.js
var engine_exports = {};
__export(engine_exports, {
  default: () => createLocks
});
module.exports = __toCommonJS(engine_exports);
var import_reducer = __toESM(require("./reducer.cjs"));
var import_selectors = require("./selectors.cjs");
function createLocks() {
  let state = (0, import_reducer.default)(void 0, { type: "@@INIT" });
  function processPendingLockRequests() {
    for (const request of (0, import_selectors.getPendingLockRequests)(state)) {
      const { store, path, exclusive, notifyAcquired } = request;
      if ((0, import_selectors.isLockAvailable)(state, store, path, { exclusive })) {
        const lock = { store, path, exclusive };
        state = (0, import_reducer.default)(state, {
          type: "GRANT_LOCK_REQUEST",
          lock,
          request
        });
        notifyAcquired(lock);
      }
    }
  }
  function acquire(store, path, exclusive) {
    return new Promise((resolve) => {
      state = (0, import_reducer.default)(state, {
        type: "ENQUEUE_LOCK_REQUEST",
        request: { store, path, exclusive, notifyAcquired: resolve }
      });
      processPendingLockRequests();
    });
  }
  function release(lock) {
    state = (0, import_reducer.default)(state, {
      type: "RELEASE_LOCK",
      lock
    });
    processPendingLockRequests();
  }
  return { acquire, release };
}
