// @wordpress/block-library@9.42.0/build-module/tabs-menu-item/edit.mjs
// packages/block-library/src/tabs-menu-item/edit.js
import clsx from "clsx";
import { __, sprintf } from "@wordpress/i18n";
import {
  useBlockProps,
  store as blockEditorStore,
  RichText
} from "@wordpress/block-editor";
import { useSelect, useDispatch } from "@wordpress/data";
import { useMemo, useCallback } from "@wordpress/element";
import Controls from "./controls.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var EMPTY_ARRAY = [];
function Edit({
  attributes,
  context,
  clientId,
  __unstableLayoutClassNames: layoutClassNames
}) {
  const tabsList = context["core/tabs-list"] || EMPTY_ARRAY;
  const activeTabIndex = context["core/tabs-activeTabIndex"] ?? 0;
  const editorActiveTabIndex = context["core/tabs-editorActiveTabIndex"];
  const effectiveActiveIndex = useMemo(() => {
    return editorActiveTabIndex ?? activeTabIndex;
  }, [editorActiveTabIndex, activeTabIndex]);
  const { menuItemIndex, tabsClientId, selectedTabClientId } = useSelect(
    (select) => {
      const {
        getBlockOrder,
        getBlockRootClientId,
        getSelectedBlockClientIds,
        hasSelectedInnerBlock
      } = select(blockEditorStore);
      const _tabsMenuClientId = getBlockRootClientId(clientId);
      const _tabsClientId = _tabsMenuClientId ? getBlockRootClientId(_tabsMenuClientId) : null;
      const siblings = getBlockOrder(_tabsMenuClientId);
      const _menuItemIndex = siblings.indexOf(clientId);
      const selectedIds = getSelectedBlockClientIds();
      let _selectedTabClientId = null;
      for (const tab2 of tabsList) {
        if (selectedIds.includes(tab2.clientId) || hasSelectedInnerBlock(tab2.clientId, true)) {
          _selectedTabClientId = tab2.clientId;
          break;
        }
      }
      return {
        menuItemIndex: _menuItemIndex,
        tabsClientId: _tabsClientId,
        selectedTabClientId: _selectedTabClientId
      };
    },
    [clientId, tabsList]
  );
  const tabAnchor = attributes.anchor?.replace(/-button$/, "") ?? "";
  const tab = tabAnchor && tabsList.find((t) => t.id === tabAnchor) || tabsList[menuItemIndex] || {};
  const tabListIndex = tab.index ?? menuItemIndex;
  const tabId = tab.id || `tab-${menuItemIndex}`;
  const tabClientId = tab.clientId || "";
  const label = tab.label || "";
  const isActive = tabListIndex === effectiveActiveIndex;
  const isSelected = tabClientId === selectedTabClientId;
  const { __unstableMarkNextChangeAsNotPersistent, updateBlockAttributes } = useDispatch(blockEditorStore);
  const handleTabClick = useCallback(
    (event) => {
      event.preventDefault();
      if (tabsClientId && tabListIndex !== effectiveActiveIndex) {
        __unstableMarkNextChangeAsNotPersistent();
        updateBlockAttributes(tabsClientId, {
          editorActiveTabIndex: tabListIndex
        });
      }
    },
    [
      tabsClientId,
      tabListIndex,
      effectiveActiveIndex,
      updateBlockAttributes,
      __unstableMarkNextChangeAsNotPersistent
    ]
  );
  const handleLabelChange = useCallback(
    (newLabel) => {
      if (tabClientId) {
        updateBlockAttributes(tabClientId, { label: newLabel });
      }
    },
    [tabClientId, updateBlockAttributes]
  );
  const blockProps = useBlockProps({
    className: clsx(layoutClassNames, {
      "is-active": isActive,
      "is-selected": isSelected
    }),
    "aria-controls": tabId,
    "aria-selected": isActive,
    id: `${tabId}--tab`,
    role: "tab",
    tabIndex: -1,
    onClick: handleTabClick
  });
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Controls, { tabsClientId }),
    /* @__PURE__ */ jsx("button", { ...blockProps, type: "button", children: /* @__PURE__ */ jsx(
      RichText,
      {
        tagName: "span",
        withoutInteractiveFormatting: true,
        placeholder: sprintf(
          /* translators: %d is the tab index + 1 */
          __("Tab title %d"),
          menuItemIndex + 1
        ),
        value: label,
        onChange: handleLabelChange
      }
    ) })
  ] });
}
var edit_default = Edit;
export {
  edit_default as default
};
