// @wordpress/block-library@9.42.0/build-module/tab/add-tab-toolbar-control.mjs
// packages/block-library/src/tab/add-tab-toolbar-control.js
import { sprintf, __ } from "@wordpress/i18n";
import { createBlock } from "@wordpress/blocks";
import {
  BlockControls,
  store as blockEditorStore
} from "@wordpress/block-editor";
import { ToolbarGroup, ToolbarButton } from "@wordpress/components";
import { useDispatch, useSelect } from "@wordpress/data";
import { jsx } from "react/jsx-runtime";
function AddTabToolbarControl({ tabsClientId }) {
  const { insertBlock } = useDispatch(blockEditorStore);
  const { tabPanelClientId, tabsMenuClientId, tabCount, existingAnchors } = useSelect(
    (select) => {
      if (!tabsClientId) {
        return {
          tabPanelClientId: null,
          tabsMenuClientId: null,
          existingAnchors: []
        };
      }
      const { getBlocks } = select(blockEditorStore);
      const innerBlocks = getBlocks(tabsClientId);
      const tabPanel = innerBlocks.find(
        (block) => block.name === "core/tab-panel"
      );
      const tabsMenu = innerBlocks.find(
        (block) => block.name === "core/tabs-menu"
      );
      return {
        tabPanelClientId: tabPanel?.clientId || null,
        tabsMenuClientId: tabsMenu?.clientId || null,
        tabCount: tabPanel?.innerBlocks?.length || 0,
        existingAnchors: (tabPanel?.innerBlocks || []).map((block) => block.attributes.anchor).filter(Boolean)
      };
    },
    [tabsClientId]
  );
  const addTab = () => {
    if (!tabPanelClientId) {
      return;
    }
    const existingAnchorSet = new Set(existingAnchors);
    let tabNumber = tabCount + 1;
    while (existingAnchorSet.has(`tab-${tabNumber}`)) {
      tabNumber++;
    }
    const newTabBlock = createBlock("core/tab", {
      anchor: `tab-${tabNumber}`,
      /* translators: %d: tab number */
      label: sprintf(__("Tab %d"), tabNumber)
    });
    insertBlock(newTabBlock, void 0, tabPanelClientId);
    if (tabsMenuClientId) {
      const newMenuItemBlock = createBlock("core/tabs-menu-item", {
        anchor: `tab-${tabNumber}-button`
      });
      insertBlock(newMenuItemBlock, void 0, tabsMenuClientId);
    }
  };
  return /* @__PURE__ */ jsx(BlockControls, { group: "other", children: /* @__PURE__ */ jsx(ToolbarGroup, { children: /* @__PURE__ */ jsx(
    ToolbarButton,
    {
      className: "components-toolbar__control",
      onClick: addTab,
      text: __("Add tab")
    }
  ) }) });
}
export {
  AddTabToolbarControl as default
};
