// @wordpress/block-library@9.42.0/build-module/tab/remove-tab-toolbar-control.mjs
// packages/block-library/src/tab/remove-tab-toolbar-control.js
import {
  BlockControls,
  store as blockEditorStore
} from "@wordpress/block-editor";
import { ToolbarGroup, ToolbarButton } from "@wordpress/components";
import { __ } from "@wordpress/i18n";
import { useDispatch, useSelect } from "@wordpress/data";
import { jsx } from "react/jsx-runtime";
function RemoveTabToolbarControl({ tabsClientId }) {
  const {
    removeBlock,
    updateBlockAttributes,
    selectBlock,
    __unstableMarkNextChangeAsNotPersistent
  } = useDispatch(blockEditorStore);
  const {
    activeTabClientId,
    activeMenuItemClientId,
    tabCount,
    editorActiveTabIndex
  } = useSelect(
    (select) => {
      if (!tabsClientId) {
        return {
          activeTabClientId: null,
          activeMenuItemClientId: null,
          tabCount: 0,
          editorActiveTabIndex: 0
        };
      }
      const { getBlocks, getBlockAttributes } = select(blockEditorStore);
      const tabsAttributes = getBlockAttributes(tabsClientId);
      const activeIndex = tabsAttributes?.editorActiveTabIndex ?? tabsAttributes?.activeTabIndex ?? 0;
      const innerBlocks = getBlocks(tabsClientId);
      const tabPanel = innerBlocks.find(
        (block) => block.name === "core/tab-panel"
      );
      const tabsMenu = innerBlocks.find(
        (block) => block.name === "core/tabs-menu"
      );
      const tabs = tabPanel?.innerBlocks || [];
      const menuItems = tabsMenu?.innerBlocks || [];
      const activeTab = tabs[activeIndex];
      const expectedMenuAnchor = activeTab?.attributes?.anchor ? `${activeTab.attributes.anchor}-button` : null;
      const activeMenuItem = expectedMenuAnchor ? menuItems.find(
        (m) => m.attributes?.anchor === expectedMenuAnchor
      ) : menuItems[activeIndex];
      return {
        activeTabClientId: activeTab?.clientId || null,
        activeMenuItemClientId: activeMenuItem?.clientId || null,
        tabCount: tabs.length,
        editorActiveTabIndex: activeIndex
      };
    },
    [tabsClientId]
  );
  const removeTab = () => {
    if (!activeTabClientId || tabCount <= 1) {
      return;
    }
    const newActiveIndex = editorActiveTabIndex >= tabCount - 1 ? tabCount - 2 : editorActiveTabIndex;
    __unstableMarkNextChangeAsNotPersistent();
    updateBlockAttributes(tabsClientId, {
      editorActiveTabIndex: newActiveIndex
    });
    removeBlock(activeTabClientId, false);
    if (activeMenuItemClientId) {
      removeBlock(activeMenuItemClientId, false);
    }
    if (tabsClientId) {
      selectBlock(tabsClientId);
    }
  };
  const isDisabled = tabCount <= 1 || !activeTabClientId;
  return /* @__PURE__ */ jsx(BlockControls, { group: "other", children: /* @__PURE__ */ jsx(ToolbarGroup, { children: /* @__PURE__ */ jsx(
    ToolbarButton,
    {
      className: "components-toolbar__control",
      onClick: removeTab,
      text: __("Remove tab"),
      disabled: isDisabled
    }
  ) }) });
}
export {
  RemoveTabToolbarControl as default
};
