// @wordpress/block-library@9.42.0/build-module/post-date/deprecated.mjs
// packages/block-library/src/post-date/deprecated.js
import clsx from "clsx";
import migrateFontFamily from "../utils/migrate-font-family.mjs";
import migrateTextAlign from "../utils/migrate-text-align.mjs";
var v4 = {
  attributes: {
    datetime: {
      type: "string",
      role: "content"
    },
    textAlign: {
      type: "string"
    },
    format: {
      type: "string"
    },
    isLink: {
      type: "boolean",
      default: false,
      role: "content"
    }
  },
  supports: {
    anchor: true,
    html: false,
    color: {
      gradients: true,
      link: true,
      __experimentalDefaultControls: {
        background: true,
        text: true,
        link: true
      }
    },
    spacing: {
      margin: true,
      padding: true
    },
    typography: {
      fontSize: true,
      lineHeight: true,
      __experimentalFontFamily: true,
      __experimentalFontWeight: true,
      __experimentalFontStyle: true,
      __experimentalTextTransform: true,
      __experimentalTextDecoration: true,
      __experimentalLetterSpacing: true,
      __experimentalDefaultControls: {
        fontSize: true
      }
    },
    interactivity: {
      clientNavigation: true
    },
    __experimentalBorder: {
      radius: true,
      color: true,
      width: true,
      style: true,
      __experimentalDefaultControls: {
        radius: true,
        color: true,
        width: true,
        style: true
      }
    }
  },
  save() {
    return null;
  },
  migrate: migrateTextAlign,
  isEligible(attributes) {
    return !!attributes.textAlign || !!attributes.className?.match(
      /\bhas-text-align-(left|center|right)\b/
    );
  }
};
var v3 = {
  attributes: {
    datetime: {
      type: "string",
      role: "content"
    },
    textAlign: {
      type: "string"
    },
    format: {
      type: "string"
    },
    isLink: {
      type: "boolean",
      default: false,
      role: "content"
    }
  },
  supports: {
    html: false,
    color: {
      gradients: true,
      link: true,
      __experimentalDefaultControls: {
        background: true,
        text: true,
        link: true
      }
    },
    spacing: {
      margin: true,
      padding: true
    },
    typography: {
      fontSize: true,
      lineHeight: true,
      __experimentalFontFamily: true,
      __experimentalFontWeight: true,
      __experimentalFontStyle: true,
      __experimentalTextTransform: true,
      __experimentalTextDecoration: true,
      __experimentalLetterSpacing: true,
      __experimentalDefaultControls: {
        fontSize: true
      }
    },
    interactivity: {
      clientNavigation: true
    },
    __experimentalBorder: {
      radius: true,
      color: true,
      width: true,
      style: true,
      __experimentalDefaultControls: {
        radius: true,
        color: true,
        width: true,
        style: true
      }
    }
  },
  save() {
    return null;
  },
  migrate({
    metadata: {
      bindings: {
        datetime: {
          source,
          args: { key, ...otherArgs }
        },
        ...otherBindings
      },
      ...otherMetadata
    },
    ...otherAttributes
  }) {
    return migrateTextAlign({
      metadata: {
        bindings: {
          datetime: {
            source,
            args: { field: key, ...otherArgs }
          },
          ...otherBindings
        },
        ...otherMetadata
      },
      ...otherAttributes
    });
  },
  isEligible(attributes) {
    return attributes?.metadata?.bindings?.datetime?.source === "core/post-data" && !!attributes?.metadata?.bindings?.datetime?.args?.key;
  }
};
var v2 = {
  attributes: {
    textAlign: {
      type: "string"
    },
    format: {
      type: "string"
    },
    isLink: {
      type: "boolean",
      default: false,
      role: "content"
    },
    displayType: {
      type: "string",
      default: "date"
    }
  },
  supports: {
    html: false,
    color: {
      gradients: true,
      link: true,
      __experimentalDefaultControls: {
        background: true,
        text: true,
        link: true
      }
    },
    spacing: {
      margin: true,
      padding: true
    },
    typography: {
      fontSize: true,
      lineHeight: true,
      __experimentalFontFamily: true,
      __experimentalFontWeight: true,
      __experimentalFontStyle: true,
      __experimentalTextTransform: true,
      __experimentalTextDecoration: true,
      __experimentalLetterSpacing: true,
      __experimentalDefaultControls: {
        fontSize: true
      }
    },
    interactivity: {
      clientNavigation: true
    },
    __experimentalBorder: {
      radius: true,
      color: true,
      width: true,
      style: true,
      __experimentalDefaultControls: {
        radius: true,
        color: true,
        width: true,
        style: true
      }
    }
  },
  save() {
    return null;
  },
  migrate({ className, displayType, metadata, ...otherAttributes }) {
    if (displayType === "date" || displayType === "modified") {
      if (displayType === "modified") {
        className = clsx(
          className,
          "wp-block-post-date__modified-date"
        );
      }
      return migrateTextAlign({
        ...otherAttributes,
        className,
        metadata: {
          ...metadata,
          bindings: {
            datetime: {
              source: "core/post-data",
              args: { field: displayType }
            }
          }
        }
      });
    }
  },
  isEligible(attributes) {
    return !attributes.datetime && !attributes?.metadata?.bindings?.datetime;
  }
};
var v1 = {
  attributes: {
    textAlign: {
      type: "string"
    },
    format: {
      type: "string"
    },
    isLink: {
      type: "boolean",
      default: false
    }
  },
  supports: {
    html: false,
    color: {
      gradients: true,
      link: true
    },
    typography: {
      fontSize: true,
      lineHeight: true,
      __experimentalFontFamily: true,
      __experimentalFontWeight: true,
      __experimentalFontStyle: true,
      __experimentalTextTransform: true,
      __experimentalLetterSpacing: true
    }
  },
  save() {
    return null;
  },
  migrate(attributes) {
    return migrateTextAlign(migrateFontFamily(attributes));
  },
  isEligible({ style }) {
    return style?.typography?.fontFamily;
  }
};
var deprecated_default = [v4, v3, v2, v1];
export {
  deprecated_default as default
};
