// @wordpress/blocks/build/api/raw-handling/markdown-converter.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/blocks/src/api/raw-handling/markdown-converter.js
var markdown_converter_exports = {};
__export(markdown_converter_exports, {
  default: () => markdownConverter
});
module.exports = __toCommonJS(markdown_converter_exports);
var import_showdown = __toESM(require("showdown"));
var converter = new import_showdown.default.Converter({
  noHeaderId: true,
  tables: true,
  literalMidWordUnderscores: true,
  omitExtraWLInCodeBlocks: true,
  simpleLineBreaks: true,
  strikethrough: true
});
function slackMarkdownVariantCorrector(text) {
  return text.replace(
    /((?:^|\n)```)([^\n`]+)(```(?:$|\n))/,
    (match, p1, p2, p3) => `${p1}
${p2}
${p3}`
  );
}
function bulletsToAsterisks(text) {
  return text.replace(/(^|\n)•( +)/g, "$1*$2");
}
function markdownConverter(text) {
  return converter.makeHtml(
    slackMarkdownVariantCorrector(bulletsToAsterisks(text))
  );
}
