// @wordpress/block-editor/build/components/text-transform-control/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/text-transform-control/index.js
var text_transform_control_exports = {};
__export(text_transform_control_exports, {
  default: () => TextTransformControl
});
module.exports = __toCommonJS(text_transform_control_exports);
var import_clsx = __toESM(require("clsx"));
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var import_components = require("@wordpress/components");
var import_jsx_runtime = require("react/jsx-runtime");
var TEXT_TRANSFORMS = [
  {
    label: (0, import_i18n.__)("None"),
    value: "none",
    icon: import_icons.reset
  },
  {
    label: (0, import_i18n.__)("Uppercase"),
    value: "uppercase",
    icon: import_icons.formatUppercase
  },
  {
    label: (0, import_i18n.__)("Lowercase"),
    value: "lowercase",
    icon: import_icons.formatLowercase
  },
  {
    label: (0, import_i18n.__)("Capitalize"),
    value: "capitalize",
    icon: import_icons.formatCapitalize
  }
];
function TextTransformControl({ className, value, onChange }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.__experimentalToggleGroupControl,
    {
      isDeselectable: true,
      __next40pxDefaultSize: true,
      label: (0, import_i18n.__)("Letter case"),
      className: (0, import_clsx.default)(
        "block-editor-text-transform-control",
        className
      ),
      value,
      onChange: (newValue) => {
        onChange(newValue === value ? void 0 : newValue);
      },
      children: TEXT_TRANSFORMS.map((option) => {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.__experimentalToggleGroupControlOptionIcon,
          {
            value: option.value,
            icon: option.icon,
            label: option.label
          },
          option.value
        );
      })
    }
  );
}
