// @wordpress/upload-media/build-module/store/types.mjs
// packages/upload-media/src/store/types.ts
var Type = /* @__PURE__ */ ((Type2) => {
  Type2["Unknown"] = "REDUX_UNKNOWN";
  Type2["Add"] = "ADD_ITEM";
  Type2["Prepare"] = "PREPARE_ITEM";
  Type2["Cancel"] = "CANCEL_ITEM";
  Type2["Remove"] = "REMOVE_ITEM";
  Type2["RetryItem"] = "RETRY_ITEM";
  Type2["PauseItem"] = "PAUSE_ITEM";
  Type2["ResumeItem"] = "RESUME_ITEM";
  Type2["PauseQueue"] = "PAUSE_QUEUE";
  Type2["ResumeQueue"] = "RESUME_QUEUE";
  Type2["OperationStart"] = "OPERATION_START";
  Type2["OperationFinish"] = "OPERATION_FINISH";
  Type2["AddOperations"] = "ADD_OPERATIONS";
  Type2["CacheBlobUrl"] = "CACHE_BLOB_URL";
  Type2["RevokeBlobUrls"] = "REVOKE_BLOB_URLS";
  Type2["UpdateProgress"] = "UPDATE_PROGRESS";
  Type2["UpdateSettings"] = "UPDATE_SETTINGS";
  return Type2;
})(Type || {});
var ItemStatus = /* @__PURE__ */ ((ItemStatus2) => {
  ItemStatus2["Queued"] = "QUEUED";
  ItemStatus2["Processing"] = "PROCESSING";
  ItemStatus2["Paused"] = "PAUSED";
  ItemStatus2["Uploaded"] = "UPLOADED";
  ItemStatus2["Error"] = "ERROR";
  return ItemStatus2;
})(ItemStatus || {});
var OperationType = /* @__PURE__ */ ((OperationType2) => {
  OperationType2["Prepare"] = "PREPARE";
  OperationType2["Upload"] = "UPLOAD";
  OperationType2["ResizeCrop"] = "RESIZE_CROP";
  OperationType2["Rotate"] = "ROTATE";
  OperationType2["TranscodeImage"] = "TRANSCODE_IMAGE";
  OperationType2["ThumbnailGeneration"] = "THUMBNAIL_GENERATION";
  OperationType2["Finalize"] = "FINALIZE";
  return OperationType2;
})(OperationType || {});
export {
  ItemStatus,
  OperationType,
  Type
};
