// .ignored/@wordpress/components/build/menu/checkbox-item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuCheckboxItem = void 0;
var Ariakit = _interopRequireWildcard(require("@ariakit/react"));
var _element = require("@wordpress/element");
var _icons = require("@wordpress/icons");
var _context = require("./context");
var Styled = _interopRequireWildcard(require("./styles"));
var _useTemporaryFocusVisibleFix = require("./use-temporary-focus-visible-fix");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const MenuCheckboxItem = exports.MenuCheckboxItem = (0, _element.forwardRef)(function MenuCheckboxItem({
  suffix,
  children,
  onBlur,
  hideOnClick = false,
  ...props
}, ref) {
  // TODO: Remove when https://github.com/ariakit/ariakit/issues/4083 is fixed
  const focusVisibleFixProps = (0, _useTemporaryFocusVisibleFix.useTemporaryFocusVisibleFix)({
    onBlur
  });
  const menuContext = (0, _element.useContext)(_context.MenuContext);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(Styled.MenuCheckboxItem, {
    ref: ref,
    ...props,
    ...focusVisibleFixProps,
    accessibleWhenDisabled: true,
    hideOnClick: hideOnClick,
    store: menuContext?.store,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(Ariakit.MenuItemCheck, {
      store: menuContext?.store,
      render: /*#__PURE__*/(0, _jsxRuntime.jsx)(Styled.ItemPrefixWrapper, {})
      // Override some ariakit inline styles
      ,
      style: {
        width: 'auto',
        height: 'auto'
      },
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_icons.Icon, {
        icon: _icons.check,
        size: 24
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(Styled.MenuItemContentWrapper, {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(Styled.MenuItemChildrenWrapper, {
        children: children
      }), suffix && /*#__PURE__*/(0, _jsxRuntime.jsx)(Styled.ItemSuffixWrapper, {
        children: suffix
      })]
    })]
  });
});
