// @wordpress/block-editor/build/components/inserter/block-patterns-tab/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/inserter/block-patterns-tab/index.js
var block_patterns_tab_exports = {};
__export(block_patterns_tab_exports, {
  default: () => block_patterns_tab_default
});
module.exports = __toCommonJS(block_patterns_tab_exports);
var import_element = require("@wordpress/element");
var import_compose = require("@wordpress/compose");
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_block_patterns_explorer = __toESM(require("../block-patterns-explorer/index.cjs"));
var import_mobile_tab_navigation = __toESM(require("../mobile-tab-navigation.cjs"));
var import_pattern_category_previews = require("./pattern-category-previews.cjs");
var import_use_pattern_categories = require("./use-pattern-categories.cjs");
var import_category_tabs = __toESM(require("../category-tabs/index.cjs"));
var import_no_results = __toESM(require("../no-results.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function BlockPatternsTab({
  onSelectCategory,
  selectedCategory,
  onInsert,
  rootClientId,
  children
}) {
  const [showPatternsExplorer, setShowPatternsExplorer] = (0, import_element.useState)(false);
  const categories = (0, import_use_pattern_categories.usePatternCategories)(rootClientId);
  const isMobile = (0, import_compose.useViewportMatch)("medium", "<");
  if (!categories.length) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_no_results.default, {});
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    !isMobile && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "block-editor-inserter__block-patterns-tabs-container", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_category_tabs.default,
        {
          categories,
          selectedCategory,
          onSelectCategory,
          children
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.Button,
        {
          __next40pxDefaultSize: true,
          className: "block-editor-inserter__patterns-explore-button",
          onClick: () => setShowPatternsExplorer(true),
          variant: "secondary",
          children: (0, import_i18n.__)("Explore all patterns")
        }
      )
    ] }),
    isMobile && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_mobile_tab_navigation.default, { categories, children: (category) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "block-editor-inserter__category-panel", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_pattern_category_previews.PatternCategoryPreviews,
      {
        onInsert,
        rootClientId,
        category
      },
      category.name
    ) }) }),
    showPatternsExplorer && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_block_patterns_explorer.default,
      {
        initialCategory: selectedCategory || categories[0],
        patternCategories: categories,
        onModalClose: () => setShowPatternsExplorer(false),
        rootClientId
      }
    )
  ] });
}
var block_patterns_tab_default = BlockPatternsTab;
