// @wordpress/blocks/build/api/node.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/blocks/src/api/node.js
var node_exports = {};
__export(node_exports, {
  default: () => node_default,
  fromDOM: () => fromDOM2,
  getNamedNodeMapAsObject: () => getNamedNodeMapAsObject,
  matcher: () => matcher,
  toHTML: () => toHTML2
});
module.exports = __toCommonJS(node_exports);
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var children = __toESM(require("./children.cjs"));
function isNodeOfType(node, type) {
  (0, import_deprecated.default)("wp.blocks.node.isNodeOfType", {
    since: "6.1",
    version: "6.3",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  return node && node.type === type;
}
function getNamedNodeMapAsObject(nodeMap) {
  const result = {};
  for (let i = 0; i < nodeMap.length; i++) {
    const { name, value } = nodeMap[i];
    result[name] = value;
  }
  return result;
}
function fromDOM2(domNode) {
  (0, import_deprecated.default)("wp.blocks.node.fromDOM", {
    since: "6.1",
    version: "6.3",
    alternative: "wp.richText.create",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  if (domNode.nodeType === domNode.TEXT_NODE) {
    return domNode.nodeValue;
  }
  if (domNode.nodeType !== domNode.ELEMENT_NODE) {
    throw new TypeError(
      "A block node can only be created from a node of type text or element."
    );
  }
  return {
    type: domNode.nodeName.toLowerCase(),
    props: {
      ...getNamedNodeMapAsObject(domNode.attributes),
      children: children.fromDOM(domNode.childNodes)
    }
  };
}
function toHTML2(node) {
  (0, import_deprecated.default)("wp.blocks.node.toHTML", {
    since: "6.1",
    version: "6.3",
    alternative: "wp.richText.toHTMLString",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  return children.toHTML([node]);
}
function matcher(selector) {
  (0, import_deprecated.default)("wp.blocks.node.matcher", {
    since: "6.1",
    version: "6.3",
    alternative: "html source",
    link: "https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/introducing-attributes-and-editable-fields/"
  });
  return (domNode) => {
    let match = domNode;
    if (selector) {
      match = domNode.querySelector(selector);
    }
    try {
      return fromDOM2(match);
    } catch (error) {
      return null;
    }
  };
}
var node_default = {
  isNodeOfType,
  fromDOM: fromDOM2,
  toHTML: toHTML2,
  matcher
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  fromDOM,
  getNamedNodeMapAsObject,
  matcher,
  toHTML
});
