// @wordpress/block-editor/build/components/block-heading-level-dropdown/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-heading-level-dropdown/index.js
var block_heading_level_dropdown_exports = {};
__export(block_heading_level_dropdown_exports, {
  default: () => HeadingLevelDropdown
});
module.exports = __toCommonJS(block_heading_level_dropdown_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_heading_level_icon = __toESM(require("./heading-level-icon.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
var HEADING_LEVELS = [1, 2, 3, 4, 5, 6];
var POPOVER_PROPS = {
  className: "block-library-heading-level-dropdown"
};
function HeadingLevelDropdown({
  options = HEADING_LEVELS,
  value,
  onChange
}) {
  const validOptions = options.filter(
    (option) => option === 0 || HEADING_LEVELS.includes(option)
  ).sort((a, b) => a - b);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.ToolbarDropdownMenu,
    {
      popoverProps: POPOVER_PROPS,
      icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_heading_level_icon.default, { level: value }),
      label: (0, import_i18n.__)("Change level"),
      controls: validOptions.map((targetLevel) => {
        const isActive = targetLevel === value;
        return {
          icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_heading_level_icon.default, { level: targetLevel }),
          title: targetLevel === 0 ? (0, import_i18n.__)("Paragraph") : (0, import_i18n.sprintf)(
            // translators: %d: heading level e.g: "1", "2", "3"
            (0, import_i18n.__)("Heading %d"),
            targetLevel
          ),
          isActive,
          onClick() {
            onChange(targetLevel);
          },
          role: "menuitemradio"
        };
      })
    }
  );
}
