// @wordpress/block-editor/build/components/block-compare/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-compare/index.js
var block_compare_exports = {};
__export(block_compare_exports, {
  default: () => block_compare_default
});
module.exports = __toCommonJS(block_compare_exports);
var import_clsx = __toESM(require("clsx"));
var import_character = require("diff/lib/diff/character");
var import_i18n = require("@wordpress/i18n");
var import_blocks = require("@wordpress/blocks");
var import_block_view = __toESM(require("./block-view.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function BlockCompare({
  block,
  onKeep,
  onConvert,
  convertor,
  convertButtonText
}) {
  function getDifference(originalContent, newContent) {
    const difference2 = (0, import_character.diffChars)(originalContent, newContent);
    return difference2.map((item, pos) => {
      const classes = (0, import_clsx.default)({
        "block-editor-block-compare__added": item.added,
        "block-editor-block-compare__removed": item.removed
      });
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: classes, children: item.value }, pos);
    });
  }
  function getConvertedContent(convertedBlock) {
    const newBlocks = Array.isArray(convertedBlock) ? convertedBlock : [convertedBlock];
    const newContent = newBlocks.map(
      (item) => (0, import_blocks.getSaveContent)(item.name, item.attributes, item.innerBlocks)
    );
    return newContent.join("");
  }
  const converted = getConvertedContent(convertor(block));
  const difference = getDifference(block.originalContent, converted);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "block-editor-block-compare__wrapper", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_block_view.default,
      {
        title: (0, import_i18n.__)("Current"),
        className: "block-editor-block-compare__current",
        action: onKeep,
        actionText: (0, import_i18n.__)("Convert to HTML"),
        rawContent: block.originalContent,
        renderedContent: block.originalContent
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_block_view.default,
      {
        title: (0, import_i18n.__)("After Conversion"),
        className: "block-editor-block-compare__converted",
        action: onConvert,
        actionText: convertButtonText,
        rawContent: difference,
        renderedContent: converted
      }
    )
  ] });
}
var block_compare_default = BlockCompare;
