// @wordpress/block-editor/build/components/list-view/leaf.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/list-view/leaf.js
var leaf_exports = {};
__export(leaf_exports, {
  default: () => leaf_default
});
module.exports = __toCommonJS(leaf_exports);
var import_web = require("@react-spring/web");
var import_clsx = __toESM(require("clsx"));
var import_components = require("@wordpress/components");
var import_compose = require("@wordpress/compose");
var import_element = require("@wordpress/element");
var import_use_moving_animation = __toESM(require("../use-moving-animation/index.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
var AnimatedTreeGridRow = (0, import_web.animated)(import_components.__experimentalTreeGridRow);
var ListViewLeaf = (0, import_element.forwardRef)(
  ({
    isDragged,
    isSelected,
    position,
    level,
    rowCount,
    children,
    className,
    path,
    ...props
  }, ref) => {
    const animationRef = (0, import_use_moving_animation.default)({
      clientId: props["data-block"],
      enableAnimation: true,
      triggerAnimationOnChange: path
    });
    const mergedRef = (0, import_compose.useMergeRefs)([ref, animationRef]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      AnimatedTreeGridRow,
      {
        ref: mergedRef,
        className: (0, import_clsx.default)("block-editor-list-view-leaf", className),
        level,
        positionInSet: position,
        setSize: rowCount,
        isExpanded: void 0,
        ...props,
        children
      }
    );
  }
);
var leaf_default = ListViewLeaf;
