// @wordpress/block-editor/build/autocompleters/link.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/autocompleters/link.js
var link_exports = {};
__export(link_exports, {
  default: () => link_default
});
module.exports = __toCommonJS(link_exports);
var import_api_fetch = __toESM(require("@wordpress/api-fetch"));
var import_url = require("@wordpress/url");
var import_icons = require("@wordpress/icons");
var import_html_entities = require("@wordpress/html-entities");
var import_jsx_runtime = require("react/jsx-runtime");
var SHOWN_SUGGESTIONS = 10;
function createLinkCompleter() {
  return {
    name: "links",
    className: "block-editor-autocompleters__link",
    triggerPrefix: "[[",
    options: async (letters) => {
      let options = await (0, import_api_fetch.default)({
        path: (0, import_url.addQueryArgs)("/wp/v2/search", {
          per_page: SHOWN_SUGGESTIONS,
          search: letters,
          type: "post",
          order_by: "menu_order"
        })
      });
      options = options.filter((option) => option.title !== "");
      return options;
    },
    getOptionKeywords(item) {
      const expansionWords = item.title.split(/\s+/);
      return [...expansionWords];
    },
    getOptionLabel(item) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_icons.Icon,
          {
            icon: item.subtype === "page" ? import_icons.page : import_icons.post
          },
          "icon"
        ),
        (0, import_html_entities.decodeEntities)(item.title)
      ] });
    },
    getOptionCompletion(item) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("a", { href: item.url, children: item.title });
    }
  };
}
var link_default = createLinkCompleter();
