// @wordpress/block-library/build-module/separator/deprecated.mjs
// packages/block-library/src/separator/deprecated.js
import clsx from "clsx";
import { getColorClassName, useBlockProps } from "@wordpress/block-editor";
import { jsx } from "react/jsx-runtime";
var v1 = {
  attributes: {
    color: {
      type: "string"
    },
    customColor: {
      type: "string"
    }
  },
  save({ attributes }) {
    const { color, customColor } = attributes;
    const backgroundClass = getColorClassName("background-color", color);
    const colorClass = getColorClassName("color", color);
    const className = clsx({
      "has-text-color has-background": color || customColor,
      [backgroundClass]: backgroundClass,
      [colorClass]: colorClass
    });
    const style = {
      backgroundColor: backgroundClass ? void 0 : customColor,
      color: colorClass ? void 0 : customColor
    };
    return /* @__PURE__ */ jsx("hr", { ...useBlockProps.save({ className, style }) });
  },
  migrate(attributes) {
    const { color, customColor, ...restAttributes } = attributes;
    return {
      ...restAttributes,
      backgroundColor: color ? color : void 0,
      opacity: "css",
      style: customColor ? { color: { background: customColor } } : void 0,
      tagName: "hr"
    };
  }
};
var deprecated_default = [v1];
export {
  deprecated_default as default
};
