// @wordpress/blocks/build/api/validation/logger.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/blocks/src/api/validation/logger.js
var logger_exports = {};
__export(logger_exports, {
  createLogger: () => createLogger,
  createQueuedLogger: () => createQueuedLogger
});
module.exports = __toCommonJS(logger_exports);
function createLogger() {
  function createLogHandler(logger) {
    return (message, ...args) => logger("Block validation: " + message, ...args);
  }
  return {
    // eslint-disable-next-line no-console
    error: createLogHandler(console.error),
    // eslint-disable-next-line no-console
    warning: createLogHandler(console.warn),
    getItems() {
      return [];
    }
  };
}
function createQueuedLogger() {
  const queue = [];
  const logger = createLogger();
  return {
    error(...args) {
      queue.push({ log: logger.error, args });
    },
    warning(...args) {
      queue.push({ log: logger.warning, args });
    },
    getItems() {
      return queue;
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createLogger,
  createQueuedLogger
});
