// @wordpress/block-library@9.42.0/build-module/video/transforms.mjs
// packages/block-library/src/video/transforms.js
import { createBlobURL, isBlobURL } from "@wordpress/blob";
import { createBlock } from "@wordpress/blocks";
var transforms = {
  from: [
    {
      type: "files",
      isMatch(files) {
        return files.length === 1 && files[0].type.indexOf("video/") === 0;
      },
      transform(files) {
        const file = files[0];
        const block = createBlock("core/video", {
          blob: createBlobURL(file)
        });
        return block;
      }
    },
    {
      type: "shortcode",
      tag: "video",
      attributes: {
        src: {
          type: "string",
          shortcode: ({
            named: { src, mp4, m4v, webm, ogv, flv }
          }) => {
            return src || mp4 || m4v || webm || ogv || flv;
          }
        },
        poster: {
          type: "string",
          shortcode: ({ named: { poster } }) => {
            return poster;
          }
        },
        loop: {
          type: "string",
          shortcode: ({ named: { loop } }) => {
            return loop;
          }
        },
        autoplay: {
          type: "string",
          shortcode: ({ named: { autoplay } }) => {
            return autoplay;
          }
        },
        preload: {
          type: "string",
          shortcode: ({ named: { preload } }) => {
            return preload;
          }
        }
      }
    },
    {
      type: "raw",
      isMatch: (node) => node.nodeName === "P" && node.children.length === 1 && node.firstChild.nodeName === "VIDEO",
      transform: (node) => {
        const videoElement = node.firstChild;
        const attributes = {
          autoplay: videoElement.hasAttribute("autoplay") ? true : void 0,
          controls: videoElement.hasAttribute("controls") ? void 0 : false,
          loop: videoElement.hasAttribute("loop") ? true : void 0,
          muted: videoElement.hasAttribute("muted") ? true : void 0,
          preload: videoElement.getAttribute("preload") || void 0,
          playsInline: videoElement.hasAttribute("playsinline") ? true : void 0,
          poster: videoElement.getAttribute("poster") || void 0,
          src: videoElement.getAttribute("src") || void 0
        };
        if (isBlobURL(attributes.src)) {
          attributes.blob = attributes.src;
          delete attributes.src;
        }
        return createBlock("core/video", attributes);
      }
    }
  ]
};
var transforms_default = transforms;
export {
  transforms_default as default
};
