// @wordpress/patterns/build/components/duplicate-pattern-modal.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/patterns/src/components/duplicate-pattern-modal.js
var duplicate_pattern_modal_exports = {};
__export(duplicate_pattern_modal_exports, {
  default: () => DuplicatePatternModal,
  useDuplicatePatternProps: () => useDuplicatePatternProps
});
module.exports = __toCommonJS(duplicate_pattern_modal_exports);
var import_core_data = require("@wordpress/core-data");
var import_data = require("@wordpress/data");
var import_i18n = require("@wordpress/i18n");
var import_notices = require("@wordpress/notices");
var import_create_pattern_modal = __toESM(require("./create-pattern-modal.cjs"));
var import_constants = require("../constants.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function getTermLabels(pattern, categories) {
  if (pattern.type !== import_constants.PATTERN_TYPES.user) {
    return categories.core?.filter(
      (category) => pattern.categories?.includes(category.name)
    ).map((category) => category.label);
  }
  return categories.user?.filter(
    (category) => pattern.wp_pattern_category?.includes(category.id)
  ).map((category) => category.label);
}
function useDuplicatePatternProps({ pattern, onSuccess }) {
  const { createSuccessNotice } = (0, import_data.useDispatch)(import_notices.store);
  const categories = (0, import_data.useSelect)((select) => {
    const { getUserPatternCategories, getBlockPatternCategories } = select(import_core_data.store);
    return {
      core: getBlockPatternCategories(),
      user: getUserPatternCategories()
    };
  });
  if (!pattern) {
    return null;
  }
  return {
    content: pattern.content,
    defaultCategories: getTermLabels(pattern, categories),
    defaultSyncType: pattern.type !== import_constants.PATTERN_TYPES.user ? import_constants.PATTERN_SYNC_TYPES.unsynced : pattern.wp_pattern_sync_status || import_constants.PATTERN_SYNC_TYPES.full,
    defaultTitle: (0, import_i18n.sprintf)(
      /* translators: %s: Existing pattern title */
      (0, import_i18n._x)("%s (Copy)", "pattern"),
      typeof pattern.title === "string" ? pattern.title : pattern.title.raw
    ),
    onSuccess: ({ pattern: newPattern }) => {
      createSuccessNotice(
        (0, import_i18n.sprintf)(
          // translators: %s: The new pattern's title e.g. 'Call to action (copy)'.
          (0, import_i18n._x)('"%s" duplicated.', "pattern"),
          newPattern.title.raw
        ),
        {
          type: "snackbar",
          id: "patterns-create"
        }
      );
      onSuccess?.({ pattern: newPattern });
    }
  };
}
function DuplicatePatternModal({
  pattern,
  onClose,
  onSuccess
}) {
  const duplicatedProps = useDuplicatePatternProps({ pattern, onSuccess });
  if (!pattern) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_create_pattern_modal.default,
    {
      modalTitle: (0, import_i18n.__)("Duplicate pattern"),
      confirmLabel: (0, import_i18n.__)("Duplicate"),
      onClose,
      onError: onClose,
      ...duplicatedProps
    }
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useDuplicatePatternProps
});
