// @wordpress/patterns/build/components/pattern-convert-button.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/patterns/src/components/pattern-convert-button.js
var pattern_convert_button_exports = {};
__export(pattern_convert_button_exports, {
  default: () => PatternConvertButton
});
module.exports = __toCommonJS(pattern_convert_button_exports);
var import_blocks = require("@wordpress/blocks");
var import_block_editor = require("@wordpress/block-editor");
var import_element = require("@wordpress/element");
var import_components = require("@wordpress/components");
var import_icons = require("@wordpress/icons");
var import_data = require("@wordpress/data");
var import_core_data = require("@wordpress/core-data");
var import_i18n = require("@wordpress/i18n");
var import_notices = require("@wordpress/notices");
var import_store = require("../store/index.cjs");
var import_create_pattern_modal = __toESM(require("./create-pattern-modal.cjs"));
var import_lock_unlock = require("../lock-unlock.cjs");
var import_constants = require("../constants.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function PatternConvertButton({
  clientIds,
  rootClientId,
  closeBlockSettingsMenu
}) {
  const { createSuccessNotice } = (0, import_data.useDispatch)(import_notices.store);
  const { replaceBlocks, updateBlockAttributes } = (0, import_data.useDispatch)(import_block_editor.store);
  const { setEditingPattern } = (0, import_lock_unlock.unlock)((0, import_data.useDispatch)(import_store.store));
  const [isModalOpen, setIsModalOpen] = (0, import_element.useState)(false);
  const { getBlockAttributes } = (0, import_data.useSelect)(import_block_editor.store);
  const canConvert = (0, import_data.useSelect)(
    (select) => {
      const { canUser } = select(import_core_data.store);
      const {
        getBlocksByClientId: getBlocksByClientId2,
        canInsertBlockType,
        getBlockRootClientId
      } = select(import_block_editor.store);
      const rootId = rootClientId || (clientIds.length > 0 ? getBlockRootClientId(clientIds[0]) : void 0);
      const blocks = getBlocksByClientId2(clientIds) ?? [];
      const hasReusableBlockSupport = (blockName) => {
        const blockType = (0, import_blocks.getBlockType)(blockName);
        const hasParent = blockType && "parent" in blockType;
        return (0, import_blocks.hasBlockSupport)(blockName, "reusable", !hasParent);
      };
      const isSyncedPattern = blocks.length === 1 && blocks[0] && (0, import_blocks.isReusableBlock)(blocks[0]) && !!select(import_core_data.store).getEntityRecord(
        "postType",
        "wp_block",
        blocks[0].attributes.ref
      );
      const isUnsyncedPattern = blocks.length === 1 && blocks?.[0]?.attributes?.metadata?.patternName;
      const _canConvert = (
        // Hide when this is already a pattern.
        !isUnsyncedPattern && !isSyncedPattern && // Hide when patterns are disabled.
        canInsertBlockType("core/block", rootId) && blocks.every(
          (block) => (
            // Guard against the case where a regular block has *just* been converted.
            !!block && // Hide on invalid blocks.
            block.isValid && // Hide when block doesn't support being made into a pattern.
            hasReusableBlockSupport(block.name)
          )
        ) && // Hide when current doesn't have permission to do that.
        // Blocks refers to the wp_block post type, this checks the ability to create a post of that type.
        !!canUser("create", {
          kind: "postType",
          name: "wp_block"
        })
      );
      return _canConvert;
    },
    [clientIds, rootClientId]
  );
  const { getBlocksByClientId } = (0, import_data.useSelect)(import_block_editor.store);
  const getContent = (0, import_element.useCallback)(
    () => (0, import_blocks.serialize)(getBlocksByClientId(clientIds)),
    [getBlocksByClientId, clientIds]
  );
  if (!canConvert) {
    return null;
  }
  const handleSuccess = ({ pattern }) => {
    if (pattern.wp_pattern_sync_status === import_constants.PATTERN_SYNC_TYPES.unsynced) {
      if (clientIds?.length === 1) {
        const existingAttributes = getBlockAttributes(clientIds[0]);
        updateBlockAttributes(clientIds[0], {
          metadata: {
            ...existingAttributes?.metadata ? existingAttributes.metadata : {},
            patternName: `core/block/${pattern.id}`,
            name: pattern.title.raw
          }
        });
      }
    } else {
      const newBlock = (0, import_blocks.createBlock)("core/block", {
        ref: pattern.id
      });
      replaceBlocks(clientIds, newBlock);
      setEditingPattern(newBlock.clientId, true);
    }
    createSuccessNotice(
      pattern.wp_pattern_sync_status === import_constants.PATTERN_SYNC_TYPES.unsynced ? (0, import_i18n.sprintf)(
        // translators: %s: the name the user has given to the pattern.
        (0, import_i18n.__)("Unsynced pattern created: %s"),
        pattern.title.raw
      ) : (0, import_i18n.sprintf)(
        // translators: %s: the name the user has given to the pattern.
        (0, import_i18n.__)("Synced pattern created: %s"),
        pattern.title.raw
      ),
      {
        type: "snackbar",
        id: "convert-to-pattern-success"
      }
    );
    setIsModalOpen(false);
    closeBlockSettingsMenu();
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.MenuItem,
      {
        icon: import_icons.symbol,
        onClick: () => setIsModalOpen(true),
        "aria-expanded": isModalOpen,
        "aria-haspopup": "dialog",
        children: (0, import_i18n.__)("Create pattern")
      }
    ),
    isModalOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_create_pattern_modal.default,
      {
        content: getContent,
        onSuccess: (pattern) => {
          handleSuccess(pattern);
        },
        onError: () => {
          setIsModalOpen(false);
        },
        onClose: () => {
          setIsModalOpen(false);
          closeBlockSettingsMenu();
        }
      }
    )
  ] });
}
