// @wordpress/patterns/build/components/pattern-overrides-controls.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/patterns/src/components/pattern-overrides-controls.js
var pattern_overrides_controls_exports = {};
__export(pattern_overrides_controls_exports, {
  default: () => pattern_overrides_controls_default
});
module.exports = __toCommonJS(pattern_overrides_controls_exports);
var import_element = require("@wordpress/element");
var import_block_editor = require("@wordpress/block-editor");
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_constants = require("../constants.cjs");
var import_allow_overrides_modal = require("./allow-overrides-modal.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function PatternOverridesControls({
  attributes,
  setAttributes,
  name: blockName
}) {
  const controlId = (0, import_element.useId)();
  const [showAllowOverridesModal, setShowAllowOverridesModal] = (0, import_element.useState)(false);
  const [showDisallowOverridesModal, setShowDisallowOverridesModal] = (0, import_element.useState)(false);
  const hasName = !!attributes.metadata?.name;
  const defaultBindings = attributes.metadata?.bindings?.__default;
  const hasOverrides = hasName && defaultBindings?.source === import_constants.PATTERN_OVERRIDES_BINDING_SOURCE;
  const isConnectedToOtherSources = defaultBindings?.source && defaultBindings.source !== import_constants.PATTERN_OVERRIDES_BINDING_SOURCE;
  const { updateBlockBindings } = (0, import_block_editor.useBlockBindingsUtils)();
  function updateBindings(isChecked, customName) {
    if (customName) {
      setAttributes({
        metadata: {
          ...attributes.metadata,
          name: customName
        }
      });
    }
    updateBlockBindings({
      __default: isChecked ? { source: import_constants.PATTERN_OVERRIDES_BINDING_SOURCE } : void 0
    });
  }
  if (isConnectedToOtherSources) {
    return null;
  }
  const hasUnsupportedImageAttributes = blockName === "core/image" && !!attributes.href?.length;
  const helpText = !hasOverrides && hasUnsupportedImageAttributes ? (0, import_i18n.__)(
    `Overrides currently don't support image links. Remove the link first before enabling overrides.`
  ) : (0, import_i18n.__)(
    "Allow changes to this block throughout instances of this pattern."
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_block_editor.InspectorControls, { group: "advanced", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.BaseControl,
      {
        id: controlId,
        label: (0, import_i18n.__)("Overrides"),
        help: helpText,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.Button,
          {
            __next40pxDefaultSize: true,
            className: "pattern-overrides-control__allow-overrides-button",
            variant: "secondary",
            "aria-haspopup": "dialog",
            onClick: () => {
              if (hasOverrides) {
                setShowDisallowOverridesModal(true);
              } else {
                setShowAllowOverridesModal(true);
              }
            },
            disabled: !hasOverrides && hasUnsupportedImageAttributes,
            accessibleWhenDisabled: true,
            children: hasOverrides ? (0, import_i18n.__)("Disable overrides") : (0, import_i18n.__)("Enable overrides")
          }
        )
      }
    ) }),
    showAllowOverridesModal && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_allow_overrides_modal.AllowOverridesModal,
      {
        initialName: attributes.metadata?.name,
        onClose: () => setShowAllowOverridesModal(false),
        onSave: (newName) => {
          updateBindings(true, newName);
        }
      }
    ),
    showDisallowOverridesModal && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_allow_overrides_modal.DisallowOverridesModal,
      {
        onClose: () => setShowDisallowOverridesModal(false),
        onSave: () => updateBindings(false)
      }
    )
  ] });
}
var pattern_overrides_controls_default = PatternOverridesControls;
