// @wordpress/patterns/build/components/rename-pattern-modal.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/patterns/src/components/rename-pattern-modal.js
var rename_pattern_modal_exports = {};
__export(rename_pattern_modal_exports, {
  default: () => RenamePatternModal
});
module.exports = __toCommonJS(rename_pattern_modal_exports);
var import_components = require("@wordpress/components");
var import_core_data = require("@wordpress/core-data");
var import_data = require("@wordpress/data");
var import_element = require("@wordpress/element");
var import_html_entities = require("@wordpress/html-entities");
var import_i18n = require("@wordpress/i18n");
var import_notices = require("@wordpress/notices");
var import_jsx_runtime = require("react/jsx-runtime");
function RenamePatternModal({
  onClose,
  onError,
  onSuccess,
  pattern,
  ...props
}) {
  const originalName = (0, import_html_entities.decodeEntities)(pattern.title);
  const [name, setName] = (0, import_element.useState)(originalName);
  const [isSaving, setIsSaving] = (0, import_element.useState)(false);
  const {
    editEntityRecord,
    __experimentalSaveSpecifiedEntityEdits: saveSpecifiedEntityEdits
  } = (0, import_data.useDispatch)(import_core_data.store);
  const { createSuccessNotice, createErrorNotice } = (0, import_data.useDispatch)(import_notices.store);
  const onRename = async (event) => {
    event.preventDefault();
    if (!name || name === pattern.title || isSaving) {
      return;
    }
    try {
      await editEntityRecord("postType", pattern.type, pattern.id, {
        title: name
      });
      setIsSaving(true);
      setName("");
      onClose?.();
      const savedRecord = await saveSpecifiedEntityEdits(
        "postType",
        pattern.type,
        pattern.id,
        ["title"],
        { throwOnError: true }
      );
      onSuccess?.(savedRecord);
      createSuccessNotice((0, import_i18n.__)("Pattern renamed"), {
        type: "snackbar",
        id: "pattern-update"
      });
    } catch (error) {
      onError?.();
      const errorMessage = error.message && error.code !== "unknown_error" ? error.message : (0, import_i18n.__)("An error occurred while renaming the pattern.");
      createErrorNotice(errorMessage, {
        type: "snackbar",
        id: "pattern-update"
      });
    } finally {
      setIsSaving(false);
      setName("");
    }
  };
  const onRequestClose = () => {
    onClose?.();
    setName("");
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Modal,
    {
      title: (0, import_i18n.__)("Rename"),
      ...props,
      onRequestClose: onClose,
      focusOnMount: "firstContentElement",
      size: "small",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("form", { onSubmit: onRename, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalVStack, { spacing: "5", children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.TextControl,
          {
            __next40pxDefaultSize: true,
            label: (0, import_i18n.__)("Name"),
            value: name,
            onChange: setName,
            required: true
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalHStack, { justify: "right", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              __next40pxDefaultSize: true,
              variant: "tertiary",
              onClick: onRequestClose,
              children: (0, import_i18n.__)("Cancel")
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              __next40pxDefaultSize: true,
              variant: "primary",
              type: "submit",
              children: (0, import_i18n.__)("Save")
            }
          )
        ] })
      ] }) })
    }
  );
}
