// @wordpress/dataviews/build/components/dataviews-footer/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/dataviews/src/components/dataviews-footer/index.tsx
var dataviews_footer_exports = {};
__export(dataviews_footer_exports, {
  default: () => DataViewsFooter
});
module.exports = __toCommonJS(dataviews_footer_exports);
var import_clsx = __toESM(require("clsx"));
var import_element = require("@wordpress/element");
var import_ui = require("@wordpress/ui");
var import_dataviews_context = __toESM(require("../dataviews-context/index.cjs"));
var import_dataviews_pagination = __toESM(require("../dataviews-pagination/index.cjs"));
var import_dataviews_bulk_actions = require("../dataviews-bulk-actions/index.cjs");
var import_constants = require("../../constants.cjs");
var import_use_delayed_loading = require("../../hooks/use-delayed-loading.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var EMPTY_ARRAY = [];
function DataViewsFooter() {
  const {
    view,
    paginationInfo: { totalItems = 0, totalPages },
    data,
    actions = EMPTY_ARRAY,
    isLoading,
    hasInitiallyLoaded,
    hasInfiniteScrollHandler
  } = (0, import_element.useContext)(import_dataviews_context.default);
  const isRefreshing = !!isLoading && hasInitiallyLoaded && !hasInfiniteScrollHandler && !!data?.length;
  const isDelayedRefreshing = (0, import_use_delayed_loading.useDelayedLoading)(!!isRefreshing);
  const hasBulkActions = (0, import_dataviews_bulk_actions.useSomeItemHasAPossibleBulkAction)(actions, data) && [import_constants.LAYOUT_TABLE, import_constants.LAYOUT_GRID].includes(view.type);
  if (!isRefreshing && (!totalItems || !totalPages || totalPages <= 1 && !hasBulkActions)) {
    return null;
  }
  return (!!totalItems || isRefreshing) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: "dataviews-footer",
      inert: isRefreshing ? "true" : void 0,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_ui.Stack,
        {
          direction: "row",
          justify: "end",
          align: "center",
          className: (0, import_clsx.default)("dataviews-footer__content", {
            "is-refreshing": isDelayedRefreshing
          }),
          gap: "sm",
          children: [
            hasBulkActions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_bulk_actions.BulkActionsFooter, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_pagination.default, {})
          ]
        }
      )
    }
  );
}
