// @wordpress/dataviews/build/components/dataviews-picker-footer/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/dataviews/src/components/dataviews-picker-footer/index.tsx
var dataviews_picker_footer_exports = {};
__export(dataviews_picker_footer_exports, {
  DataViewsPickerFooter: () => DataViewsPickerFooter,
  useIsMultiselectPicker: () => useIsMultiselectPicker
});
module.exports = __toCommonJS(dataviews_picker_footer_exports);
var import_components = require("@wordpress/components");
var import_data = require("@wordpress/data");
var import_element = require("@wordpress/element");
var import_ui = require("@wordpress/ui");
var import_i18n = require("@wordpress/i18n");
var import_dataviews_pagination = __toESM(require("../dataviews-pagination/index.cjs"));
var import_dataviews_context = __toESM(require("../dataviews-context/index.cjs"));
var import_get_footer_message = __toESM(require("../../utils/get-footer-message.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
var EMPTY_ARRAY = [];
function useIsMultiselectPicker(actions) {
  return (0, import_element.useMemo)(() => {
    return actions?.every((action) => action.supportsBulk);
  }, [actions]);
}
function BulkSelectionCheckbox({
  selection,
  selectedItems,
  onChangeSelection,
  data,
  getItemId
}) {
  const areAllSelected = selectedItems.length === data.length;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.CheckboxControl,
    {
      className: "dataviews-view-table-selection-checkbox",
      checked: areAllSelected,
      indeterminate: !areAllSelected && !!selectedItems.length,
      onChange: () => {
        if (areAllSelected) {
          onChangeSelection(
            selection.filter(
              (id) => !data.some(
                (item) => id === getItemId(item)
              )
            )
          );
        } else {
          const selectionSet = /* @__PURE__ */ new Set([
            ...selection,
            ...data.map((item) => getItemId(item))
          ]);
          onChangeSelection(Array.from(selectionSet));
        }
      },
      "aria-label": areAllSelected ? (0, import_i18n.__)("Deselect all") : (0, import_i18n.__)("Select all")
    }
  );
}
function ActionButtons({
  actions,
  items,
  selection
}) {
  const registry = (0, import_data.useRegistry)();
  const [actionInProgress, setActionInProgress] = (0, import_element.useState)(
    null
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ui.Stack, { direction: "row", gap: "xs", children: actions.map((action) => {
    if (!("callback" in action)) {
      return null;
    }
    const { id, label, icon, isPrimary, callback } = action;
    const _label = typeof label === "string" ? label : label(items);
    const variant = isPrimary ? "primary" : "tertiary";
    const isInProgress = id === actionInProgress;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.Button,
      {
        accessibleWhenDisabled: true,
        icon,
        disabled: isInProgress || !selection?.length,
        isBusy: isInProgress,
        onClick: async () => {
          setActionInProgress(id);
          await callback(items, {
            registry
          });
          setActionInProgress(null);
        },
        size: "compact",
        variant,
        children: _label
      },
      id
    );
  }) });
}
function DataViewsPickerFooter() {
  const {
    data,
    selection,
    onChangeSelection,
    getItemId,
    actions = EMPTY_ARRAY,
    paginationInfo
  } = (0, import_element.useContext)(import_dataviews_context.default);
  const isMultiselect = useIsMultiselectPicker(actions);
  const message = (0, import_get_footer_message.default)(
    selection.length,
    data.length,
    paginationInfo.totalItems
  );
  const selectedItems = (0, import_element.useMemo)(
    () => data.filter((item) => selection.includes(getItemId(item))),
    [selection, getItemId, data]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ui.Stack,
    {
      direction: "row",
      justify: "space-between",
      align: "center",
      className: "dataviews-footer",
      gap: "sm",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_ui.Stack,
          {
            direction: "row",
            className: "dataviews-picker-footer__bulk-selection",
            gap: "md",
            align: "center",
            children: [
              isMultiselect && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                BulkSelectionCheckbox,
                {
                  selection,
                  selectedItems,
                  onChangeSelection,
                  data,
                  getItemId
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: "dataviews-bulk-actions-footer__item-count", children: message })
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_dataviews_pagination.default, {}),
        Boolean(actions?.length) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "dataviews-picker-footer__actions", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          ActionButtons,
          {
            actions,
            items: selectedItems,
            selection
          }
        ) })
      ]
    }
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DataViewsPickerFooter,
  useIsMultiselectPicker
});
