// @wordpress/global-styles-engine@1.9.0/build/utils/get-global-styles-changes.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/global-styles-engine/src/utils/get-global-styles-changes.ts
var get_global_styles_changes_exports = {};
__export(get_global_styles_changes_exports, {
  default: () => getGlobalStylesChanges,
  getGlobalStylesChangelist: () => getGlobalStylesChangelist
});
module.exports = __toCommonJS(get_global_styles_changes_exports);
var import_memize = __toESM(require("memize"));
var import_i18n = require("@wordpress/i18n");
var import_blocks = require("@wordpress/blocks");
var globalStylesChangesCache = /* @__PURE__ */ new Map();
var EMPTY_ARRAY = [];
var translationMap = {
  caption: (0, import_i18n.__)("Caption"),
  link: (0, import_i18n.__)("Link"),
  button: (0, import_i18n.__)("Button"),
  heading: (0, import_i18n.__)("Heading"),
  h1: (0, import_i18n.__)("H1"),
  h2: (0, import_i18n.__)("H2"),
  h3: (0, import_i18n.__)("H3"),
  h4: (0, import_i18n.__)("H4"),
  h5: (0, import_i18n.__)("H5"),
  h6: (0, import_i18n.__)("H6"),
  "settings.color": (0, import_i18n.__)("Color"),
  "settings.typography": (0, import_i18n.__)("Typography"),
  "settings.shadow": (0, import_i18n.__)("Shadow"),
  "settings.layout": (0, import_i18n.__)("Layout"),
  "styles.color": (0, import_i18n.__)("Colors"),
  "styles.spacing": (0, import_i18n.__)("Spacing"),
  "styles.background": (0, import_i18n.__)("Background"),
  "styles.typography": (0, import_i18n.__)("Typography")
};
var getBlockNames = (0, import_memize.default)(
  () => (0, import_blocks.getBlockTypes)().reduce(
    (accumulator, {
      name,
      title
    }) => {
      accumulator[name] = title;
      return accumulator;
    },
    {}
  )
);
var isObject = (obj) => obj !== null && typeof obj === "object";
function getTranslation(key) {
  if (translationMap[key]) {
    return translationMap[key];
  }
  const keyArray = key.split(".");
  if (keyArray?.[0] === "blocks") {
    const blockName = getBlockNames()?.[keyArray[1]];
    return blockName || keyArray[1];
  }
  if (keyArray?.[0] === "elements") {
    return translationMap[keyArray[1]] || keyArray[1];
  }
  return void 0;
}
function deepCompare(changedObject, originalObject, parentPath = "") {
  if (!isObject(changedObject) && !isObject(originalObject)) {
    return changedObject !== originalObject ? parentPath.split(".").slice(0, 2).join(".") : void 0;
  }
  changedObject = isObject(changedObject) ? changedObject : {};
  originalObject = isObject(originalObject) ? originalObject : {};
  const allKeys = /* @__PURE__ */ new Set([
    ...Object.keys(changedObject),
    ...Object.keys(originalObject)
  ]);
  let diffs = [];
  for (const key of allKeys) {
    const path = parentPath ? parentPath + "." + key : key;
    const changedPath = deepCompare(
      changedObject[key],
      originalObject[key],
      path
    );
    if (changedPath) {
      diffs = diffs.concat(changedPath);
    }
  }
  return diffs;
}
function getGlobalStylesChangelist(next, previous) {
  const cacheKey = JSON.stringify({ next, previous });
  if (globalStylesChangesCache.has(cacheKey)) {
    return globalStylesChangesCache.get(cacheKey);
  }
  const changedValueTree = deepCompare(
    {
      styles: {
        background: next?.styles?.background,
        color: next?.styles?.color,
        typography: next?.styles?.typography,
        spacing: next?.styles?.spacing
      },
      blocks: next?.styles?.blocks,
      elements: next?.styles?.elements,
      settings: next?.settings
    },
    {
      styles: {
        background: previous?.styles?.background,
        color: previous?.styles?.color,
        typography: previous?.styles?.typography,
        spacing: previous?.styles?.spacing
      },
      blocks: previous?.styles?.blocks,
      elements: previous?.styles?.elements,
      settings: previous?.settings
    }
  );
  if (!changedValueTree || Array.isArray(changedValueTree) && !changedValueTree.length) {
    globalStylesChangesCache.set(cacheKey, []);
    return [];
  }
  const changedValueArray = Array.isArray(changedValueTree) ? changedValueTree : [changedValueTree];
  const result = [...new Set(changedValueArray)].reduce((acc, curr) => {
    const translation = getTranslation(curr);
    if (translation) {
      acc.push([curr.split(".")[0], translation]);
    }
    return acc;
  }, []);
  globalStylesChangesCache.set(cacheKey, result);
  return result;
}
function getGlobalStylesChanges(next, previous, options = {}) {
  let changeList = getGlobalStylesChangelist(next, previous);
  const changesLength = changeList.length;
  const { maxResults } = options;
  if (changesLength) {
    if (!!maxResults && changesLength > maxResults) {
      changeList = changeList.slice(0, maxResults);
    }
    return Object.entries(
      changeList.reduce((acc, curr) => {
        const group = acc[curr[0]] || [];
        if (!group.includes(curr[1])) {
          acc[curr[0]] = [...group, curr[1]];
        }
        return acc;
      }, {})
    ).map(([key, changeValues]) => {
      const changeValuesLength = changeValues.length;
      const joinedChangesValue = changeValues.join(
        /* translators: Used between list items, there is a space after the comma. */
        (0, import_i18n.__)(", ")
        // eslint-disable-line @wordpress/i18n-no-flanking-whitespace
      );
      switch (key) {
        case "blocks": {
          return (0, import_i18n.sprintf)(
            // translators: %s: a list of block names separated by a comma.
            (0, import_i18n._n)("%s block.", "%s blocks.", changeValuesLength),
            joinedChangesValue
          );
        }
        case "elements": {
          return (0, import_i18n.sprintf)(
            // translators: %s: a list of element names separated by a comma.
            (0, import_i18n._n)("%s element.", "%s elements.", changeValuesLength),
            joinedChangesValue
          );
        }
        case "settings": {
          return (0, import_i18n.sprintf)(
            // translators: %s: a list of theme.json setting labels separated by a comma.
            (0, import_i18n.__)("%s settings."),
            joinedChangesValue
          );
        }
        case "styles": {
          return (0, import_i18n.sprintf)(
            // translators: %s: a list of theme.json top-level styles labels separated by a comma.
            (0, import_i18n.__)("%s styles."),
            joinedChangesValue
          );
        }
        default: {
          return (0, import_i18n.sprintf)(
            // translators: %s: a list of global styles changes separated by a comma.
            (0, import_i18n.__)("%s."),
            joinedChangesValue
          );
        }
      }
    });
  }
  return EMPTY_ARRAY;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getGlobalStylesChangelist
});
