// @wordpress/patterns@2.42.0/build/components/category-selector.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/patterns/src/components/category-selector.js
var category_selector_exports = {};
__export(category_selector_exports, {
  CATEGORY_SLUG: () => CATEGORY_SLUG,
  default: () => CategorySelector
});
module.exports = __toCommonJS(category_selector_exports);
var import_i18n = require("@wordpress/i18n");
var import_element = require("@wordpress/element");
var import_components = require("@wordpress/components");
var import_compose = require("@wordpress/compose");
var import_html_entities = require("@wordpress/html-entities");
var import_jsx_runtime = require("react/jsx-runtime");
var unescapeString = (arg) => {
  return (0, import_html_entities.decodeEntities)(arg);
};
var CATEGORY_SLUG = "wp_pattern_category";
function CategorySelector({
  categoryTerms,
  onChange,
  categoryMap
}) {
  const [search, setSearch] = (0, import_element.useState)("");
  const debouncedSearch = (0, import_compose.useDebounce)(setSearch, 500);
  const suggestions = (0, import_element.useMemo)(() => {
    return Array.from(categoryMap.values()).map((category) => unescapeString(category.label)).filter((category) => {
      if (search !== "") {
        return category.toLowerCase().includes(search.toLowerCase());
      }
      return true;
    }).sort((a, b) => a.localeCompare(b));
  }, [search, categoryMap]);
  function handleChange(termNames) {
    const uniqueTerms = termNames.reduce((terms, newTerm) => {
      if (!terms.some(
        (term) => term.toLowerCase() === newTerm.toLowerCase()
      )) {
        terms.push(newTerm);
      }
      return terms;
    }, []);
    onChange(uniqueTerms);
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.FormTokenField,
    {
      className: "patterns-menu-items__convert-modal-categories",
      value: categoryTerms,
      suggestions,
      onChange: handleChange,
      onInputChange: debouncedSearch,
      label: (0, import_i18n.__)("Categories"),
      tokenizeOnBlur: true,
      __experimentalExpandOnFocus: true,
      __next40pxDefaultSize: true
    }
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CATEGORY_SLUG
});
