// @wordpress/patterns@2.42.0/build/components/rename-pattern-category-modal.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/patterns/src/components/rename-pattern-category-modal.js
var rename_pattern_category_modal_exports = {};
__export(rename_pattern_category_modal_exports, {
  default: () => RenamePatternCategoryModal
});
module.exports = __toCommonJS(rename_pattern_category_modal_exports);
var import_components = require("@wordpress/components");
var import_core_data = require("@wordpress/core-data");
var import_data = require("@wordpress/data");
var import_element = require("@wordpress/element");
var import_html_entities = require("@wordpress/html-entities");
var import_i18n = require("@wordpress/i18n");
var import_notices = require("@wordpress/notices");
var import_a11y = require("@wordpress/a11y");
var import_category_selector = require("./category-selector.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function RenamePatternCategoryModal({
  category,
  existingCategories,
  onClose,
  onError,
  onSuccess,
  ...props
}) {
  const id = (0, import_element.useId)();
  const textControlRef = (0, import_element.useRef)();
  const [name, setName] = (0, import_element.useState)((0, import_html_entities.decodeEntities)(category.name));
  const [isSaving, setIsSaving] = (0, import_element.useState)(false);
  const [validationMessage, setValidationMessage] = (0, import_element.useState)(false);
  const validationMessageId = validationMessage ? `patterns-rename-pattern-category-modal__validation-message-${id}` : void 0;
  const { saveEntityRecord, invalidateResolution } = (0, import_data.useDispatch)(import_core_data.store);
  const { createErrorNotice, createSuccessNotice } = (0, import_data.useDispatch)(import_notices.store);
  const onChange = (newName) => {
    if (validationMessage) {
      setValidationMessage(void 0);
    }
    setName(newName);
  };
  const onSave = async (event) => {
    event.preventDefault();
    if (isSaving) {
      return;
    }
    if (!name || name === category.name) {
      const message = (0, import_i18n.__)("Please enter a new name for this category.");
      (0, import_a11y.speak)(message, "assertive");
      setValidationMessage(message);
      textControlRef.current?.focus();
      return;
    }
    if (existingCategories.patternCategories.find((existingCategory) => {
      return existingCategory.id !== category.id && existingCategory.label.toLowerCase() === name.toLowerCase();
    })) {
      const message = (0, import_i18n.__)(
        "This category already exists. Please use a different name."
      );
      (0, import_a11y.speak)(message, "assertive");
      setValidationMessage(message);
      textControlRef.current?.focus();
      return;
    }
    try {
      setIsSaving(true);
      const savedRecord = await saveEntityRecord(
        "taxonomy",
        import_category_selector.CATEGORY_SLUG,
        {
          id: category.id,
          slug: category.slug,
          name
        }
      );
      invalidateResolution("getUserPatternCategories");
      onSuccess?.(savedRecord);
      onClose();
      createSuccessNotice((0, import_i18n.__)("Pattern category renamed."), {
        type: "snackbar",
        id: "pattern-category-update"
      });
    } catch (error) {
      onError?.();
      createErrorNotice(error.message, {
        type: "snackbar",
        id: "pattern-category-update"
      });
    } finally {
      setIsSaving(false);
      setName("");
    }
  };
  const onRequestClose = () => {
    onClose();
    setName("");
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Modal,
    {
      title: (0, import_i18n.__)("Rename"),
      onRequestClose,
      ...props,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("form", { onSubmit: onSave, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalVStack, { spacing: "5", children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalVStack, { spacing: "2", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.TextControl,
            {
              ref: textControlRef,
              __next40pxDefaultSize: true,
              label: (0, import_i18n.__)("Name"),
              value: name,
              onChange,
              "aria-describedby": validationMessageId,
              required: true
            }
          ),
          validationMessage && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "span",
            {
              className: "patterns-rename-pattern-category-modal__validation-message",
              id: validationMessageId,
              children: validationMessage
            }
          )
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalHStack, { justify: "right", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              __next40pxDefaultSize: true,
              variant: "tertiary",
              onClick: onRequestClose,
              children: (0, import_i18n.__)("Cancel")
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Button,
            {
              __next40pxDefaultSize: true,
              variant: "primary",
              type: "submit",
              "aria-disabled": !name || name === category.name || isSaving,
              isBusy: isSaving,
              children: (0, import_i18n.__)("Save")
            }
          )
        ] })
      ] }) })
    }
  );
}
