// @wordpress/block-library/build/terms-query/edit/inspector-controls/inherit-control.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/terms-query/edit/inspector-controls/inherit-control.js
var inherit_control_exports = {};
__export(inherit_control_exports, {
  default: () => InheritControl
});
module.exports = __toCommonJS(inherit_control_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_jsx_runtime = require("react/jsx-runtime");
function InheritControl({ value, onChange, label }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_components.__experimentalToggleGroupControl,
    {
      __next40pxDefaultSize: true,
      label,
      isBlock: true,
      onChange: (newValue) => {
        onChange({
          inherit: newValue === "default"
        });
      },
      help: value ? (0, import_i18n.__)(
        "Display terms based on the current taxonomy archive. For hierarchical taxonomies, shows children of the current term. For non-hierarchical taxonomies, shows all terms."
      ) : (0, import_i18n.__)("Display terms based on specific criteria."),
      value: value ? "default" : "custom",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.__experimentalToggleGroupControlOption,
          {
            value: "default",
            label: (0, import_i18n.__)("Default")
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.__experimentalToggleGroupControlOption, { value: "custom", label: (0, import_i18n.__)("Custom") })
      ]
    }
  );
}
