// @wordpress/reusable-blocks/build/components/reusable-blocks-menu-items/reusable-block-convert-button.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/reusable-blocks/src/components/reusable-blocks-menu-items/reusable-block-convert-button.js
var reusable_block_convert_button_exports = {};
__export(reusable_block_convert_button_exports, {
  default: () => ReusableBlockConvertButton
});
module.exports = __toCommonJS(reusable_block_convert_button_exports);
var import_blocks = require("@wordpress/blocks");
var import_block_editor = require("@wordpress/block-editor");
var import_element = require("@wordpress/element");
var import_components = require("@wordpress/components");
var import_icons = require("@wordpress/icons");
var import_data = require("@wordpress/data");
var import_i18n = require("@wordpress/i18n");
var import_notices = require("@wordpress/notices");
var import_core_data = require("@wordpress/core-data");
var import_store = require("../../store/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function ReusableBlockConvertButton({
  clientIds,
  rootClientId,
  onClose
}) {
  const [syncType, setSyncType] = (0, import_element.useState)(void 0);
  const [isModalOpen, setIsModalOpen] = (0, import_element.useState)(false);
  const [title, setTitle] = (0, import_element.useState)("");
  const canConvert = (0, import_data.useSelect)(
    (select) => {
      const { canUser } = select(import_core_data.store);
      const {
        getBlocksByClientId,
        canInsertBlockType,
        getBlockRootClientId
      } = select(import_block_editor.store);
      const rootId = rootClientId || (clientIds.length > 0 ? getBlockRootClientId(clientIds[0]) : void 0);
      const blocks = getBlocksByClientId(clientIds) ?? [];
      const isReusable = blocks.length === 1 && blocks[0] && (0, import_blocks.isReusableBlock)(blocks[0]) && !!select(import_core_data.store).getEntityRecord(
        "postType",
        "wp_block",
        blocks[0].attributes.ref
      );
      const _canConvert = (
        // Hide when this is already a reusable block.
        !isReusable && // Hide when reusable blocks are disabled.
        canInsertBlockType("core/block", rootId) && blocks.every(
          (block) => (
            // Guard against the case where a regular block has *just* been converted.
            !!block && // Hide on invalid blocks.
            block.isValid && // Hide when block doesn't support being made reusable.
            (0, import_blocks.hasBlockSupport)(block.name, "reusable", true)
          )
        ) && // Hide when current doesn't have permission to do that.
        // Blocks refers to the wp_block post type, this checks the ability to create a post of that type.
        !!canUser("create", {
          kind: "postType",
          name: "wp_block"
        })
      );
      return _canConvert;
    },
    [clientIds, rootClientId]
  );
  const { __experimentalConvertBlocksToReusable: convertBlocksToReusable } = (0, import_data.useDispatch)(import_store.store);
  const { createSuccessNotice, createErrorNotice } = (0, import_data.useDispatch)(import_notices.store);
  const onConvert = (0, import_element.useCallback)(
    async function(reusableBlockTitle) {
      try {
        await convertBlocksToReusable(
          clientIds,
          reusableBlockTitle,
          syncType
        );
        createSuccessNotice(
          !syncType ? (0, import_i18n.sprintf)(
            // translators: %s: the name the user has given to the pattern.
            (0, import_i18n.__)("Synced pattern created: %s"),
            reusableBlockTitle
          ) : (0, import_i18n.sprintf)(
            // translators: %s: the name the user has given to the pattern.
            (0, import_i18n.__)("Unsynced pattern created: %s"),
            reusableBlockTitle
          ),
          {
            type: "snackbar",
            id: "convert-to-reusable-block-success"
          }
        );
      } catch (error) {
        createErrorNotice(error.message, {
          type: "snackbar",
          id: "convert-to-reusable-block-error"
        });
      }
    },
    [
      convertBlocksToReusable,
      clientIds,
      syncType,
      createSuccessNotice,
      createErrorNotice
    ]
  );
  if (!canConvert) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.MenuItem, { icon: import_icons.symbol, onClick: () => setIsModalOpen(true), children: (0, import_i18n.__)("Create pattern") }),
    isModalOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.Modal,
      {
        title: (0, import_i18n.__)("Create pattern"),
        onRequestClose: () => {
          setIsModalOpen(false);
          setTitle("");
        },
        overlayClassName: "reusable-blocks-menu-items__convert-modal",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "form",
          {
            onSubmit: (event) => {
              event.preventDefault();
              onConvert(title);
              setIsModalOpen(false);
              setTitle("");
              onClose();
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalVStack, { spacing: "5", children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_components.TextControl,
                {
                  __next40pxDefaultSize: true,
                  label: (0, import_i18n.__)("Name"),
                  value: title,
                  onChange: setTitle,
                  placeholder: (0, import_i18n.__)("My pattern")
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_components.ToggleControl,
                {
                  label: (0, import_i18n._x)("Synced", "pattern (singular)"),
                  help: (0, import_i18n.__)(
                    "Sync this pattern across multiple locations."
                  ),
                  checked: !syncType,
                  onChange: () => {
                    setSyncType(
                      !syncType ? "unsynced" : void 0
                    );
                  }
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_components.__experimentalHStack, { justify: "right", children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_components.Button,
                  {
                    __next40pxDefaultSize: true,
                    variant: "tertiary",
                    onClick: () => {
                      setIsModalOpen(false);
                      setTitle("");
                    },
                    children: (0, import_i18n.__)("Cancel")
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_components.Button,
                  {
                    __next40pxDefaultSize: true,
                    variant: "primary",
                    type: "submit",
                    children: (0, import_i18n.__)("Create")
                  }
                )
              ] })
            ] })
          }
        )
      }
    )
  ] });
}
