// @wordpress/block-editor/build/components/block-popover/drop-zone.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-popover/drop-zone.js
var drop_zone_exports = {};
__export(drop_zone_exports, {
  default: () => drop_zone_default
});
module.exports = __toCommonJS(drop_zone_exports);
var import_data = require("@wordpress/data");
var import_compose = require("@wordpress/compose");
var import_components = require("@wordpress/components");
var import_store = require("../../store/index.cjs");
var import_cover = __toESM(require("./cover.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
var animateVariants = {
  hide: { opacity: 0, scaleY: 0.75 },
  show: { opacity: 1, scaleY: 1 },
  exit: { opacity: 0, scaleY: 0.9 }
};
function BlockDropZonePopover({
  __unstablePopoverSlot,
  __unstableContentRef
}) {
  const { clientId } = (0, import_data.useSelect)((select) => {
    const { getBlockOrder, getBlockInsertionPoint } = select(import_store.store);
    const insertionPoint = getBlockInsertionPoint();
    const order = getBlockOrder(insertionPoint.rootClientId);
    if (!order.length) {
      return {};
    }
    return {
      clientId: order[insertionPoint.index]
    };
  }, []);
  const reducedMotion = (0, import_compose.useReducedMotion)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_cover.default,
    {
      clientId,
      __unstablePopoverSlot,
      __unstableContentRef,
      className: "block-editor-block-popover__drop-zone",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.__unstableMotion.div,
        {
          "data-testid": "block-popover-drop-zone",
          initial: reducedMotion ? animateVariants.show : animateVariants.hide,
          animate: animateVariants.show,
          exit: reducedMotion ? animateVariants.show : animateVariants.exit,
          className: "block-editor-block-popover__drop-zone-foreground"
        }
      )
    }
  );
}
var drop_zone_default = BlockDropZonePopover;
