// .ignored/@wordpress/edit-post/build/components/keyboard-shortcuts/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/edit-post/src/components/keyboard-shortcuts/index.js
var keyboard_shortcuts_exports = {};
__export(keyboard_shortcuts_exports, {
  default: () => keyboard_shortcuts_default
});
module.exports = __toCommonJS(keyboard_shortcuts_exports);
var import_element = require("@wordpress/element");
var import_data = require("@wordpress/data");
var import_keyboard_shortcuts = require("@wordpress/keyboard-shortcuts");
var import_i18n = require("@wordpress/i18n");
var import_store = require("../../store/index.cjs");
function KeyboardShortcuts() {
  const { toggleFullscreenMode } = (0, import_data.useDispatch)(import_store.store);
  const { registerShortcut } = (0, import_data.useDispatch)(import_keyboard_shortcuts.store);
  (0, import_element.useEffect)(() => {
    registerShortcut({
      name: "core/edit-post/toggle-fullscreen",
      category: "global",
      description: (0, import_i18n.__)("Enable or disable fullscreen mode."),
      keyCombination: {
        modifier: "secondary",
        character: "f"
      }
    });
  }, []);
  (0, import_keyboard_shortcuts.useShortcut)("core/edit-post/toggle-fullscreen", () => {
    toggleFullscreenMode();
  });
  return null;
}
var keyboard_shortcuts_default = KeyboardShortcuts;
