// @wordpress/upload-media/build/image-file.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/upload-media/src/image-file.ts
var image_file_exports = {};
__export(image_file_exports, {
  ImageFile: () => ImageFile
});
module.exports = __toCommonJS(image_file_exports);
var ImageFile = class extends File {
  width = 0;
  height = 0;
  originalWidth = 0;
  originalHeight = 0;
  get wasResized() {
    return (this.originalWidth || 0) > this.width || (this.originalHeight || 0) > this.height;
  }
  constructor(file, width, height, originalWidth, originalHeight) {
    super([file], file.name, {
      type: file.type,
      lastModified: file.lastModified
    });
    this.width = width;
    this.height = height;
    this.originalWidth = originalWidth;
    this.originalHeight = originalHeight;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ImageFile
});
