// @wordpress/block-library/build/tabs/edit.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/tabs/edit.js
var edit_exports = {};
__export(edit_exports, {
  default: () => edit_default
});
module.exports = __toCommonJS(edit_exports);
var import_block_editor = require("@wordpress/block-editor");
var import_data = require("@wordpress/data");
var import_element = require("@wordpress/element");
var import_controls = __toESM(require("./controls.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
var TABS_TEMPLATE = [
  [
    "core/tabs-menu",
    {
      lock: {
        remove: true
      }
    },
    [
      ["core/tabs-menu-item", { anchor: "tab-1-button" }],
      ["core/tabs-menu-item", { anchor: "tab-2-button" }]
    ]
  ],
  [
    "core/tab-panel",
    {
      lock: {
        remove: true
      }
    },
    [
      [
        "core/tab",
        {
          anchor: "tab-1",
          label: "Tab 1"
        },
        [["core/paragraph"]]
      ],
      [
        "core/tab",
        {
          anchor: "tab-2",
          label: "Tab 2"
        },
        [["core/paragraph"]]
      ]
    ]
  ]
];
function Edit({
  clientId,
  attributes,
  setAttributes,
  __unstableLayoutClassNames: layoutClassNames
}) {
  const { anchor, activeTabIndex, editorActiveTabIndex } = attributes;
  (0, import_element.useEffect)(() => {
    if (editorActiveTabIndex === void 0) {
      setAttributes({ editorActiveTabIndex: activeTabIndex });
    }
  }, []);
  const { removeBlock } = (0, import_data.useDispatch)(import_block_editor.store);
  const { tabs, menuItems } = (0, import_data.useSelect)(
    (select) => {
      const { getBlocks } = select(import_block_editor.store);
      const innerBlocks = getBlocks(clientId);
      const tabPanel = innerBlocks.find(
        (block) => block.name === "core/tab-panel"
      );
      const tabsMenu = innerBlocks.find(
        (block) => block.name === "core/tabs-menu"
      );
      return {
        tabs: tabPanel ? tabPanel.innerBlocks.filter(
          (block) => block.name === "core/tab"
        ) : [],
        menuItems: tabsMenu ? getBlocks(tabsMenu.clientId).filter((b) => b.name === "core/tabs-menu-item").map((b) => ({
          clientId: b.clientId,
          anchor: b.attributes.anchor ?? ""
        })) : []
      };
    },
    [clientId]
  );
  const prevSyncStateRef = (0, import_element.useRef)(null);
  (0, import_element.useEffect)(() => {
    const currentTabs = tabs.map((tab) => ({
      clientId: tab.clientId,
      anchor: tab.attributes.anchor ?? ""
    }));
    if (prevSyncStateRef.current === null) {
      prevSyncStateRef.current = {
        tabs: currentTabs,
        menuItems: [...menuItems]
      };
      return;
    }
    const { tabs: prevTabs, menuItems: prevMenuItems } = prevSyncStateRef.current;
    const tabsRemoved = currentTabs.length < prevTabs.length;
    const menuItemsRemoved = menuItems.length < prevMenuItems.length;
    prevSyncStateRef.current = {
      tabs: currentTabs,
      menuItems: [...menuItems]
    };
    if (!tabsRemoved && !menuItemsRemoved || tabsRemoved && menuItemsRemoved) {
      return;
    }
    const currentTabIds = new Set(currentTabs.map((t) => t.clientId));
    const currentMenuItemIds = new Set(
      menuItems.map((m) => m.clientId)
    );
    if (tabsRemoved) {
      prevTabs.forEach((prevTab) => {
        if (currentTabIds.has(prevTab.clientId)) {
          return;
        }
        const expectedMenuAnchor = prevTab.anchor ? `${prevTab.anchor}-button` : null;
        const menuItemToRemove = expectedMenuAnchor ? menuItems.find((m) => m.anchor === expectedMenuAnchor) : null;
        if (menuItemToRemove) {
          removeBlock(menuItemToRemove.clientId, false);
          prevSyncStateRef.current.menuItems = prevSyncStateRef.current.menuItems.filter(
            (m) => m.clientId !== menuItemToRemove.clientId
          );
        }
      });
    } else {
      prevMenuItems.forEach((prevItem) => {
        if (currentMenuItemIds.has(prevItem.clientId)) {
          return;
        }
        const expectedTabAnchor = prevItem.anchor?.replace(/-button$/, "") ?? "";
        const tabToRemove = tabs.find(
          (tab) => (tab.attributes.anchor ?? "") === expectedTabAnchor
        );
        if (tabToRemove) {
          removeBlock(tabToRemove.clientId, false);
          prevSyncStateRef.current.tabs = prevSyncStateRef.current.tabs.filter(
            (t) => t.clientId !== tabToRemove.clientId
          );
        }
      });
    }
  }, [tabs, menuItems, removeBlock]);
  const contextValue = (0, import_element.useMemo)(() => {
    const tabList = tabs.map((tab, index) => ({
      id: tab.attributes.anchor || `tab-${index}`,
      label: tab.attributes.label || "",
      clientId: tab.clientId,
      index
    }));
    return {
      "core/tabs-list": tabList,
      "core/tabs-id": anchor,
      "core/tabs-activeTabIndex": activeTabIndex,
      "core/tabs-editorActiveTabIndex": editorActiveTabIndex
    };
  }, [tabs, anchor, activeTabIndex, editorActiveTabIndex]);
  const blockProps = (0, import_block_editor.useBlockProps)({
    className: layoutClassNames
  });
  const innerBlockProps = (0, import_block_editor.useInnerBlocksProps)(blockProps, {
    __experimentalCaptureToolbars: true,
    template: TABS_TEMPLATE,
    templateLock: false,
    renderAppender: false
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_block_editor.BlockContextProvider, { value: contextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { ...innerBlockProps, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_controls.default,
      {
        clientId,
        attributes,
        setAttributes
      }
    ),
    innerBlockProps.children
  ] }) });
}
var edit_default = Edit;
