// @wordpress/dataviews/build-module/utils/filter-sort-and-paginate.mjs
// packages/dataviews/src/utils/filter-sort-and-paginate.ts
import removeAccents from "remove-accents";
import deprecated from "@wordpress/deprecated";
import { OPERATOR_IS_NOT_ALL } from "../constants.mjs";
import normalizeFields from "../field-types/index.mjs";
function normalizeSearchInput(input = "") {
  return removeAccents(input.trim().toLowerCase());
}
var EMPTY_ARRAY = [];
function filterSortAndPaginate(data, view, fields) {
  if (!data) {
    return {
      data: EMPTY_ARRAY,
      paginationInfo: { totalItems: 0, totalPages: 0 }
    };
  }
  const _fields = normalizeFields(fields);
  let filteredData = [...data];
  if (view.search) {
    const normalizedSearch = normalizeSearchInput(view.search);
    filteredData = filteredData.filter((item) => {
      return _fields.filter((field) => field.enableGlobalSearch).some((field) => {
        const fieldValue = field.getValue({ item });
        const values = Array.isArray(fieldValue) ? fieldValue : [fieldValue];
        return values.some(
          (value) => normalizeSearchInput(String(value)).includes(
            normalizedSearch
          )
        );
      });
    });
  }
  if (view.filters && view.filters?.length > 0) {
    view.filters.forEach((filter) => {
      const field = _fields.find(
        (_field) => _field.id === filter.field
      );
      if (field) {
        if (filter.operator === OPERATOR_IS_NOT_ALL) {
          deprecated("The 'isNotAll' filter operator", {
            since: "7.0",
            alternative: "'isNone'"
          });
        }
        const handler = field.filter[filter.operator];
        if (handler) {
          filteredData = filteredData.filter(
            (item) => handler(item, field, filter.value)
          );
        }
      }
    });
  }
  const sortByField = view.sort?.field ? _fields.find((field) => {
    return field.enableSorting !== false && field.id === view.sort?.field;
  }) : null;
  const groupByField = view.groupBy?.field ? _fields.find((field) => {
    return field.enableSorting !== false && field.id === view.groupBy?.field;
  }) : null;
  if (sortByField || groupByField) {
    filteredData.sort((a, b) => {
      if (groupByField) {
        const groupCompare = groupByField.sort(
          a,
          b,
          view.groupBy?.direction ?? "asc"
        );
        if (groupCompare !== 0) {
          return groupCompare;
        }
      }
      if (sortByField) {
        return sortByField.sort(a, b, view.sort?.direction ?? "desc");
      }
      return 0;
    });
  }
  let totalItems = filteredData.length;
  let totalPages = 1;
  if (view.page !== void 0 && view.perPage !== void 0) {
    const start = (view.page - 1) * view.perPage;
    totalItems = filteredData?.length || 0;
    totalPages = Math.ceil(totalItems / view.perPage);
    filteredData = filteredData?.slice(start, start + view.perPage);
  }
  return {
    data: filteredData,
    paginationInfo: {
      totalItems,
      totalPages
    }
  };
}
export {
  filterSortAndPaginate as default
};
