// @wordpress/block-library@9.42.0/build/template-part/edit/placeholder.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-library/src/template-part/edit/placeholder.js
var placeholder_exports = {};
__export(placeholder_exports, {
  default: () => TemplatePartPlaceholder
});
module.exports = __toCommonJS(placeholder_exports);
var import_i18n = require("@wordpress/i18n");
var import_components = require("@wordpress/components");
var import_element = require("@wordpress/element");
var import_data = require("@wordpress/data");
var import_core_data = require("@wordpress/core-data");
var import_hooks = require("./utils/hooks.cjs");
var import_title_modal = __toESM(require("./title-modal.cjs"));
var import_get_template_part_icon = require("./utils/get-template-part-icon.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function TemplatePartPlaceholder({
  area,
  clientId,
  templatePartId,
  onOpenSelectionModal,
  setAttributes
}) {
  const { templateParts, isResolving } = (0, import_hooks.useAlternativeTemplateParts)(
    area,
    templatePartId
  );
  const blockPatterns = (0, import_hooks.useAlternativeBlockPatterns)(area, clientId);
  const { isBlockBasedTheme, canCreateTemplatePart } = (0, import_data.useSelect)(
    (select) => {
      const { getCurrentTheme, canUser } = select(import_core_data.store);
      return {
        isBlockBasedTheme: getCurrentTheme()?.is_block_theme,
        canCreateTemplatePart: canUser("create", {
          kind: "postType",
          name: "wp_template_part"
        })
      };
    },
    []
  );
  const [showTitleModal, setShowTitleModal] = (0, import_element.useState)(false);
  const areaObject = (0, import_hooks.useTemplatePartArea)(area);
  const createFromBlocks = (0, import_hooks.useCreateTemplatePartFromBlocks)(
    area,
    setAttributes
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_components.Placeholder,
    {
      icon: (0, import_get_template_part_icon.getTemplatePartIcon)(areaObject.icon),
      label: areaObject.label,
      instructions: isBlockBasedTheme ? (0, import_i18n.sprintf)(
        // Translators: %s as template part area title ("Header", "Footer", etc.).
        (0, import_i18n.__)("Choose an existing %s or create a new one."),
        areaObject.label.toLowerCase()
      ) : (0, import_i18n.sprintf)(
        // Translators: %s as template part area title ("Header", "Footer", etc.).
        (0, import_i18n.__)("Choose an existing %s."),
        areaObject.label.toLowerCase()
      ),
      children: [
        isResolving && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.Spinner, {}),
        !isResolving && !!(templateParts.length || blockPatterns.length) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.Button,
          {
            __next40pxDefaultSize: true,
            variant: "primary",
            onClick: onOpenSelectionModal,
            children: (0, import_i18n.__)("Choose")
          }
        ),
        !isResolving && isBlockBasedTheme && canCreateTemplatePart && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_components.Button,
          {
            __next40pxDefaultSize: true,
            variant: "secondary",
            onClick: () => {
              setShowTitleModal(true);
            },
            children: (0, import_i18n.__)("Start blank")
          }
        ),
        showTitleModal && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_title_modal.default,
          {
            areaLabel: areaObject.label,
            onClose: () => setShowTitleModal(false),
            onSubmit: (title) => {
              createFromBlocks([], title);
            }
          }
        )
      ]
    }
  );
}
