// @wordpress/block-editor/build/components/image-editor/zoom-dropdown.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/image-editor/zoom-dropdown.js
var zoom_dropdown_exports = {};
__export(zoom_dropdown_exports, {
  default: () => ZoomDropdown
});
module.exports = __toCommonJS(zoom_dropdown_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_icons = require("@wordpress/icons");
var import_constants = require("./constants.cjs");
var import_context = require("./context.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function ZoomDropdown() {
  const { isInProgress, zoom, setZoom } = (0, import_context.useImageEditingContext)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Dropdown,
    {
      contentClassName: "wp-block-image__zoom",
      popoverProps: import_constants.POPOVER_PROPS,
      renderToggle: ({ isOpen, onToggle }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.ToolbarButton,
        {
          icon: import_icons.search,
          label: (0, import_i18n.__)("Zoom"),
          onClick: onToggle,
          "aria-expanded": isOpen,
          disabled: isInProgress
        }
      ),
      renderContent: () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.__experimentalDropdownContentWrapper, { paddingSize: "medium", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_components.RangeControl,
        {
          __next40pxDefaultSize: true,
          label: (0, import_i18n.__)("Zoom"),
          min: import_constants.MIN_ZOOM,
          max: import_constants.MAX_ZOOM,
          value: Math.round(zoom * 100),
          onChange: (newZoom) => setZoom(newZoom / 100)
        }
      ) })
    }
  );
}
