// @wordpress/block-editor/build/hooks/line-height.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/hooks/line-height.js
var line_height_exports = {};
__export(line_height_exports, {
  LINE_HEIGHT_SUPPORT_KEY: () => LINE_HEIGHT_SUPPORT_KEY,
  LineHeightEdit: () => LineHeightEdit,
  useIsLineHeightDisabled: () => useIsLineHeightDisabled
});
module.exports = __toCommonJS(line_height_exports);
var import_blocks = require("@wordpress/blocks");
var import_line_height_control = __toESM(require("../components/line-height-control/index.cjs"));
var import_utils = require("./utils.cjs");
var import_use_settings = require("../components/use-settings/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var LINE_HEIGHT_SUPPORT_KEY = "typography.lineHeight";
function LineHeightEdit(props) {
  const {
    attributes: { style },
    setAttributes
  } = props;
  const onChange = (newLineHeightValue) => {
    const newStyle = {
      ...style,
      typography: {
        ...style?.typography,
        lineHeight: newLineHeightValue
      }
    };
    setAttributes({ style: (0, import_utils.cleanEmptyObject)(newStyle) });
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_line_height_control.default,
    {
      __unstableInputWidth: "100%",
      value: style?.typography?.lineHeight,
      onChange,
      size: "__unstable-large"
    }
  );
}
function useIsLineHeightDisabled({ name: blockName } = {}) {
  const [isEnabled] = (0, import_use_settings.useSettings)("typography.lineHeight");
  return !isEnabled || !(0, import_blocks.hasBlockSupport)(blockName, LINE_HEIGHT_SUPPORT_KEY);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LINE_HEIGHT_SUPPORT_KEY,
  LineHeightEdit,
  useIsLineHeightDisabled
});
