// @wordpress/fields/build-module/actions/index.mjs
// packages/fields/src/actions/index.ts
import { default as default2 } from "./view-post.mjs";
import { default as default3 } from "./reorder-page.mjs";
import { default as default4 } from "./duplicate-post.mjs";
import { default as default5 } from "./rename-post.mjs";
import { default as default6 } from "./reset-post.mjs";
import { default as default7 } from "./duplicate-pattern.mjs";
import { default as default8 } from "./export-pattern.mjs";
import { default as default9 } from "./view-post-revisions.mjs";
import { default as default10 } from "./permanently-delete-post.mjs";
import { default as default11 } from "./restore-post.mjs";
import { default as default12 } from "./trash-post.mjs";
import { default as default13 } from "./delete-post.mjs";
import { default as default14 } from "./duplicate-template-part.mjs";
export {
  default13 as deletePost,
  default7 as duplicatePattern,
  default4 as duplicatePost,
  default14 as duplicateTemplatePart,
  default8 as exportPattern,
  default10 as permanentlyDeletePost,
  default5 as renamePost,
  default3 as reorderPage,
  default6 as resetPost,
  default11 as restorePost,
  default12 as trashPost,
  default2 as viewPost,
  default9 as viewPostRevisions
};
