// .ignored/@wordpress/core-data/build-module/utils/crdt-utils.mjs
// packages/core-data/src/utils/crdt-utils.ts
import { Y } from "@wordpress/sync";
import { create, insert, toHTMLString } from "@wordpress/rich-text";
import { CRDT_RECORD_MAP_KEY } from "../sync.mjs";
function getRootMap(doc, key) {
  return doc.getMap(key);
}
function createYMap(partial = {}) {
  return new Y.Map(Object.entries(partial));
}
function isYMap(value) {
  return value instanceof Y.Map;
}
function findBlockByClientIdInDoc(blockId, ydoc) {
  const ymap = getRootMap(ydoc, CRDT_RECORD_MAP_KEY);
  const blocks = ymap.get("blocks");
  if (!(blocks instanceof Y.Array)) {
    return null;
  }
  return findBlockByClientIdInBlocks(blockId, blocks);
}
var MARKER_START = 57344;
function pickMarker(text) {
  const tryCount = 16;
  for (let code = MARKER_START; code < MARKER_START + tryCount; code++) {
    const candidate = String.fromCharCode(code);
    if (!text.includes(candidate)) {
      return candidate;
    }
  }
  return null;
}
function htmlIndexToRichTextOffset(html, htmlIndex) {
  if (!html.includes("<") && !html.includes("&")) {
    return htmlIndex;
  }
  const marker = pickMarker(html);
  if (!marker) {
    return htmlIndex;
  }
  const withMarker = html.slice(0, htmlIndex) + marker + html.slice(htmlIndex);
  const value = create({ html: withMarker });
  const markerPos = value.text.indexOf(marker);
  return markerPos === -1 ? htmlIndex : markerPos;
}
function richTextOffsetToHtmlIndex(html, richTextOffset) {
  if (!html.includes("<") && !html.includes("&")) {
    return richTextOffset;
  }
  const marker = pickMarker(html);
  if (!marker) {
    return richTextOffset;
  }
  const value = create({ html });
  const markerValue = create({ text: marker });
  if (value.formats[richTextOffset]) {
    markerValue.formats[0] = value.formats[richTextOffset];
  }
  const withMarker = insert(
    value,
    markerValue,
    richTextOffset,
    richTextOffset
  );
  const htmlWithMarker = toHTMLString({ value: withMarker });
  const markerIndex = htmlWithMarker.indexOf(marker);
  return markerIndex === -1 ? richTextOffset : markerIndex;
}
function findBlockByClientIdInBlocks(blockId, blocks) {
  for (const block of blocks) {
    if (block.get("clientId") === blockId) {
      return block;
    }
    const innerBlocks = block.get("innerBlocks");
    if (innerBlocks && innerBlocks.length > 0) {
      const innerBlock = findBlockByClientIdInBlocks(
        blockId,
        innerBlocks
      );
      if (innerBlock) {
        return innerBlock;
      }
    }
  }
  return null;
}
export {
  createYMap,
  findBlockByClientIdInDoc,
  getRootMap,
  htmlIndexToRichTextOffset,
  isYMap,
  richTextOffsetToHtmlIndex
};
