// .ignored/@wordpress/core-data/build-module/utils/index.mjs
// packages/core-data/src/utils/index.js
import { default as default2 } from "./conservative-map-item.mjs";
import { default as default3 } from "./get-normalized-comma-separable.mjs";
import { default as default4 } from "./if-matching-action.mjs";
import { default as default5 } from "./forward-resolver.mjs";
import { default as default6 } from "./on-sub-key.mjs";
import { default as default7 } from "./replace-action.mjs";
import { default as default8 } from "./with-weak-map-cache.mjs";
import { default as default9 } from "./is-raw-attribute.mjs";
import { default as default10 } from "./set-nested-value.mjs";
import { default as default11 } from "./get-nested-value.mjs";
import { default as default12 } from "./is-numeric-id.mjs";
import {
  getUserPermissionCacheKey,
  getUserPermissionsFromAllowHeader,
  ALLOWED_RESOURCE_ACTIONS
} from "./user-permissions.mjs";
import { RECEIVE_INTERMEDIATE_RESULTS } from "./receive-intermediate-results.mjs";
import { default as default13 } from "./normalize-query-for-resolution.mjs";
export {
  ALLOWED_RESOURCE_ACTIONS,
  RECEIVE_INTERMEDIATE_RESULTS,
  default2 as conservativeMapItem,
  default5 as forwardResolver,
  default11 as getNestedValue,
  default3 as getNormalizedCommaSeparable,
  getUserPermissionCacheKey,
  getUserPermissionsFromAllowHeader,
  default4 as ifMatchingAction,
  default12 as isNumericID,
  default9 as isRawAttribute,
  default13 as normalizeQueryForResolution,
  default6 as onSubKey,
  default7 as replaceAction,
  default10 as setNestedValue,
  default8 as withWeakMapCache
};
