// @wordpress/ui/build/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/index.ts
var index_exports = {};
__export(index_exports, {
  Card: () => Card,
  Collapsible: () => Collapsible,
  CollapsibleCard: () => CollapsibleCard,
  Dialog: () => Dialog,
  Notice: () => Notice,
  Tabs: () => Tabs,
  Tooltip: () => Tooltip
});
module.exports = __toCommonJS(index_exports);
__reExport(index_exports, require("./badge/index.cjs"), module.exports);
__reExport(index_exports, require("./button/index.cjs"), module.exports);
var Card = __toESM(require("./card/index.cjs"));
var Collapsible = __toESM(require("./collapsible/index.cjs"));
var CollapsibleCard = __toESM(require("./collapsible-card/index.cjs"));
var Dialog = __toESM(require("./dialog/index.cjs"));
__reExport(index_exports, require("./form/primitives/index.cjs"), module.exports);
__reExport(index_exports, require("./icon/index.cjs"), module.exports);
__reExport(index_exports, require("./icon-button/index.cjs"), module.exports);
__reExport(index_exports, require("./link/index.cjs"), module.exports);
var Notice = __toESM(require("./notice/index.cjs"));
__reExport(index_exports, require("./stack/index.cjs"), module.exports);
var Tabs = __toESM(require("./tabs/index.cjs"));
__reExport(index_exports, require("./text/index.cjs"), module.exports);
var Tooltip = __toESM(require("./tooltip/index.cjs"));
__reExport(index_exports, require("./visually-hidden/index.cjs"), module.exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Card,
  Collapsible,
  CollapsibleCard,
  Dialog,
  Notice,
  Tabs,
  Tooltip,
  ...require("./badge/index.cjs"),
  ...require("./button/index.cjs"),
  ...require("./form/primitives/index.cjs"),
  ...require("./icon/index.cjs"),
  ...require("./icon-button/index.cjs"),
  ...require("./link/index.cjs"),
  ...require("./stack/index.cjs"),
  ...require("./text/index.cjs"),
  ...require("./visually-hidden/index.cjs")
});
