// @wordpress/blocks/build-module/store/actions.mjs
// packages/blocks/src/store/actions.js
import deprecated from "@wordpress/deprecated";
import { processBlockType } from "./process-block-type.mjs";
function addBlockTypes(blockTypes) {
  return {
    type: "ADD_BLOCK_TYPES",
    blockTypes: Array.isArray(blockTypes) ? blockTypes : [blockTypes]
  };
}
function reapplyBlockTypeFilters() {
  return ({ dispatch, select }) => {
    const processedBlockTypes = [];
    for (const [name, settings] of Object.entries(
      select.getUnprocessedBlockTypes()
    )) {
      const result = dispatch(processBlockType(name, settings));
      if (result) {
        processedBlockTypes.push(result);
      }
    }
    if (!processedBlockTypes.length) {
      return;
    }
    dispatch.addBlockTypes(processedBlockTypes);
  };
}
function __experimentalReapplyBlockFilters() {
  deprecated(
    'wp.data.dispatch( "core/blocks" ).__experimentalReapplyBlockFilters',
    {
      since: "6.4",
      alternative: "reapplyBlockFilters"
    }
  );
  return reapplyBlockTypeFilters();
}
function removeBlockTypes(names) {
  return {
    type: "REMOVE_BLOCK_TYPES",
    names: Array.isArray(names) ? names : [names]
  };
}
function addBlockStyles(blockNames, styles) {
  return {
    type: "ADD_BLOCK_STYLES",
    styles: Array.isArray(styles) ? styles : [styles],
    blockNames: Array.isArray(blockNames) ? blockNames : [blockNames]
  };
}
function removeBlockStyles(blockName, styleNames) {
  return {
    type: "REMOVE_BLOCK_STYLES",
    styleNames: Array.isArray(styleNames) ? styleNames : [styleNames],
    blockName
  };
}
function addBlockVariations(blockName, variations) {
  return {
    type: "ADD_BLOCK_VARIATIONS",
    variations: Array.isArray(variations) ? variations : [variations],
    blockName
  };
}
function removeBlockVariations(blockName, variationNames) {
  return {
    type: "REMOVE_BLOCK_VARIATIONS",
    variationNames: Array.isArray(variationNames) ? variationNames : [variationNames],
    blockName
  };
}
function setDefaultBlockName(name) {
  return {
    type: "SET_DEFAULT_BLOCK_NAME",
    name
  };
}
function setFreeformFallbackBlockName(name) {
  return {
    type: "SET_FREEFORM_FALLBACK_BLOCK_NAME",
    name
  };
}
function setUnregisteredFallbackBlockName(name) {
  return {
    type: "SET_UNREGISTERED_FALLBACK_BLOCK_NAME",
    name
  };
}
function setGroupingBlockName(name) {
  return {
    type: "SET_GROUPING_BLOCK_NAME",
    name
  };
}
function setCategories(categories) {
  return {
    type: "SET_CATEGORIES",
    categories
  };
}
function updateCategory(slug, category) {
  return {
    type: "UPDATE_CATEGORY",
    slug,
    category
  };
}
function addBlockCollection(namespace, title, icon) {
  return {
    type: "ADD_BLOCK_COLLECTION",
    namespace,
    title,
    icon
  };
}
function removeBlockCollection(namespace) {
  return {
    type: "REMOVE_BLOCK_COLLECTION",
    namespace
  };
}
export {
  __experimentalReapplyBlockFilters,
  addBlockCollection,
  addBlockStyles,
  addBlockTypes,
  addBlockVariations,
  reapplyBlockTypeFilters,
  removeBlockCollection,
  removeBlockStyles,
  removeBlockTypes,
  removeBlockVariations,
  setCategories,
  setDefaultBlockName,
  setFreeformFallbackBlockName,
  setGroupingBlockName,
  setUnregisteredFallbackBlockName,
  updateCategory
};
