// @wordpress/media-fields/build-module/index.mjs
// packages/media-fields/src/index.ts
import { default as default2 } from "./alt_text/index.mjs";
import { default as default3 } from "./attached_to/index.mjs";
import { default as default4 } from "./author/index.mjs";
import { default as default5 } from "./caption/index.mjs";
import { default as default6 } from "./date_added/index.mjs";
import { default as default7 } from "./date_modified/index.mjs";
import { default as default8 } from "./description/index.mjs";
import { default as default9 } from "./filename/index.mjs";
import { default as default10 } from "./filesize/index.mjs";
import { default as default11 } from "./media_dimensions/index.mjs";
import { default as default12 } from "./media_thumbnail/index.mjs";
import { default as default13 } from "./mime_type/index.mjs";
export {
  default2 as altTextField,
  default3 as attachedToField,
  default4 as authorField,
  default5 as captionField,
  default6 as dateAddedField,
  default7 as dateModifiedField,
  default8 as descriptionField,
  default9 as filenameField,
  default10 as filesizeField,
  default11 as mediaDimensionsField,
  default12 as mediaThumbnailField,
  default13 as mimeTypeField
};
