// .ignored/@wordpress/core-data/build/selectors.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core-data/src/selectors.ts
var selectors_exports = {};
__export(selectors_exports, {
  __experimentalGetCurrentGlobalStylesId: () => __experimentalGetCurrentGlobalStylesId,
  __experimentalGetCurrentThemeBaseGlobalStyles: () => __experimentalGetCurrentThemeBaseGlobalStyles,
  __experimentalGetCurrentThemeGlobalStylesVariations: () => __experimentalGetCurrentThemeGlobalStylesVariations,
  __experimentalGetDirtyEntityRecords: () => __experimentalGetDirtyEntityRecords,
  __experimentalGetEntitiesBeingSaved: () => __experimentalGetEntitiesBeingSaved,
  __experimentalGetEntityRecordNoResolver: () => __experimentalGetEntityRecordNoResolver,
  canUser: () => canUser,
  canUserEditEntityRecord: () => canUserEditEntityRecord,
  getAuthors: () => getAuthors,
  getAutosave: () => getAutosave,
  getAutosaves: () => getAutosaves,
  getBlockPatternCategories: () => getBlockPatternCategories,
  getBlockPatterns: () => getBlockPatterns,
  getCurrentTheme: () => getCurrentTheme,
  getCurrentThemeGlobalStylesRevisions: () => getCurrentThemeGlobalStylesRevisions,
  getCurrentUser: () => getCurrentUser,
  getDefaultTemplateId: () => getDefaultTemplateId,
  getEditedEntityRecord: () => getEditedEntityRecord,
  getEmbedPreview: () => getEmbedPreview,
  getEntitiesByKind: () => getEntitiesByKind,
  getEntitiesConfig: () => getEntitiesConfig,
  getEntity: () => getEntity,
  getEntityConfig: () => getEntityConfig,
  getEntityRecord: () => getEntityRecord,
  getEntityRecordEdits: () => getEntityRecordEdits,
  getEntityRecordNonTransientEdits: () => getEntityRecordNonTransientEdits,
  getEntityRecords: () => getEntityRecords,
  getEntityRecordsTotalItems: () => getEntityRecordsTotalItems,
  getEntityRecordsTotalPages: () => getEntityRecordsTotalPages,
  getLastEntityDeleteError: () => getLastEntityDeleteError,
  getLastEntitySaveError: () => getLastEntitySaveError,
  getRawEntityRecord: () => getRawEntityRecord,
  getRedoEdit: () => getRedoEdit,
  getReferenceByDistinctEdits: () => getReferenceByDistinctEdits,
  getRevision: () => getRevision,
  getRevisions: () => getRevisions,
  getSyncConnectionStatus: () => getSyncConnectionStatus,
  getThemeSupports: () => getThemeSupports,
  getUndoEdit: () => getUndoEdit,
  getUserPatternCategories: () => getUserPatternCategories,
  getUserQueryResults: () => getUserQueryResults,
  hasEditsForEntityRecord: () => hasEditsForEntityRecord,
  hasEntityRecord: () => hasEntityRecord,
  hasEntityRecords: () => hasEntityRecords,
  hasFetchedAutosaves: () => hasFetchedAutosaves,
  hasRedo: () => hasRedo,
  hasRevision: () => hasRevision,
  hasUndo: () => hasUndo,
  isAutosavingEntityRecord: () => isAutosavingEntityRecord,
  isDeletingEntityRecord: () => isDeletingEntityRecord,
  isPreviewEmbedFallback: () => isPreviewEmbedFallback,
  isRequestingEmbedPreview: () => isRequestingEmbedPreview,
  isSavingEntityRecord: () => isSavingEntityRecord
});
module.exports = __toCommonJS(selectors_exports);
var import_data = require("@wordpress/data");
var import_url = require("@wordpress/url");
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var import_name = require("./name.cjs");
var import_queried_data = require("./queried-data/index.cjs");
var import_entities = require("./entities.cjs");
var import_private_selectors = require("./private-selectors.cjs");
var import_utils = require("./utils/index.cjs");
var import_log_entity_deprecation = __toESM(require("./utils/log-entity-deprecation.cjs"));
var EMPTY_OBJECT = {};
var isRequestingEmbedPreview = (0, import_data.createRegistrySelector)(
  (select) => (state, url) => {
    return select(import_name.STORE_NAME).isResolving("getEmbedPreview", [
      url
    ]);
  }
);
function getAuthors(state, query) {
  (0, import_deprecated.default)("select( 'core' ).getAuthors()", {
    since: "5.9",
    alternative: "select( 'core' ).getUsers({ who: 'authors' })"
  });
  const path = (0, import_url.addQueryArgs)(
    "/wp/v2/users/?who=authors&per_page=100",
    query
  );
  return getUserQueryResults(state, path);
}
function getCurrentUser(state) {
  return state.currentUser;
}
var getUserQueryResults = (0, import_data.createSelector)(
  (state, queryID) => {
    const queryResults = state.users.queries[queryID] ?? [];
    return queryResults.map((id) => state.users.byId[id]);
  },
  (state, queryID) => [
    state.users.queries[queryID],
    state.users.byId
  ]
);
function getEntitiesByKind(state, kind) {
  (0, import_deprecated.default)("wp.data.select( 'core' ).getEntitiesByKind()", {
    since: "6.0",
    alternative: "wp.data.select( 'core' ).getEntitiesConfig()"
  });
  return getEntitiesConfig(state, kind);
}
var getEntitiesConfig = (0, import_data.createSelector)(
  (state, kind) => state.entities.config.filter((entity) => entity.kind === kind),
  /* eslint-disable @typescript-eslint/no-unused-vars */
  (state, kind) => state.entities.config
  /* eslint-enable @typescript-eslint/no-unused-vars */
);
function getEntity(state, kind, name) {
  (0, import_deprecated.default)("wp.data.select( 'core' ).getEntity()", {
    since: "6.0",
    alternative: "wp.data.select( 'core' ).getEntityConfig()"
  });
  return getEntityConfig(state, kind, name);
}
function getEntityConfig(state, kind, name) {
  (0, import_log_entity_deprecation.default)(kind, name, "getEntityConfig");
  return state.entities.config?.find(
    (config) => config.kind === kind && config.name === name
  );
}
var getEntityRecord = (0, import_data.createSelector)(
  ((state, kind, name, key, query) => {
    (0, import_log_entity_deprecation.default)(kind, name, "getEntityRecord");
    const queriedState = state.entities.records?.[kind]?.[name]?.queriedData;
    if (!queriedState) {
      return void 0;
    }
    const context = query?.context ?? "default";
    if (!query || !query._fields) {
      if (!queriedState.itemIsComplete[context]?.[key]) {
        return void 0;
      }
      return queriedState.items[context][key];
    }
    const item = queriedState.items[context]?.[key];
    if (!item) {
      return item;
    }
    const filteredItem = {};
    const fields = (0, import_utils.getNormalizedCommaSeparable)(query._fields) ?? [];
    for (let f = 0; f < fields.length; f++) {
      const field = fields[f].split(".");
      let value = item;
      field.forEach((fieldName) => {
        value = value?.[fieldName];
      });
      (0, import_utils.setNestedValue)(filteredItem, field, value);
    }
    return filteredItem;
  }),
  (state, kind, name, recordId, query) => {
    const context = query?.context ?? "default";
    const queriedState = state.entities.records?.[kind]?.[name]?.queriedData;
    return [
      queriedState?.items[context]?.[recordId],
      queriedState?.itemIsComplete[context]?.[recordId]
    ];
  }
);
getEntityRecord.__unstableNormalizeArgs = (args) => {
  const newArgs = [...args];
  const recordKey = newArgs?.[2];
  newArgs[2] = (0, import_utils.isNumericID)(recordKey) ? Number(recordKey) : recordKey;
  return newArgs;
};
function hasEntityRecord(state, kind, name, key, query) {
  const queriedState = state.entities.records?.[kind]?.[name]?.queriedData;
  if (!queriedState) {
    return false;
  }
  const context = query?.context ?? "default";
  if (!query || !query._fields) {
    return !!queriedState.itemIsComplete[context]?.[key];
  }
  const item = queriedState.items[context]?.[key];
  if (!item) {
    return false;
  }
  const fields = (0, import_utils.getNormalizedCommaSeparable)(query._fields) ?? [];
  for (let i = 0; i < fields.length; i++) {
    const path = fields[i].split(".");
    let value = item;
    for (let p = 0; p < path.length; p++) {
      const part = path[p];
      if (!value || !Object.hasOwn(value, part)) {
        return false;
      }
      value = value[part];
    }
  }
  return true;
}
function __experimentalGetEntityRecordNoResolver(state, kind, name, key) {
  return getEntityRecord(state, kind, name, key);
}
var getRawEntityRecord = (0, import_data.createSelector)(
  (state, kind, name, key) => {
    (0, import_log_entity_deprecation.default)(kind, name, "getRawEntityRecord");
    const record = getEntityRecord(
      state,
      kind,
      name,
      key
    );
    return record && Object.keys(record).reduce((accumulator, _key) => {
      if ((0, import_utils.isRawAttribute)(getEntityConfig(state, kind, name), _key)) {
        accumulator[_key] = record[_key]?.raw !== void 0 ? record[_key]?.raw : record[_key];
      } else {
        accumulator[_key] = record[_key];
      }
      return accumulator;
    }, {});
  },
  (state, kind, name, recordId, query) => {
    const context = query?.context ?? "default";
    return [
      state.entities.config,
      state.entities.records?.[kind]?.[name]?.queriedData?.items[context]?.[recordId],
      state.entities.records?.[kind]?.[name]?.queriedData?.itemIsComplete[context]?.[recordId]
    ];
  }
);
function hasEntityRecords(state, kind, name, query) {
  (0, import_log_entity_deprecation.default)(kind, name, "hasEntityRecords");
  return Array.isArray(getEntityRecords(state, kind, name, query));
}
var getEntityRecords = ((state, kind, name, query) => {
  (0, import_log_entity_deprecation.default)(kind, name, "getEntityRecords");
  const queriedState = state.entities.records?.[kind]?.[name]?.queriedData;
  if (!queriedState) {
    return null;
  }
  return (0, import_queried_data.getQueriedItems)(queriedState, query);
});
var getEntityRecordsTotalItems = (state, kind, name, query) => {
  (0, import_log_entity_deprecation.default)(kind, name, "getEntityRecordsTotalItems");
  const queriedState = state.entities.records?.[kind]?.[name]?.queriedData;
  if (!queriedState) {
    return null;
  }
  return (0, import_queried_data.getQueriedTotalItems)(queriedState, query);
};
var getEntityRecordsTotalPages = (state, kind, name, query) => {
  (0, import_log_entity_deprecation.default)(kind, name, "getEntityRecordsTotalPages");
  const queriedState = state.entities.records?.[kind]?.[name]?.queriedData;
  if (!queriedState) {
    return null;
  }
  if (query?.per_page === -1) {
    return 1;
  }
  const totalItems = (0, import_queried_data.getQueriedTotalItems)(queriedState, query);
  if (!totalItems) {
    return totalItems;
  }
  if (!query?.per_page) {
    return (0, import_queried_data.getQueriedTotalPages)(queriedState, query);
  }
  return Math.ceil(totalItems / query.per_page);
};
var __experimentalGetDirtyEntityRecords = (0, import_data.createSelector)(
  (state) => {
    const {
      entities: { records }
    } = state;
    const dirtyRecords = [];
    Object.keys(records).forEach((kind) => {
      Object.keys(records[kind]).forEach((name) => {
        const primaryKeys = Object.keys(records[kind][name].edits).filter(
          (primaryKey) => (
            // The entity record must exist (not be deleted),
            // and it must have edits.
            getEntityRecord(state, kind, name, primaryKey) && hasEditsForEntityRecord(state, kind, name, primaryKey)
          )
        );
        if (primaryKeys.length) {
          const entityConfig = getEntityConfig(state, kind, name);
          primaryKeys.forEach((primaryKey) => {
            const entityRecord = getEditedEntityRecord(
              state,
              kind,
              name,
              primaryKey
            );
            dirtyRecords.push({
              // We avoid using primaryKey because it's transformed into a string
              // when it's used as an object key.
              key: entityRecord ? entityRecord[entityConfig.key || import_entities.DEFAULT_ENTITY_KEY] : void 0,
              title: entityConfig?.getTitle?.(entityRecord) || "",
              name,
              kind
            });
          });
        }
      });
    });
    return dirtyRecords;
  },
  (state) => [state.entities.records]
);
var __experimentalGetEntitiesBeingSaved = (0, import_data.createSelector)(
  (state) => {
    const {
      entities: { records }
    } = state;
    const recordsBeingSaved = [];
    Object.keys(records).forEach((kind) => {
      Object.keys(records[kind]).forEach((name) => {
        const primaryKeys = Object.keys(records[kind][name].saving).filter(
          (primaryKey) => isSavingEntityRecord(state, kind, name, primaryKey)
        );
        if (primaryKeys.length) {
          const entityConfig = getEntityConfig(state, kind, name);
          primaryKeys.forEach((primaryKey) => {
            const entityRecord = getEditedEntityRecord(
              state,
              kind,
              name,
              primaryKey
            );
            recordsBeingSaved.push({
              // We avoid using primaryKey because it's transformed into a string
              // when it's used as an object key.
              key: entityRecord ? entityRecord[entityConfig.key || import_entities.DEFAULT_ENTITY_KEY] : void 0,
              title: entityConfig?.getTitle?.(entityRecord) || "",
              name,
              kind
            });
          });
        }
      });
    });
    return recordsBeingSaved;
  },
  (state) => [state.entities.records]
);
function getEntityRecordEdits(state, kind, name, recordId) {
  (0, import_log_entity_deprecation.default)(kind, name, "getEntityRecordEdits");
  return state.entities.records?.[kind]?.[name]?.edits?.[recordId];
}
var getEntityRecordNonTransientEdits = (0, import_data.createSelector)(
  (state, kind, name, recordId) => {
    (0, import_log_entity_deprecation.default)(kind, name, "getEntityRecordNonTransientEdits");
    const { transientEdits } = getEntityConfig(state, kind, name) || {};
    const edits = getEntityRecordEdits(state, kind, name, recordId) || {};
    if (!transientEdits) {
      return edits;
    }
    return Object.keys(edits).reduce((acc, key) => {
      if (!transientEdits[key]) {
        acc[key] = edits[key];
      }
      return acc;
    }, {});
  },
  (state, kind, name, recordId) => [
    state.entities.config,
    state.entities.records?.[kind]?.[name]?.edits?.[recordId]
  ]
);
function hasEditsForEntityRecord(state, kind, name, recordId) {
  (0, import_log_entity_deprecation.default)(kind, name, "hasEditsForEntityRecord");
  return isSavingEntityRecord(state, kind, name, recordId) || Object.keys(
    getEntityRecordNonTransientEdits(state, kind, name, recordId)
  ).length > 0;
}
var getEditedEntityRecord = (0, import_data.createSelector)(
  (state, kind, name, recordId) => {
    (0, import_log_entity_deprecation.default)(kind, name, "getEditedEntityRecord");
    const raw = getRawEntityRecord(state, kind, name, recordId);
    const edited = getEntityRecordEdits(state, kind, name, recordId);
    if (!raw && !edited) {
      return false;
    }
    return {
      ...raw,
      ...edited
    };
  },
  (state, kind, name, recordId, query) => {
    const context = query?.context ?? "default";
    return [
      state.entities.config,
      state.entities.records?.[kind]?.[name]?.queriedData.items[context]?.[recordId],
      state.entities.records?.[kind]?.[name]?.queriedData.itemIsComplete[context]?.[recordId],
      state.entities.records?.[kind]?.[name]?.edits?.[recordId]
    ];
  }
);
function isAutosavingEntityRecord(state, kind, name, recordId) {
  (0, import_log_entity_deprecation.default)(kind, name, "isAutosavingEntityRecord");
  const { pending, isAutosave } = state.entities.records?.[kind]?.[name]?.saving?.[recordId] ?? {};
  return Boolean(pending && isAutosave);
}
function isSavingEntityRecord(state, kind, name, recordId) {
  (0, import_log_entity_deprecation.default)(kind, name, "isSavingEntityRecord");
  return state.entities.records?.[kind]?.[name]?.saving?.[recordId]?.pending ?? false;
}
function isDeletingEntityRecord(state, kind, name, recordId) {
  (0, import_log_entity_deprecation.default)(kind, name, "isDeletingEntityRecord");
  return state.entities.records?.[kind]?.[name]?.deleting?.[recordId]?.pending ?? false;
}
function getLastEntitySaveError(state, kind, name, recordId) {
  (0, import_log_entity_deprecation.default)(kind, name, "getLastEntitySaveError");
  return state.entities.records?.[kind]?.[name]?.saving?.[recordId]?.error;
}
function getLastEntityDeleteError(state, kind, name, recordId) {
  (0, import_log_entity_deprecation.default)(kind, name, "getLastEntityDeleteError");
  return state.entities.records?.[kind]?.[name]?.deleting?.[recordId]?.error;
}
function getUndoEdit(state) {
  (0, import_deprecated.default)("select( 'core' ).getUndoEdit()", {
    since: "6.3"
  });
  return void 0;
}
function getRedoEdit(state) {
  (0, import_deprecated.default)("select( 'core' ).getRedoEdit()", {
    since: "6.3"
  });
  return void 0;
}
function hasUndo(state) {
  return (0, import_private_selectors.getUndoManager)(state).hasUndo();
}
function hasRedo(state) {
  return (0, import_private_selectors.getUndoManager)(state).hasRedo();
}
function getCurrentTheme(state) {
  if (!state.currentTheme) {
    return null;
  }
  return getEntityRecord(state, "root", "theme", state.currentTheme);
}
function __experimentalGetCurrentGlobalStylesId(state) {
  return state.currentGlobalStylesId;
}
function getThemeSupports(state) {
  return getCurrentTheme(state)?.theme_supports ?? EMPTY_OBJECT;
}
function getEmbedPreview(state, url) {
  return state.embedPreviews[url];
}
function isPreviewEmbedFallback(state, url) {
  const preview = state.embedPreviews[url];
  const oEmbedLinkCheck = '<a href="' + url + '">' + url + "</a>";
  if (!preview) {
    return false;
  }
  return preview.html === oEmbedLinkCheck;
}
function canUser(state, action, resource, id) {
  const isEntity = typeof resource === "object";
  if (isEntity && (!resource.kind || !resource.name)) {
    return false;
  }
  if (isEntity) {
    (0, import_log_entity_deprecation.default)(resource.kind, resource.name, "canUser");
  }
  const key = (0, import_utils.getUserPermissionCacheKey)(action, resource, id);
  return state.userPermissions[key];
}
function canUserEditEntityRecord(state, kind, name, recordId) {
  (0, import_deprecated.default)(`wp.data.select( 'core' ).canUserEditEntityRecord()`, {
    since: "6.7",
    alternative: `wp.data.select( 'core' ).canUser( 'update', { kind, name, id } )`
  });
  return canUser(state, "update", { kind, name, id: recordId });
}
function getAutosaves(state, postType, postId) {
  return state.autosaves[postId];
}
function getAutosave(state, postType, postId, authorId) {
  if (authorId === void 0) {
    return;
  }
  const autosaves = state.autosaves[postId];
  return autosaves?.find(
    (autosave) => autosave.author === authorId
  );
}
var hasFetchedAutosaves = (0, import_data.createRegistrySelector)(
  (select) => (state, postType, postId) => {
    return select(import_name.STORE_NAME).hasFinishedResolution("getAutosaves", [
      postType,
      postId
    ]);
  }
);
function getReferenceByDistinctEdits(state) {
  return state.editsReference;
}
function __experimentalGetCurrentThemeBaseGlobalStyles(state) {
  const currentTheme = getCurrentTheme(state);
  if (!currentTheme) {
    return null;
  }
  return state.themeBaseGlobalStyles[currentTheme.stylesheet];
}
function __experimentalGetCurrentThemeGlobalStylesVariations(state) {
  const currentTheme = getCurrentTheme(state);
  if (!currentTheme) {
    return null;
  }
  return state.themeGlobalStyleVariations[currentTheme.stylesheet];
}
function getBlockPatterns(state) {
  return state.blockPatterns;
}
function getBlockPatternCategories(state) {
  return state.blockPatternCategories;
}
function getUserPatternCategories(state) {
  return state.userPatternCategories;
}
function getCurrentThemeGlobalStylesRevisions(state) {
  (0, import_deprecated.default)("select( 'core' ).getCurrentThemeGlobalStylesRevisions()", {
    since: "6.5.0",
    alternative: "select( 'core' ).getRevisions( 'root', 'globalStyles', ${ recordKey } )"
  });
  const currentGlobalStylesId = __experimentalGetCurrentGlobalStylesId(state);
  if (!currentGlobalStylesId) {
    return null;
  }
  return state.themeGlobalStyleRevisions[currentGlobalStylesId];
}
function getDefaultTemplateId(state, query) {
  return state.defaultTemplates[JSON.stringify(query)];
}
var getRevisions = (state, kind, name, recordKey, query) => {
  (0, import_log_entity_deprecation.default)(kind, name, "getRevisions");
  const queriedStateRevisions = state.entities.records?.[kind]?.[name]?.revisions?.[recordKey];
  if (!queriedStateRevisions) {
    return null;
  }
  return (0, import_queried_data.getQueriedItems)(queriedStateRevisions, query);
};
function hasRevision(state, kind, name, recordKey, revisionKey, query) {
  const queriedState = state.entities.records?.[kind]?.[name]?.revisions?.[recordKey];
  if (!queriedState) {
    return false;
  }
  const context = query?.context ?? "default";
  if (!query || !query._fields) {
    return !!queriedState.itemIsComplete[context]?.[revisionKey];
  }
  const item = queriedState.items[context]?.[revisionKey];
  if (!item) {
    return false;
  }
  const fields = (0, import_utils.getNormalizedCommaSeparable)(query._fields) ?? [];
  for (let i = 0; i < fields.length; i++) {
    const path = fields[i].split(".");
    let value = item;
    for (let p = 0; p < path.length; p++) {
      const part = path[p];
      if (!value || !Object.hasOwn(value, part)) {
        return false;
      }
      value = value[part];
    }
  }
  return true;
}
var getRevision = (0, import_data.createSelector)(
  (state, kind, name, recordKey, revisionKey, query) => {
    (0, import_log_entity_deprecation.default)(kind, name, "getRevision");
    const queriedState = state.entities.records?.[kind]?.[name]?.revisions?.[recordKey];
    if (!queriedState) {
      return void 0;
    }
    const context = query?.context ?? "default";
    if (!query || !query._fields) {
      if (!queriedState.itemIsComplete[context]?.[revisionKey]) {
        return void 0;
      }
      return queriedState.items[context][revisionKey];
    }
    const item = queriedState.items[context]?.[revisionKey];
    if (!item) {
      return item;
    }
    const filteredItem = {};
    const fields = (0, import_utils.getNormalizedCommaSeparable)(query._fields) ?? [];
    for (let f = 0; f < fields.length; f++) {
      const field = fields[f].split(".");
      let value = item;
      field.forEach((fieldName) => {
        value = value?.[fieldName];
      });
      (0, import_utils.setNestedValue)(filteredItem, field, value);
    }
    return filteredItem;
  },
  (state, kind, name, recordKey, revisionKey, query) => {
    const context = query?.context ?? "default";
    const queriedState = state.entities.records?.[kind]?.[name]?.revisions?.[recordKey];
    return [
      queriedState?.items?.[context]?.[revisionKey],
      queriedState?.itemIsComplete?.[context]?.[revisionKey]
    ];
  }
);
function getSyncConnectionStatus(state) {
  if (!state.syncConnectionStatuses) {
    return void 0;
  }
  const PRIORITIZED_STATUSES = ["disconnected", "connecting", "connected"];
  let coalesced;
  for (const status of Object.values(state.syncConnectionStatuses)) {
    if (!coalesced || PRIORITIZED_STATUSES.indexOf(status.status) < PRIORITIZED_STATUSES.indexOf(coalesced.status)) {
      coalesced = status;
    }
  }
  return coalesced;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  __experimentalGetCurrentGlobalStylesId,
  __experimentalGetCurrentThemeBaseGlobalStyles,
  __experimentalGetCurrentThemeGlobalStylesVariations,
  __experimentalGetDirtyEntityRecords,
  __experimentalGetEntitiesBeingSaved,
  __experimentalGetEntityRecordNoResolver,
  canUser,
  canUserEditEntityRecord,
  getAuthors,
  getAutosave,
  getAutosaves,
  getBlockPatternCategories,
  getBlockPatterns,
  getCurrentTheme,
  getCurrentThemeGlobalStylesRevisions,
  getCurrentUser,
  getDefaultTemplateId,
  getEditedEntityRecord,
  getEmbedPreview,
  getEntitiesByKind,
  getEntitiesConfig,
  getEntity,
  getEntityConfig,
  getEntityRecord,
  getEntityRecordEdits,
  getEntityRecordNonTransientEdits,
  getEntityRecords,
  getEntityRecordsTotalItems,
  getEntityRecordsTotalPages,
  getLastEntityDeleteError,
  getLastEntitySaveError,
  getRawEntityRecord,
  getRedoEdit,
  getReferenceByDistinctEdits,
  getRevision,
  getRevisions,
  getSyncConnectionStatus,
  getThemeSupports,
  getUndoEdit,
  getUserPatternCategories,
  getUserQueryResults,
  hasEditsForEntityRecord,
  hasEntityRecord,
  hasEntityRecords,
  hasFetchedAutosaves,
  hasRedo,
  hasRevision,
  hasUndo,
  isAutosavingEntityRecord,
  isDeletingEntityRecord,
  isPreviewEmbedFallback,
  isRequestingEmbedPreview,
  isSavingEntityRecord
});
